{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9b7b095d-dfd2-430d-85a3-dbc51e4cc9af",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:23:48.589147Z",
     "iopub.status.busy": "2023-05-22T07:23:48.588127Z",
     "iopub.status.idle": "2023-05-22T07:23:54.158775Z",
     "shell.execute_reply": "2023-05-22T07:23:54.156708Z",
     "shell.execute_reply.started": "2023-05-22T07:23:48.589113Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from pybedtools import BedTool"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59e21d55-3103-487a-b764-9eb0d67da344",
   "metadata": {},
   "source": [
    "# 查看ATAC-seq在目标位置的信号变化"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "724976cc-9ed5-4276-8a11-db409fba4ad6",
   "metadata": {},
   "source": [
    "## 准备df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60c4497a-9596-45f2-ba19-9dcb9797ac3b",
   "metadata": {},
   "source": [
    "### df_ctcf\n",
    "- weak\n",
    "- medium\n",
    "- strong\n",
    "- overlapped with offs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "62b7018c-78cd-483c-b767-665fbb81ec26",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:25:54.219288Z",
     "iopub.status.busy": "2023-05-22T07:25:54.218543Z",
     "iopub.status.idle": "2023-05-22T07:25:55.102906Z",
     "shell.execute_reply": "2023-05-22T07:25:55.100940Z",
     "shell.execute_reply.started": "2023-05-22T07:25:54.219228Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr11</td>\n",
       "      <td>28107123</td>\n",
       "      <td>28107196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>127523472</td>\n",
       "      <td>127523528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr10</td>\n",
       "      <td>79996559</td>\n",
       "      <td>79996604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr20</td>\n",
       "      <td>36808712</td>\n",
       "      <td>36808741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr9</td>\n",
       "      <td>115952242</td>\n",
       "      <td>115952303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15392</th>\n",
       "      <td>chr1</td>\n",
       "      <td>148725246</td>\n",
       "      <td>148725266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15393</th>\n",
       "      <td>chr8</td>\n",
       "      <td>81158725</td>\n",
       "      <td>81158758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15394</th>\n",
       "      <td>chr8</td>\n",
       "      <td>100191816</td>\n",
       "      <td>100191830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15395</th>\n",
       "      <td>chr17</td>\n",
       "      <td>32831978</td>\n",
       "      <td>32831991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15396</th>\n",
       "      <td>chrX</td>\n",
       "      <td>79459870</td>\n",
       "      <td>79459896</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12809 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom      start        end\n",
       "0      chr11   28107123   28107196\n",
       "1       chr4  127523472  127523528\n",
       "2      chr10   79996559   79996604\n",
       "3      chr20   36808712   36808741\n",
       "4       chr9  115952242  115952303\n",
       "...      ...        ...        ...\n",
       "15392   chr1  148725246  148725266\n",
       "15393   chr8   81158725   81158758\n",
       "15394   chr8  100191816  100191830\n",
       "15395  chr17   32831978   32831991\n",
       "15396   chrX   79459870   79459896\n",
       "\n",
       "[12809 rows x 3 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_new_offs = pd.read_csv(\n",
    "    '/lustre1/chengqiyi_pkuhpc/zhaohn/3.project/2022_DdCBE-3D-Genome_topic/'\n",
    "    '2022-09-30_Detect-seq_batch-1_ATP8_JAK2_SIRT6/final_list_after_igv_check/'\n",
    "    '2023-04-24_merged_final_list_add_header_poisson_result.csv',\n",
    "    header=0, \n",
    ")\n",
    "df_new_offs = df_new_offs[['chr_name', 'region_start', 'region_end']].copy()\n",
    "df_new_offs.columns = ['chrom', 'start', 'end']\n",
    "df_new_offs.drop_duplicates(keep='first', inplace=True)\n",
    "df_new_offs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "735b03e2-2892-4f31-a9e7-cbd131a1ced9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:25:58.983830Z",
     "iopub.status.busy": "2023-05-22T07:25:58.983210Z",
     "iopub.status.idle": "2023-05-22T07:25:59.104879Z",
     "shell.execute_reply": "2023-05-22T07:25:59.103139Z",
     "shell.execute_reply.started": "2023-05-22T07:25:58.983773Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2044988</td>\n",
       "      <td>2044998</td>\n",
       "      <td>ND5.1-TAS.IND-1_chr1_2044988_2044998</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9702414</td>\n",
       "      <td>9702449</td>\n",
       "      <td>ND5.1-TAS.IND-2_ND6-TAS.IND-1_chr1_9702414_970...</td>\n",
       "      <td>11</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12618165</td>\n",
       "      <td>12618174</td>\n",
       "      <td>ND5.1-TAS.IND-3_chr1_12618165_12618174</td>\n",
       "      <td>5</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13019998</td>\n",
       "      <td>13020019</td>\n",
       "      <td>ND6-TAS.IND-2_chr1_13019998_13020019</td>\n",
       "      <td>9</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13354705</td>\n",
       "      <td>13354738</td>\n",
       "      <td>ND6-TAS.IND-3_chr1_13354705_13354738</td>\n",
       "      <td>11</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1027</th>\n",
       "      <td>chrX</td>\n",
       "      <td>136925330</td>\n",
       "      <td>136925338</td>\n",
       "      <td>ND6-TAS.IND-538_chrX_136925330_136925338</td>\n",
       "      <td>5</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1028</th>\n",
       "      <td>chrX</td>\n",
       "      <td>138128211</td>\n",
       "      <td>138128237</td>\n",
       "      <td>ND6-TAS.IND-539_chrX_138128211_138128237</td>\n",
       "      <td>9</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1029</th>\n",
       "      <td>chrX</td>\n",
       "      <td>150683735</td>\n",
       "      <td>150683758</td>\n",
       "      <td>ND5.1-TAS.IND-452_ND6-TAS.IND-540_chrX_1506837...</td>\n",
       "      <td>11</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153669436</td>\n",
       "      <td>153669453</td>\n",
       "      <td>ND5.1-TAS.IND-453_ND6-TAS.IND-541_chrX_1536694...</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1034</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155049205</td>\n",
       "      <td>155049215</td>\n",
       "      <td>ND5.1-TAS.IND-454_ND6-TAS.IND-542_chrX_1550492...</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>650 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end   \n",
       "0     chr1    2044988    2044998  \\\n",
       "1     chr1    9702414    9702449   \n",
       "2     chr1   12618165   12618174   \n",
       "3     chr1   13019998   13020019   \n",
       "4     chr1   13354705   13354738   \n",
       "...    ...        ...        ...   \n",
       "1027  chrX  136925330  136925338   \n",
       "1028  chrX  138128211  138128237   \n",
       "1029  chrX  150683735  150683758   \n",
       "1032  chrX  153669436  153669453   \n",
       "1034  chrX  155049205  155049215   \n",
       "\n",
       "                                                   name  score strand  \n",
       "0                  ND5.1-TAS.IND-1_chr1_2044988_2044998      7      +  \n",
       "1     ND5.1-TAS.IND-2_ND6-TAS.IND-1_chr1_9702414_970...     11      -  \n",
       "2                ND5.1-TAS.IND-3_chr1_12618165_12618174      5      +  \n",
       "3                  ND6-TAS.IND-2_chr1_13019998_13020019      9      -  \n",
       "4                  ND6-TAS.IND-3_chr1_13354705_13354738     11      +  \n",
       "...                                                 ...    ...    ...  \n",
       "1027           ND6-TAS.IND-538_chrX_136925330_136925338      5      +  \n",
       "1028           ND6-TAS.IND-539_chrX_138128211_138128237      9      -  \n",
       "1029  ND5.1-TAS.IND-452_ND6-TAS.IND-540_chrX_1506837...     11      -  \n",
       "1032  ND5.1-TAS.IND-453_ND6-TAS.IND-541_chrX_1536694...     10      +  \n",
       "1034  ND5.1-TAS.IND-454_ND6-TAS.IND-542_chrX_1550492...      7      -  \n",
       "\n",
       "[650 rows x 6 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_old_offs = pd.read_csv(\n",
    "    '/lustre1/chengqiyi_pkuhpc/zhaohn/3.project/2022_DdCBE-3D-Genome_topic/'\n",
    "    '2022-09-30_Detect-seq_batch-1_ATP8_JAK2_SIRT6/tables/'\n",
    "    '2022-10-05_WT_ND4-5.1-6_final-list.bed',\n",
    "    header=None, sep='\\t', names=['chrom', 'start', 'end', 'name', 'score', 'strand']\n",
    ")\n",
    "df_old_offs = df_old_offs.query('name.str.contains(\"IND\")').copy()\n",
    "df_old_offs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f7b40a50-6e24-4c36-b666-1248c6351123",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:25:59.795798Z",
     "iopub.status.busy": "2023-05-22T07:25:59.795198Z",
     "iopub.status.idle": "2023-05-22T07:25:59.820344Z",
     "shell.execute_reply": "2023-05-22T07:25:59.818813Z",
     "shell.execute_reply.started": "2023-05-22T07:25:59.795743Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2044988</td>\n",
       "      <td>2044998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9702414</td>\n",
       "      <td>9702449</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12618165</td>\n",
       "      <td>12618174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13019998</td>\n",
       "      <td>13020019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13354705</td>\n",
       "      <td>13354738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15392</th>\n",
       "      <td>chr1</td>\n",
       "      <td>148725246</td>\n",
       "      <td>148725266</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15393</th>\n",
       "      <td>chr8</td>\n",
       "      <td>81158725</td>\n",
       "      <td>81158758</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15394</th>\n",
       "      <td>chr8</td>\n",
       "      <td>100191816</td>\n",
       "      <td>100191830</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15395</th>\n",
       "      <td>chr17</td>\n",
       "      <td>32831978</td>\n",
       "      <td>32831991</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15396</th>\n",
       "      <td>chrX</td>\n",
       "      <td>79459870</td>\n",
       "      <td>79459896</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13459 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom      start        end\n",
       "0       chr1    2044988    2044998\n",
       "1       chr1    9702414    9702449\n",
       "2       chr1   12618165   12618174\n",
       "3       chr1   13019998   13020019\n",
       "4       chr1   13354705   13354738\n",
       "...      ...        ...        ...\n",
       "15392   chr1  148725246  148725266\n",
       "15393   chr8   81158725   81158758\n",
       "15394   chr8  100191816  100191830\n",
       "15395  chr17   32831978   32831991\n",
       "15396   chrX   79459870   79459896\n",
       "\n",
       "[13459 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_offs = pd.concat(\n",
    "    [\n",
    "        df_old_offs.iloc[:,:3],\n",
    "        df_new_offs.iloc[:,:3]\n",
    "    ]\n",
    ")\n",
    "df_offs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c81f9922-79d4-4d4d-9efd-239ce7bf84f2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:26:01.405038Z",
     "iopub.status.busy": "2023-05-22T07:26:01.404448Z",
     "iopub.status.idle": "2023-05-22T07:26:01.597810Z",
     "shell.execute_reply": "2023-05-22T07:26:01.596171Z",
     "shell.execute_reply.started": "2023-05-22T07:26:01.404983Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>RPKM</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>267790</td>\n",
       "      <td>268214</td>\n",
       "      <td>chr1_267790_268214_weak_peak</td>\n",
       "      <td>1.568319</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>778701</td>\n",
       "      <td>779047</td>\n",
       "      <td>chr1_778701_779047_strong_peak</td>\n",
       "      <td>5.326924</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>778863</td>\n",
       "      <td>779287</td>\n",
       "      <td>chr1_778863_779287_medium_peak</td>\n",
       "      <td>2.778652</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>869675</td>\n",
       "      <td>870099</td>\n",
       "      <td>chr1_869675_870099_weak_peak</td>\n",
       "      <td>1.125098</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>904753</td>\n",
       "      <td>904883</td>\n",
       "      <td>chr1_904753_904883_strong_peak</td>\n",
       "      <td>6.115919</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38030</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155264125</td>\n",
       "      <td>155264549</td>\n",
       "      <td>chrX_155264125_155264549_medium_peak</td>\n",
       "      <td>3.204826</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38031</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155686260</td>\n",
       "      <td>155686684</td>\n",
       "      <td>chrX_155686260_155686684_weak_peak</td>\n",
       "      <td>1.295568</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38032</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155820126</td>\n",
       "      <td>155820550</td>\n",
       "      <td>chrX_155820126_155820550_medium_peak</td>\n",
       "      <td>3.426436</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38033</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155881180</td>\n",
       "      <td>155881604</td>\n",
       "      <td>chrX_155881180_155881604_medium_peak</td>\n",
       "      <td>3.068450</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38034</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155966835</td>\n",
       "      <td>155967259</td>\n",
       "      <td>chrX_155966835_155967259_medium_peak</td>\n",
       "      <td>2.471807</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38035 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom      start        end                                  name   \n",
       "0      chr1     267790     268214          chr1_267790_268214_weak_peak  \\\n",
       "1      chr1     778701     779047        chr1_778701_779047_strong_peak   \n",
       "2      chr1     778863     779287        chr1_778863_779287_medium_peak   \n",
       "3      chr1     869675     870099          chr1_869675_870099_weak_peak   \n",
       "4      chr1     904753     904883        chr1_904753_904883_strong_peak   \n",
       "...     ...        ...        ...                                   ...   \n",
       "38030  chrX  155264125  155264549  chrX_155264125_155264549_medium_peak   \n",
       "38031  chrX  155686260  155686684    chrX_155686260_155686684_weak_peak   \n",
       "38032  chrX  155820126  155820550  chrX_155820126_155820550_medium_peak   \n",
       "38033  chrX  155881180  155881604  chrX_155881180_155881604_medium_peak   \n",
       "38034  chrX  155966835  155967259  chrX_155966835_155967259_medium_peak   \n",
       "\n",
       "           RPKM strand  \n",
       "0      1.568319      .  \n",
       "1      5.326924      .  \n",
       "2      2.778652      .  \n",
       "3      1.125098      .  \n",
       "4      6.115919      .  \n",
       "...         ...    ...  \n",
       "38030  3.204826      .  \n",
       "38031  1.295568      .  \n",
       "38032  3.426436      .  \n",
       "38033  3.068450      .  \n",
       "38034  2.471807      .  \n",
       "\n",
       "[38035 rows x 6 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ctcf = pd.read_csv('/lustre1/chengqiyi_pkuhpc/zhaohn/1.database/public_data/CTCF_ATAC_data/ENCFF285QVL_CTCF_binding_sites_fix_range.bed',\n",
    "                     sep='\\t')\n",
    "df_ctcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "8376cc24-fc42-405b-8fb9-b7b30aa7396f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:26:11.644710Z",
     "iopub.status.busy": "2023-05-22T07:26:11.644065Z",
     "iopub.status.idle": "2023-05-22T07:26:12.291195Z",
     "shell.execute_reply": "2023-05-22T07:26:12.290223Z",
     "shell.execute_reply.started": "2023-05-22T07:26:11.644653Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>RPKM</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>267790</td>\n",
       "      <td>268214</td>\n",
       "      <td>chr1_267790_268214_weak_peak</td>\n",
       "      <td>1.568319</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>778701</td>\n",
       "      <td>779047</td>\n",
       "      <td>chr1_778701_779047_strong_peak</td>\n",
       "      <td>5.326924</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>778863</td>\n",
       "      <td>779287</td>\n",
       "      <td>chr1_778863_779287_medium_peak</td>\n",
       "      <td>2.778652</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>869675</td>\n",
       "      <td>870099</td>\n",
       "      <td>chr1_869675_870099_weak_peak</td>\n",
       "      <td>1.125098</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>904753</td>\n",
       "      <td>904883</td>\n",
       "      <td>chr1_904753_904883_strong_peak</td>\n",
       "      <td>6.115919</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36777</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155264125</td>\n",
       "      <td>155264549</td>\n",
       "      <td>chrX_155264125_155264549_medium_peak</td>\n",
       "      <td>3.204826</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36778</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155686260</td>\n",
       "      <td>155686684</td>\n",
       "      <td>chrX_155686260_155686684_weak_peak</td>\n",
       "      <td>1.295568</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36779</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155820126</td>\n",
       "      <td>155820550</td>\n",
       "      <td>chrX_155820126_155820550_medium_peak</td>\n",
       "      <td>3.426436</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36780</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155881180</td>\n",
       "      <td>155881604</td>\n",
       "      <td>chrX_155881180_155881604_medium_peak</td>\n",
       "      <td>3.068450</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36781</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155966835</td>\n",
       "      <td>155967259</td>\n",
       "      <td>chrX_155966835_155967259_medium_peak</td>\n",
       "      <td>2.471807</td>\n",
       "      <td>.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36782 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom      start        end                                  name   \n",
       "0      chr1     267790     268214          chr1_267790_268214_weak_peak  \\\n",
       "1      chr1     778701     779047        chr1_778701_779047_strong_peak   \n",
       "2      chr1     778863     779287        chr1_778863_779287_medium_peak   \n",
       "3      chr1     869675     870099          chr1_869675_870099_weak_peak   \n",
       "4      chr1     904753     904883        chr1_904753_904883_strong_peak   \n",
       "...     ...        ...        ...                                   ...   \n",
       "36777  chrX  155264125  155264549  chrX_155264125_155264549_medium_peak   \n",
       "36778  chrX  155686260  155686684    chrX_155686260_155686684_weak_peak   \n",
       "36779  chrX  155820126  155820550  chrX_155820126_155820550_medium_peak   \n",
       "36780  chrX  155881180  155881604  chrX_155881180_155881604_medium_peak   \n",
       "36781  chrX  155966835  155967259  chrX_155966835_155967259_medium_peak   \n",
       "\n",
       "           RPKM strand  \n",
       "0      1.568319      .  \n",
       "1      5.326924      .  \n",
       "2      2.778652      .  \n",
       "3      1.125098      .  \n",
       "4      6.115919      .  \n",
       "...         ...    ...  \n",
       "36777  3.204826      .  \n",
       "36778  1.295568      .  \n",
       "36779  3.426436      .  \n",
       "36780  3.068450      .  \n",
       "36781  2.471807      .  \n",
       "\n",
       "[36782 rows x 6 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a = BedTool.from_dataframe(df_offs.iloc[:,:3])\n",
    "b = BedTool.from_dataframe(df_ctcf.iloc[:,:3])\n",
    "# 去掉在off里的ctcf\n",
    "df_ctcf_out_offs = b.intersect(a, loj=True).to_dataframe().query('score==-1').iloc[:, :3].merge(df_ctcf, on=['chrom', 'start', 'end'], how='left')\n",
    "df_ctcf_in_offs = b.intersect(a, loj=True).to_dataframe().query('score!=-1').iloc[:, :3].merge(df_ctcf, on=['chrom', 'start', 'end'], how='left')\n",
    "df_ctcf_out_offs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0b9597cd-11e2-4a26-a25d-f16b41369e37",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:26:12.293282Z",
     "iopub.status.busy": "2023-05-22T07:26:12.292891Z",
     "iopub.status.idle": "2023-05-22T07:26:12.367466Z",
     "shell.execute_reply": "2023-05-22T07:26:12.366606Z",
     "shell.execute_reply.started": "2023-05-22T07:26:12.293251Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>RPKM</th>\n",
       "      <th>strand</th>\n",
       "      <th>signal_strength</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>267790</td>\n",
       "      <td>268214</td>\n",
       "      <td>chr1_267790_268214_weak_peak</td>\n",
       "      <td>1.568319</td>\n",
       "      <td>.</td>\n",
       "      <td>weak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>778701</td>\n",
       "      <td>779047</td>\n",
       "      <td>chr1_778701_779047_strong_peak</td>\n",
       "      <td>5.326924</td>\n",
       "      <td>.</td>\n",
       "      <td>strong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>778863</td>\n",
       "      <td>779287</td>\n",
       "      <td>chr1_778863_779287_medium_peak</td>\n",
       "      <td>2.778652</td>\n",
       "      <td>.</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>869675</td>\n",
       "      <td>870099</td>\n",
       "      <td>chr1_869675_870099_weak_peak</td>\n",
       "      <td>1.125098</td>\n",
       "      <td>.</td>\n",
       "      <td>weak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>904753</td>\n",
       "      <td>904883</td>\n",
       "      <td>chr1_904753_904883_strong_peak</td>\n",
       "      <td>6.115919</td>\n",
       "      <td>.</td>\n",
       "      <td>strong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36777</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155264125</td>\n",
       "      <td>155264549</td>\n",
       "      <td>chrX_155264125_155264549_medium_peak</td>\n",
       "      <td>3.204826</td>\n",
       "      <td>.</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36778</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155686260</td>\n",
       "      <td>155686684</td>\n",
       "      <td>chrX_155686260_155686684_weak_peak</td>\n",
       "      <td>1.295568</td>\n",
       "      <td>.</td>\n",
       "      <td>weak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36779</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155820126</td>\n",
       "      <td>155820550</td>\n",
       "      <td>chrX_155820126_155820550_medium_peak</td>\n",
       "      <td>3.426436</td>\n",
       "      <td>.</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36780</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155881180</td>\n",
       "      <td>155881604</td>\n",
       "      <td>chrX_155881180_155881604_medium_peak</td>\n",
       "      <td>3.068450</td>\n",
       "      <td>.</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36781</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155966835</td>\n",
       "      <td>155967259</td>\n",
       "      <td>chrX_155966835_155967259_medium_peak</td>\n",
       "      <td>2.471807</td>\n",
       "      <td>.</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>36782 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom      start        end                                  name   \n",
       "0      chr1     267790     268214          chr1_267790_268214_weak_peak  \\\n",
       "1      chr1     778701     779047        chr1_778701_779047_strong_peak   \n",
       "2      chr1     778863     779287        chr1_778863_779287_medium_peak   \n",
       "3      chr1     869675     870099          chr1_869675_870099_weak_peak   \n",
       "4      chr1     904753     904883        chr1_904753_904883_strong_peak   \n",
       "...     ...        ...        ...                                   ...   \n",
       "36777  chrX  155264125  155264549  chrX_155264125_155264549_medium_peak   \n",
       "36778  chrX  155686260  155686684    chrX_155686260_155686684_weak_peak   \n",
       "36779  chrX  155820126  155820550  chrX_155820126_155820550_medium_peak   \n",
       "36780  chrX  155881180  155881604  chrX_155881180_155881604_medium_peak   \n",
       "36781  chrX  155966835  155967259  chrX_155966835_155967259_medium_peak   \n",
       "\n",
       "           RPKM strand signal_strength  \n",
       "0      1.568319      .            weak  \n",
       "1      5.326924      .          strong  \n",
       "2      2.778652      .          medium  \n",
       "3      1.125098      .            weak  \n",
       "4      6.115919      .          strong  \n",
       "...         ...    ...             ...  \n",
       "36777  3.204826      .          medium  \n",
       "36778  1.295568      .            weak  \n",
       "36779  3.426436      .          medium  \n",
       "36780  3.068450      .          medium  \n",
       "36781  2.471807      .          medium  \n",
       "\n",
       "[36782 rows x 7 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ctcf_out_offs['signal_strength'] = df_ctcf_out_offs['name'].str.split('_').str[-2]\n",
    "df_ctcf_out_offs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "074210a2-a44e-4e6c-8b08-8800970f0892",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:26:50.199053Z",
     "iopub.status.busy": "2023-05-22T07:26:50.198383Z",
     "iopub.status.idle": "2023-05-22T07:26:50.234859Z",
     "shell.execute_reply": "2023-05-22T07:26:50.233533Z",
     "shell.execute_reply.started": "2023-05-22T07:26:50.198994Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>RPKM</th>\n",
       "      <th>strand</th>\n",
       "      <th>signal_strength</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2044897</td>\n",
       "      <td>2045321</td>\n",
       "      <td>chr1_2044897_2045321_medium_peak</td>\n",
       "      <td>3.477577</td>\n",
       "      <td>.</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2044897</td>\n",
       "      <td>2045321</td>\n",
       "      <td>chr1_2044897_2045321_medium_peak</td>\n",
       "      <td>3.477577</td>\n",
       "      <td>.</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2139943</td>\n",
       "      <td>2140169</td>\n",
       "      <td>chr1_2139943_2140169_strong_peak</td>\n",
       "      <td>8.826999</td>\n",
       "      <td>.</td>\n",
       "      <td>strong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9428572</td>\n",
       "      <td>9428996</td>\n",
       "      <td>chr1_9428572_9428996_strong_peak</td>\n",
       "      <td>4.704957</td>\n",
       "      <td>.</td>\n",
       "      <td>strong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9702095</td>\n",
       "      <td>9702519</td>\n",
       "      <td>chr1_9702095_9702519_weak_peak</td>\n",
       "      <td>1.892211</td>\n",
       "      <td>.</td>\n",
       "      <td>weak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1575</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153922850</td>\n",
       "      <td>153923274</td>\n",
       "      <td>chrX_153922850_153923274_strong_peak</td>\n",
       "      <td>3.682140</td>\n",
       "      <td>.</td>\n",
       "      <td>strong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1576</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154013581</td>\n",
       "      <td>154014005</td>\n",
       "      <td>chrX_154013581_154014005_strong_peak</td>\n",
       "      <td>4.261736</td>\n",
       "      <td>.</td>\n",
       "      <td>strong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1577</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154980943</td>\n",
       "      <td>154981142</td>\n",
       "      <td>chrX_154980943_154981142_strong_peak</td>\n",
       "      <td>7.627437</td>\n",
       "      <td>.</td>\n",
       "      <td>strong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1578</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155048969</td>\n",
       "      <td>155049393</td>\n",
       "      <td>chrX_155048969_155049393_medium_peak</td>\n",
       "      <td>2.574089</td>\n",
       "      <td>.</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1579</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155048969</td>\n",
       "      <td>155049393</td>\n",
       "      <td>chrX_155048969_155049393_medium_peak</td>\n",
       "      <td>2.574089</td>\n",
       "      <td>.</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1580 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end                                  name   \n",
       "0     chr1    2044897    2045321      chr1_2044897_2045321_medium_peak  \\\n",
       "1     chr1    2044897    2045321      chr1_2044897_2045321_medium_peak   \n",
       "2     chr1    2139943    2140169      chr1_2139943_2140169_strong_peak   \n",
       "3     chr1    9428572    9428996      chr1_9428572_9428996_strong_peak   \n",
       "4     chr1    9702095    9702519        chr1_9702095_9702519_weak_peak   \n",
       "...    ...        ...        ...                                   ...   \n",
       "1575  chrX  153922850  153923274  chrX_153922850_153923274_strong_peak   \n",
       "1576  chrX  154013581  154014005  chrX_154013581_154014005_strong_peak   \n",
       "1577  chrX  154980943  154981142  chrX_154980943_154981142_strong_peak   \n",
       "1578  chrX  155048969  155049393  chrX_155048969_155049393_medium_peak   \n",
       "1579  chrX  155048969  155049393  chrX_155048969_155049393_medium_peak   \n",
       "\n",
       "          RPKM strand signal_strength  \n",
       "0     3.477577      .          medium  \n",
       "1     3.477577      .          medium  \n",
       "2     8.826999      .          strong  \n",
       "3     4.704957      .          strong  \n",
       "4     1.892211      .            weak  \n",
       "...        ...    ...             ...  \n",
       "1575  3.682140      .          strong  \n",
       "1576  4.261736      .          strong  \n",
       "1577  7.627437      .          strong  \n",
       "1578  2.574089      .          medium  \n",
       "1579  2.574089      .          medium  \n",
       "\n",
       "[1580 rows x 7 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ctcf_in_offs['signal_strength'] = df_ctcf_in_offs['name'].str.split('_').str[-2]\n",
    "df_ctcf_in_offs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9696048b-6e2a-4b8c-a4b1-521e4564165b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:39:32.795139Z",
     "iopub.status.busy": "2023-05-22T07:39:32.794321Z",
     "iopub.status.idle": "2023-05-22T07:39:32.818469Z",
     "shell.execute_reply": "2023-05-22T07:39:32.816941Z",
     "shell.execute_reply.started": "2023-05-22T07:39:32.795062Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "signal_strength\n",
       "strong    12473\n",
       "medium    12166\n",
       "weak      12143\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ctcf_out_offs.signal_strength.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a9926bdd-0949-478d-a093-569262548338",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:39:33.542096Z",
     "iopub.status.busy": "2023-05-22T07:39:33.541462Z",
     "iopub.status.idle": "2023-05-22T07:39:33.555142Z",
     "shell.execute_reply": "2023-05-22T07:39:33.553727Z",
     "shell.execute_reply.started": "2023-05-22T07:39:33.542029Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "signal_strength\n",
       "strong    663\n",
       "medium    520\n",
       "weak      397\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_ctcf_in_offs.signal_strength.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b1d94c6d-2231-4c8f-8ddd-1862a16ede23",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:39:42.645853Z",
     "iopub.status.busy": "2023-05-22T07:39:42.645236Z",
     "iopub.status.idle": "2023-05-22T07:39:42.656268Z",
     "shell.execute_reply": "2023-05-22T07:39:42.654633Z",
     "shell.execute_reply.started": "2023-05-22T07:39:42.645799Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1570.6785927899516"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "36565 / (110 + 107 + 36565) * 1580\n",
    "\n",
    "# 所以从CTCF强度上看不出有无offtarget的差别"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b154948f-1b99-447d-91d3-519fce315c50",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:39:52.304948Z",
     "iopub.status.busy": "2023-05-22T07:39:52.304321Z",
     "iopub.status.idle": "2023-05-22T07:39:52.798662Z",
     "shell.execute_reply": "2023-05-22T07:39:52.797275Z",
     "shell.execute_reply.started": "2023-05-22T07:39:52.304893Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "for signal, df in df_ctcf_out_offs.groupby('signal_strength'):\n",
    "    # print(signal)\n",
    "    # print(df.drop(columns='signal_strength'))\n",
    "    df.drop(columns='signal_strength').to_csv(f'../bed_for_count/2023-05-12_CTCF_{signal}_for_count.bed', header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0fa7b436-00b0-48bd-a8f0-507895942f9c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:39:54.397556Z",
     "iopub.status.busy": "2023-05-22T07:39:54.396909Z",
     "iopub.status.idle": "2023-05-22T07:39:54.448695Z",
     "shell.execute_reply": "2023-05-22T07:39:54.446968Z",
     "shell.execute_reply.started": "2023-05-22T07:39:54.397497Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_ctcf_in_offs.drop(columns='signal_strength').to_csv(f'../bed_for_count/2023-05-12_CTCF_within-offs_for_count.bed', header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fda2dce6-23d7-4b3b-8682-af259335e72a",
   "metadata": {},
   "source": [
    "### df_offs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "3b6c4751-2793-4d53-b44d-af36c8fcdc07",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:39:56.967296Z",
     "iopub.status.busy": "2023-05-22T07:39:56.966658Z",
     "iopub.status.idle": "2023-05-22T07:39:57.004470Z",
     "shell.execute_reply": "2023-05-22T07:39:57.002908Z",
     "shell.execute_reply.started": "2023-05-22T07:39:56.967240Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>2044988</td>\n",
       "      <td>2044998</td>\n",
       "      <td>ND5.1-TAS.IND-1_chr1_2044988_2044998</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9702414</td>\n",
       "      <td>9702449</td>\n",
       "      <td>ND5.1-TAS.IND-2_ND6-TAS.IND-1_chr1_9702414_970...</td>\n",
       "      <td>11</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>12618165</td>\n",
       "      <td>12618174</td>\n",
       "      <td>ND5.1-TAS.IND-3_chr1_12618165_12618174</td>\n",
       "      <td>5</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13019998</td>\n",
       "      <td>13020019</td>\n",
       "      <td>ND6-TAS.IND-2_chr1_13019998_13020019</td>\n",
       "      <td>9</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13354705</td>\n",
       "      <td>13354738</td>\n",
       "      <td>ND6-TAS.IND-3_chr1_13354705_13354738</td>\n",
       "      <td>11</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1030</th>\n",
       "      <td>chrX</td>\n",
       "      <td>150772346</td>\n",
       "      <td>150772356</td>\n",
       "      <td>ND6-Extend-38_chrX_150772346_150772356</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1031</th>\n",
       "      <td>chrX</td>\n",
       "      <td>151608905</td>\n",
       "      <td>151608918</td>\n",
       "      <td>ND5.1-TAS.DR-91_chrX_151608905_151608918</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153669436</td>\n",
       "      <td>153669453</td>\n",
       "      <td>ND5.1-TAS.IND-453_ND6-TAS.IND-541_chrX_1536694...</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1033</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154373542</td>\n",
       "      <td>154373556</td>\n",
       "      <td>ND5.1-Extend-75_chrX_154373542_154373556</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1034</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155049205</td>\n",
       "      <td>155049215</td>\n",
       "      <td>ND5.1-TAS.IND-454_ND6-TAS.IND-542_chrX_1550492...</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1035 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end   \n",
       "0     chr1    2044988    2044998  \\\n",
       "1     chr1    9702414    9702449   \n",
       "2     chr1   12618165   12618174   \n",
       "3     chr1   13019998   13020019   \n",
       "4     chr1   13354705   13354738   \n",
       "...    ...        ...        ...   \n",
       "1030  chrX  150772346  150772356   \n",
       "1031  chrX  151608905  151608918   \n",
       "1032  chrX  153669436  153669453   \n",
       "1033  chrX  154373542  154373556   \n",
       "1034  chrX  155049205  155049215   \n",
       "\n",
       "                                                   name  score strand  \n",
       "0                  ND5.1-TAS.IND-1_chr1_2044988_2044998      7      +  \n",
       "1     ND5.1-TAS.IND-2_ND6-TAS.IND-1_chr1_9702414_970...     11      -  \n",
       "2                ND5.1-TAS.IND-3_chr1_12618165_12618174      5      +  \n",
       "3                  ND6-TAS.IND-2_chr1_13019998_13020019      9      -  \n",
       "4                  ND6-TAS.IND-3_chr1_13354705_13354738     11      +  \n",
       "...                                                 ...    ...    ...  \n",
       "1030             ND6-Extend-38_chrX_150772346_150772356      6      -  \n",
       "1031           ND5.1-TAS.DR-91_chrX_151608905_151608918      7      +  \n",
       "1032  ND5.1-TAS.IND-453_ND6-TAS.IND-541_chrX_1536694...     10      +  \n",
       "1033           ND5.1-Extend-75_chrX_154373542_154373556      6      -  \n",
       "1034  ND5.1-TAS.IND-454_ND6-TAS.IND-542_chrX_1550492...      7      -  \n",
       "\n",
       "[1035 rows x 6 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_offs = pd.read_csv(\n",
    "    '/lustre1/chengqiyi_pkuhpc/zhaohn/3.project/2022_DdCBE-3D-Genome_topic/'\n",
    "    '2022-09-30_Detect-seq_batch-1_ATP8_JAK2_SIRT6/tables/'\n",
    "    '2022-10-05_WT_ND4-5.1-6_final-list.bed',\n",
    "    header=None, sep='\\t', names=['chrom', 'start', 'end', 'name', 'score', 'strand']\n",
    ")\n",
    "df_offs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "26112522-c801-4895-a740-a64a02a2c0fe",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:39:57.139541Z",
     "iopub.status.busy": "2023-05-22T07:39:57.138909Z",
     "iopub.status.idle": "2023-05-22T07:39:57.172289Z",
     "shell.execute_reply": "2023-05-22T07:39:57.171029Z",
     "shell.execute_reply.started": "2023-05-22T07:39:57.139474Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>9702414</td>\n",
       "      <td>9702449</td>\n",
       "      <td>ND5.1-TAS.IND-2_ND6-TAS.IND-1_chr1_9702414_970...</td>\n",
       "      <td>11</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13019998</td>\n",
       "      <td>13020019</td>\n",
       "      <td>ND6-TAS.IND-2_chr1_13019998_13020019</td>\n",
       "      <td>9</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>13354705</td>\n",
       "      <td>13354738</td>\n",
       "      <td>ND6-TAS.IND-3_chr1_13354705_13354738</td>\n",
       "      <td>11</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>15046702</td>\n",
       "      <td>15046715</td>\n",
       "      <td>ND6-TAS.IND-4_chr1_15046702_15046715</td>\n",
       "      <td>7</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16082505</td>\n",
       "      <td>16082527</td>\n",
       "      <td>ND5.1-TAS.IND-4_ND6-TAS.IND-5_chr1_16082505_16...</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1027</th>\n",
       "      <td>chrX</td>\n",
       "      <td>136925330</td>\n",
       "      <td>136925338</td>\n",
       "      <td>ND6-TAS.IND-538_chrX_136925330_136925338</td>\n",
       "      <td>5</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1028</th>\n",
       "      <td>chrX</td>\n",
       "      <td>138128211</td>\n",
       "      <td>138128237</td>\n",
       "      <td>ND6-TAS.IND-539_chrX_138128211_138128237</td>\n",
       "      <td>9</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1029</th>\n",
       "      <td>chrX</td>\n",
       "      <td>150683735</td>\n",
       "      <td>150683758</td>\n",
       "      <td>ND5.1-TAS.IND-452_ND6-TAS.IND-540_chrX_1506837...</td>\n",
       "      <td>11</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1032</th>\n",
       "      <td>chrX</td>\n",
       "      <td>153669436</td>\n",
       "      <td>153669453</td>\n",
       "      <td>ND5.1-TAS.IND-453_ND6-TAS.IND-541_chrX_1536694...</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1034</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155049205</td>\n",
       "      <td>155049215</td>\n",
       "      <td>ND5.1-TAS.IND-454_ND6-TAS.IND-542_chrX_1550492...</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>542 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom      start        end   \n",
       "1     chr1    9702414    9702449  \\\n",
       "3     chr1   13019998   13020019   \n",
       "4     chr1   13354705   13354738   \n",
       "5     chr1   15046702   15046715   \n",
       "6     chr1   16082505   16082527   \n",
       "...    ...        ...        ...   \n",
       "1027  chrX  136925330  136925338   \n",
       "1028  chrX  138128211  138128237   \n",
       "1029  chrX  150683735  150683758   \n",
       "1032  chrX  153669436  153669453   \n",
       "1034  chrX  155049205  155049215   \n",
       "\n",
       "                                                   name  score strand  \n",
       "1     ND5.1-TAS.IND-2_ND6-TAS.IND-1_chr1_9702414_970...     11      -  \n",
       "3                  ND6-TAS.IND-2_chr1_13019998_13020019      9      -  \n",
       "4                  ND6-TAS.IND-3_chr1_13354705_13354738     11      +  \n",
       "5                  ND6-TAS.IND-4_chr1_15046702_15046715      7      +  \n",
       "6     ND5.1-TAS.IND-4_ND6-TAS.IND-5_chr1_16082505_16...     10      -  \n",
       "...                                                 ...    ...    ...  \n",
       "1027           ND6-TAS.IND-538_chrX_136925330_136925338      5      +  \n",
       "1028           ND6-TAS.IND-539_chrX_138128211_138128237      9      -  \n",
       "1029  ND5.1-TAS.IND-452_ND6-TAS.IND-540_chrX_1506837...     11      -  \n",
       "1032  ND5.1-TAS.IND-453_ND6-TAS.IND-541_chrX_1536694...     10      +  \n",
       "1034  ND5.1-TAS.IND-454_ND6-TAS.IND-542_chrX_1550492...      7      -  \n",
       "\n",
       "[542 rows x 6 columns]"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_offs_IND = pd.concat(\n",
    "    [\n",
    "        # df_offs.query('name.str.contains(\"ND6-TAS.DL\")'),\n",
    "        # df_offs.query('name.str.contains(\"ND6-TAS.DR\")'),\n",
    "        df_offs.query('name.str.contains(\"ND6-TAS.IND\")')\n",
    "    ]\n",
    ")\n",
    "df_offs_IND"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "cf980d99-6cf1-4197-8b9c-12862ed5de3d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:39:57.311724Z",
     "iopub.status.busy": "2023-05-22T07:39:57.311135Z",
     "iopub.status.idle": "2023-05-22T07:39:57.349525Z",
     "shell.execute_reply": "2023-05-22T07:39:57.348059Z",
     "shell.execute_reply.started": "2023-05-22T07:39:57.311669Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chr1</td>\n",
       "      <td>31956473</td>\n",
       "      <td>31956495</td>\n",
       "      <td>ND6-TAS.DL-1_chr1_31956473_31956495</td>\n",
       "      <td>10</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>chr1</td>\n",
       "      <td>35079561</td>\n",
       "      <td>35079576</td>\n",
       "      <td>ND6-TAS.DL-2_chr1_35079561_35079576</td>\n",
       "      <td>8</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54578134</td>\n",
       "      <td>54578149</td>\n",
       "      <td>ND6-TAS.DL-3_chr1_54578134_54578149</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>chr1</td>\n",
       "      <td>54671922</td>\n",
       "      <td>54671938</td>\n",
       "      <td>ND6-TAS.DL-4_chr1_54671922_54671938</td>\n",
       "      <td>13</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>chr1</td>\n",
       "      <td>95036181</td>\n",
       "      <td>95036212</td>\n",
       "      <td>ND6-TAS.DL-5_chr1_95036181_95036212</td>\n",
       "      <td>8</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>974</th>\n",
       "      <td>chr22</td>\n",
       "      <td>26135728</td>\n",
       "      <td>26135754</td>\n",
       "      <td>ND6-TAS.DR-29_chr22_26135728_26135754</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>983</th>\n",
       "      <td>chr22</td>\n",
       "      <td>35676615</td>\n",
       "      <td>35676639</td>\n",
       "      <td>ND6-TAS.DR-30_chr22_35676615_35676639</td>\n",
       "      <td>8</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1003</th>\n",
       "      <td>chrX</td>\n",
       "      <td>25020353</td>\n",
       "      <td>25020385</td>\n",
       "      <td>ND6-TAS.DR-31_chrX_25020353_25020385</td>\n",
       "      <td>14</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1016</th>\n",
       "      <td>chrX</td>\n",
       "      <td>106613768</td>\n",
       "      <td>106613787</td>\n",
       "      <td>ND6-TAS.DR-32_chrX_106613768_106613787</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1022</th>\n",
       "      <td>chrX</td>\n",
       "      <td>123188225</td>\n",
       "      <td>123188237</td>\n",
       "      <td>ND6-TAS.DR-33_chrX_123188225_123188237</td>\n",
       "      <td>5</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>117 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom      start        end                                    name   \n",
       "17     chr1   31956473   31956495     ND6-TAS.DL-1_chr1_31956473_31956495  \\\n",
       "23     chr1   35079561   35079576     ND6-TAS.DL-2_chr1_35079561_35079576   \n",
       "36     chr1   54578134   54578149     ND6-TAS.DL-3_chr1_54578134_54578149   \n",
       "37     chr1   54671922   54671938     ND6-TAS.DL-4_chr1_54671922_54671938   \n",
       "65     chr1   95036181   95036212     ND6-TAS.DL-5_chr1_95036181_95036212   \n",
       "...     ...        ...        ...                                     ...   \n",
       "974   chr22   26135728   26135754   ND6-TAS.DR-29_chr22_26135728_26135754   \n",
       "983   chr22   35676615   35676639   ND6-TAS.DR-30_chr22_35676615_35676639   \n",
       "1003   chrX   25020353   25020385    ND6-TAS.DR-31_chrX_25020353_25020385   \n",
       "1016   chrX  106613768  106613787  ND6-TAS.DR-32_chrX_106613768_106613787   \n",
       "1022   chrX  123188225  123188237  ND6-TAS.DR-33_chrX_123188225_123188237   \n",
       "\n",
       "      score strand  \n",
       "17       10      +  \n",
       "23        8      -  \n",
       "36        3      -  \n",
       "37       13      -  \n",
       "65        8      -  \n",
       "...     ...    ...  \n",
       "974       7      -  \n",
       "983       8      -  \n",
       "1003     14      -  \n",
       "1016      6      -  \n",
       "1022      5      +  \n",
       "\n",
       "[117 rows x 6 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_offs_DEP = pd.concat(\n",
    "    [\n",
    "        df_offs.query('name.str.contains(\"ND6-TAS.DL\")'),\n",
    "        df_offs.query('name.str.contains(\"ND6-TAS.DR\")'),\n",
    "        # df_offs.query('name.str.contains(\"ND6-TAS.IND\")')\n",
    "    ]\n",
    ")\n",
    "df_offs_DEP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5550ac52-dda3-45af-9192-8d5fdfc65117",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:39:58.791929Z",
     "iopub.status.busy": "2023-05-22T07:39:58.791298Z",
     "iopub.status.idle": "2023-05-22T07:39:58.912031Z",
     "shell.execute_reply": "2023-05-22T07:39:58.909003Z",
     "shell.execute_reply.started": "2023-05-22T07:39:58.791874Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_offs_IND.to_csv(f'../bed_for_count/2023-05-12_offs-ND6-IND_for_count.bed', header=False, index=False, sep='\\t')\n",
    "df_offs_DEP.to_csv(f'../bed_for_count/2023-05-12_offs-ND6-DEP_for_count.bed', header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6eebaa8e-a647-4288-bba3-0faff53543dd",
   "metadata": {},
   "source": [
    "### df_random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "39fa30cd-1dcc-492e-abbc-24ac9870a6f7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:40:00.017068Z",
     "iopub.status.busy": "2023-05-22T07:40:00.016350Z",
     "iopub.status.idle": "2023-05-22T07:44:18.925386Z",
     "shell.execute_reply": "2023-05-22T07:44:18.923243Z",
     "shell.execute_reply.started": "2023-05-22T07:40:00.017012Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       chrom      start        end   name  score strand\n",
      "0       chr1  171273699  171273899      1    200      +\n",
      "1      chr10  118814584  118814784      2    200      -\n",
      "2      chr17   54532798   54532998      3    200      -\n",
      "3       chr7   70496778   70496978      4    200      -\n",
      "4       chr2    2558151    2558351      5    200      -\n",
      "...      ...        ...        ...    ...    ...    ...\n",
      "32995   chrX  100943277  100943477  32996    200      -\n",
      "32996   chr5  102140158  102140358  32997    200      -\n",
      "32997  chr16    2979521    2979721  32998    200      +\n",
      "32998   chr3  141405823  141406023  32999    200      +\n",
      "32999  chr20    7284852    7285052  33000    200      +\n",
      "\n",
      "[33000 rows x 6 columns]\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING. chromosome (chr19_KI270930v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr13_KI270843v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949746v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270728v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270850v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_GL383528v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270728v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr13_KI270841v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949753v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270886v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270921v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270925v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270882v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_KI270770v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270858v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_GL000257v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr9_KI270823v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270845v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270908v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270788v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270908v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270759v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270438v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_GL339449v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270846v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_GL383554v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270910v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270791v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270822v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949749v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_GL000214v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL000209v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270850v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL383566v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_GL000216v2) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_GL000221v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270728v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270749v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270858v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270916v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_GL000194v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_GL383534v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_GL383555v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_KI270904v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270906v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270747v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270938v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949747v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270929v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_GL000219v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949747v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270741v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270744v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL000209v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr18_GL383571v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270811v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270803v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270762v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_GL383552v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949746v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrX_KI270880v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_KI270835v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_KI270776v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr9_KI270823v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270723v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL383564v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_KI270801v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949752v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270743v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_KI270716v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr11_GL383547v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270728v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270847v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270728v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_KI270778v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270722v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270896v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270804v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270847v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_KI270773v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270847v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_KI270775v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_GL383553v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr21_GL383579v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270728v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270858v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270810v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949746v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_GL000257v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270759v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270908v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr11_GL383547v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270763v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949746v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270438v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270803v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949752v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_KI270734v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_GL000216v2) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270438v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL383573v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_KI270736v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_GL877876v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270789v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270900v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270792v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270929v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270821v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_GL000257v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL000209v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270901v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr18_GL383569v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949748v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270850v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270811v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270727v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270728v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270883v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270821v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_GL339449v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270844v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949746v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr9_KI270719v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_KI270780v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr21_KI270873v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270442v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949749v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_GL000257v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_GL383555v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_KI270835v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_KI270731v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_KI270767v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_KI270904v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949750v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270741v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_KB021644v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270938v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270908v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270749v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270897v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr11_KI270829v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270899v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270787v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_GL000225v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_GL383557v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270794v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949752v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL383563v3_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_GL383555v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_GL383520v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr18_GL383570v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_GL383520v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270759v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL000209v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270792v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949751v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr11_KI270826v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270761v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_KI270779v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949749v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr9_GL383539v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_KI270801v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_KI270784v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270897v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_KI270736v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270787v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270847v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270759v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr20_KI270869v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_GL000194v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270847v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270519v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270728v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270898v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_KI270801v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270786v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270931v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr9_KI270719v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949753v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270746v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_JH636055v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949748v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr18_KI270863v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_KI270894v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270728v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270920v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270897v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270890v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270810v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_GL339449v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270728v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270901v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270814v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_GL000213v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr13_KI270838v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270822v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr20_KI270869v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_KI270772v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270914v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270858v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_KI270924v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270906v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270909v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270897v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270858v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr11_KI270830v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270729v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_KI270778v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270851v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_GL582966v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_GL000221v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_KI270798v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270808v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270933v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270811v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270897v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270916v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270821v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL383533v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_GL339449v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270763v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_KI270799v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270821v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270756v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_KI270875v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270868v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270915v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270897v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_GL000208v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_GL000257v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_KI270924v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270796v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270897v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270747v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_KI270879v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr9_KI270823v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270589v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr11_KI270826v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949753v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_KI270875v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr21_GL383579v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270847v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_KI270784v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270727v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949749v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270744v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949746v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949749v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr20_KI270870v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL383563v3_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270846v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270812v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr18_GL383567v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270908v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270908v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270847v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_GL877876v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949747v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270910v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270855v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270804v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_GL383549v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270712v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270812v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270722v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270803v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_KI270904v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_KI270801v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL383565v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270861v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270862v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_KI270800v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr21_GL383579v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949746v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270809v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270908v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_GL383527v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949746v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270846v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270442v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270803v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270442v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270803v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_KI270904v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270803v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_GL383520v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270730v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_KB021644v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270821v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270909v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270919v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270821v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_KI270798v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270728v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270807v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270763v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_GL383554v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_GL877876v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_KB021644v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_GL877876v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_GL383549v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270925v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL000209v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270793v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270803v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_KI270776v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL383573v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_GL339449v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_KI270904v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270728v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr18_GL383570v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949751v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949752v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_KI270894v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949750v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr13_KI270841v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949751v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270847v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr11_GL383547v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrX_KI270881v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270909v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL383575v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949747v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270890v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270862v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270896v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270897v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270728v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270810v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270787v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270822v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270888v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_KI270879v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949746v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949753v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270855v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270787v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270710v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270821v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270929v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270765v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270897v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_GL383528v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270906v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270803v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270748v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270847v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270921v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_KI270835v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_KI270734v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270897v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270818v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949749v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949749v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_KI270775v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949746v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270926v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270442v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_GL000257v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270925v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949749v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270852v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL383576v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949747v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270909v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_GL383528v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270908v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270891v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270809v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_JH636055v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270908v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270763v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270855v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_GL339449v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_GL383532v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr18_KI270911v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270805v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270796v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270922v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270815v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr20_KI270869v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_GL877875v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_KI270801v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949746v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr18_GL383570v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949749v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270901v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_GL383555v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr18_GL383567v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949751v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949749v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL383564v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000205v2_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949746v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949751v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_GL383555v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270887v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270807v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr9_KI270719v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270854v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270906v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270923v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_GL339449v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270938v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr11_KI270903v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270811v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270848v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949747v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270822v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270729v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270813v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_KB021644v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270845v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL383563v3_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270925v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_JH159146v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270884v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_GL383526v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_KI270738v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_GL383527v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_GL339449v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270849v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270847v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270763v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270806v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr13_KI270839v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr11_KI270830v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949748v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_GL949742v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr13_KI270843v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270796v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949749v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270854v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr9_KI270823v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270908v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL383573v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_JH159147v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_KI270775v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270728v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270930v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270708v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_KI270716v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr20_KI270870v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270729v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270908v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270938v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270846v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270849v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949747v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000205v2_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949750v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr11_KI270831v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_GL383549v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270706v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL383563v3_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270816v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949752v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270766v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr13_KI270838v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270891v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270810v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949751v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_GL339449v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_GL000008v2_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270522v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_GL339449v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270851v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_JH159146v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270896v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270747v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270925v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949748v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949751v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270789v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949752v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL383564v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270851v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270706v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949748v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000205v2_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270806v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_GL000009v2_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949748v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270821v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270901v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrX_KI270880v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr18_GL383569v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270852v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270860v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_GL000194v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_GL383582v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270900v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_GL383528v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270852v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270932v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949748v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270763v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr21_KI270873v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL383533v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949749v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270925v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270813v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_GL383556v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_GL339449v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_GL383520v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_KI270780v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270846v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949748v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949751v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270908v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr20_KI270869v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270917v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949750v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270852v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_GL000218v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_KI270779v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270809v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270811v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270846v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270909v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_KI270774v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr20_KI270870v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_KI270875v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270803v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270727v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_GL383583v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949750v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_KI270935v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270908v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270931v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL383565v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270916v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270908v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270538v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270752v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_KI270769v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270909v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270812v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949751v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270747v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_KI270928v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_KI270801v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_GL877875v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrX_KI270913v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270917v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr11_KI270826v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270851v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270790v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270790v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_KI270879v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270810v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270812v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270897v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270727v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_KI270937v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr11_KI270927v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL383563v3_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270846v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270938v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270804v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_KI270835v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270786v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_GL383556v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949749v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_GL339449v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr11_JH159136v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_KI270758v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949750v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949753v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_KI270878v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270729v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270897v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949750v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr9_KI270823v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949747v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_KI270736v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_GL383520v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270938v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_GL383556v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270728v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270742v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr21_GL383578v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_KI270879v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr21_GL383581v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270884v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949746v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_KI270783v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270908v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270854v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270811v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_KI270904v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr11_KI270829v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270848v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949746v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_JH159148v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr18_GL383569v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270722v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270909v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr20_KI270870v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_KI270783v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_GL339449v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270847v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270438v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270729v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270760v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_GL383555v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270788v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr11_KI270832v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270810v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr9_GL383540v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270848v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270810v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr18_GL383572v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000205v2_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270763v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270750v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270938v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270925v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270938v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270764v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_JH159146v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949750v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_GL000009v2_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270849v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270821v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL383565v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_GL383554v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270908v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_KI270775v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_GL383528v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270922v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_KI270893v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270763v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_KI270774v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270765v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270821v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr2_KI270775v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270728v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr10_GL383546v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr9_GL383539v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_KI270799v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270847v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270850v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270790v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270938v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270861v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270929v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270792v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270853v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270438v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr18_GL383571v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270847v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270910v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr20_KI270870v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270794v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270850v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr11_KI270830v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270763v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270747v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949753v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270742v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270938v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270817v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr18_GL383569v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr9_GL383539v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr21_KI270872v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_GL000258v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270925v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr9_GL383542v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr13_KI270839v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270884v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_GL383551v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_KI270904v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270859v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270846v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270728v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270926v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_GL383526v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270929v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949753v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270803v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270900v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_KI270743v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_GL000219v1) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270761v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949749v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270897v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000253v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270803v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr8_KI270821v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr7_KI270808v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270847v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270908v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270763v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chrUn_GL000216v2) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270712v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr5_KI270796v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_GL000225v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_KI270896v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_KI270904v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr4_GL000257v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949749v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270906v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270857v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270762v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000255v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_JH159147v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949748v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270713v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_KI270779v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949747v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_KI270733v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr16_KI270728v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr21_GL383579v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270727v1_random) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000256v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270914v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL949751v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr1_KI270763v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_GL383583v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000251v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr17_KI270909v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_KI270938v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL383574v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr22_KI270876v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr21_GL383579v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr11_KI270903v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000252v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr19_GL383574v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr11_GL383547v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr3_KI270782v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000250v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr12_GL877875v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr15_KI270905v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr6_GL000254v2_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr21_KI270873v1_alt) was not found in the FASTA file. Skipping.\n",
      "WARNING. chromosome (chr14_KI270845v1_alt) was not found in the FASTA file. Skipping.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>info</th>\n",
       "      <th>coordinate</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:171273699-171273899\\nagagataaggtctcactatg...</td>\n",
       "      <td>chr1:171273699-171273899</td>\n",
       "      <td>agagataaggtctcactatgccgcccagggtggtcttgaactcctg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10:118814584-118814784\\ngtgagcccccttgcccaag...</td>\n",
       "      <td>chr10:118814584-118814784</td>\n",
       "      <td>gtgagcccccttgcccaaggtctcacctccttcttagagagcttgt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17:54532798-54532998\\ncaaatgatatttctgcctcta...</td>\n",
       "      <td>chr17:54532798-54532998</td>\n",
       "      <td>caaatgatatttctgcctctatgttttcttttgtttaattttgagt...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7:70496778-70496978\\ntcagacacacacacacacccca...</td>\n",
       "      <td>chr7:70496778-70496978</td>\n",
       "      <td>tcagacacacacacacaccccacacatgcacacgtcacatcagcat...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2:2558151-2558351\\nGAAGGTGAACAGACTGCCCTTCAA...</td>\n",
       "      <td>chr2:2558151-2558351</td>\n",
       "      <td>GAAGGTGAACAGACTGCCCTTCAAATCAAATTTGGAGTAATGGAGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31841</th>\n",
       "      <td>chrX:100943277-100943477\\ncataataggatattgtttta...</td>\n",
       "      <td>chrX:100943277-100943477</td>\n",
       "      <td>cataataggatattgttttatgtttttatactttacgttaatacta...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31842</th>\n",
       "      <td>chr5:102140158-102140358\\nCCATATGCACTCTTATACAG...</td>\n",
       "      <td>chr5:102140158-102140358</td>\n",
       "      <td>CCATATGCACTCTTATACAGAGAAAAGCCACCACATATCAAGCTAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31843</th>\n",
       "      <td>chr16:2979521-2979721\\nAGGGTGTTGGGATCCTGGAGGTC...</td>\n",
       "      <td>chr16:2979521-2979721</td>\n",
       "      <td>AGGGTGTTGGGATCCTGGAGGTCACTGGTCACAAGCACAGCCTCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31844</th>\n",
       "      <td>chr3:141405823-141406023\\nCTCTGTTGAGGAAAGAAGAT...</td>\n",
       "      <td>chr3:141405823-141406023</td>\n",
       "      <td>CTCTGTTGAGGAAAGAAGATATGCAGAAATAACCACGTGATAAATG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31845</th>\n",
       "      <td>chr20:7284852-7285052\\nCTAAGGAAAATGGTACCCATTAA...</td>\n",
       "      <td>chr20:7284852-7285052</td>\n",
       "      <td>CTAAGGAAAATGGTACCCATTAAGAATGATTTTTTTAAAAGCAAAT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30316 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    info   \n",
       "0      chr1:171273699-171273899\\nagagataaggtctcactatg...  \\\n",
       "1      chr10:118814584-118814784\\ngtgagcccccttgcccaag...   \n",
       "2      chr17:54532798-54532998\\ncaaatgatatttctgcctcta...   \n",
       "3      chr7:70496778-70496978\\ntcagacacacacacacacccca...   \n",
       "4      chr2:2558151-2558351\\nGAAGGTGAACAGACTGCCCTTCAA...   \n",
       "...                                                  ...   \n",
       "31841  chrX:100943277-100943477\\ncataataggatattgtttta...   \n",
       "31842  chr5:102140158-102140358\\nCCATATGCACTCTTATACAG...   \n",
       "31843  chr16:2979521-2979721\\nAGGGTGTTGGGATCCTGGAGGTC...   \n",
       "31844  chr3:141405823-141406023\\nCTCTGTTGAGGAAAGAAGAT...   \n",
       "31845  chr20:7284852-7285052\\nCTAAGGAAAATGGTACCCATTAA...   \n",
       "\n",
       "                      coordinate   \n",
       "0       chr1:171273699-171273899  \\\n",
       "1      chr10:118814584-118814784   \n",
       "2        chr17:54532798-54532998   \n",
       "3         chr7:70496778-70496978   \n",
       "4           chr2:2558151-2558351   \n",
       "...                          ...   \n",
       "31841   chrX:100943277-100943477   \n",
       "31842   chr5:102140158-102140358   \n",
       "31843      chr16:2979521-2979721   \n",
       "31844   chr3:141405823-141406023   \n",
       "31845      chr20:7284852-7285052   \n",
       "\n",
       "                                                sequence  \n",
       "0      agagataaggtctcactatgccgcccagggtggtcttgaactcctg...  \n",
       "1      gtgagcccccttgcccaaggtctcacctccttcttagagagcttgt...  \n",
       "2      caaatgatatttctgcctctatgttttcttttgtttaattttgagt...  \n",
       "3      tcagacacacacacacaccccacacatgcacacgtcacatcagcat...  \n",
       "4      GAAGGTGAACAGACTGCCCTTCAAATCAAATTTGGAGTAATGGAGA...  \n",
       "...                                                  ...  \n",
       "31841  cataataggatattgttttatgtttttatactttacgttaatacta...  \n",
       "31842  CCATATGCACTCTTATACAGAGAAAAGCCACCACATATCAAGCTAT...  \n",
       "31843  AGGGTGTTGGGATCCTGGAGGTCACTGGTCACAAGCACAGCCTCCA...  \n",
       "31844  CTCTGTTGAGGAAAGAAGATATGCAGAAATAACCACGTGATAAATG...  \n",
       "31845  CTAAGGAAAATGGTACCCATTAAGAATGATTTTTTTAAAAGCAAAT...  \n",
       "\n",
       "[30316 rows x 3 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "random = BedTool().random(l=200, n=33000, genome='hg38', seed=123)\n",
    "print(random.to_dataframe())\n",
    "random_fa = random.sequence(fi='/lustre1/chengqiyi_pkuhpc/zhaohn/1.database/db_genomes/genome_fa/genome_ucsc_hg38/genome_ucsc_hg38.fa')\n",
    "df_random = pd.DataFrame(random_fa.print_sequence().split('>')[1:], columns=['info'])\n",
    "df_random[['coordinate', 'sequence']] = df_random['info'].str.strip().str.split('\\n', expand=True)\n",
    "df_random = df_random.query('~sequence.str.upper().str.contains(\"N\")').copy()\n",
    "df_random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b2b0dd50-f72a-456e-89b7-e03a8949386c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:47:39.229012Z",
     "iopub.status.busy": "2023-05-22T07:47:39.228269Z",
     "iopub.status.idle": "2023-05-22T07:47:39.361071Z",
     "shell.execute_reply": "2023-05-22T07:47:39.359626Z",
     "shell.execute_reply.started": "2023-05-22T07:47:39.228952Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>info</th>\n",
       "      <th>coordinate</th>\n",
       "      <th>sequence</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:171273699-171273899\\nagagataaggtctcactatg...</td>\n",
       "      <td>chr1:171273699-171273899</td>\n",
       "      <td>agagataaggtctcactatgccgcccagggtggtcttgaactcctg...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>171273699</td>\n",
       "      <td>171273899</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10:118814584-118814784\\ngtgagcccccttgcccaag...</td>\n",
       "      <td>chr10:118814584-118814784</td>\n",
       "      <td>gtgagcccccttgcccaaggtctcacctccttcttagagagcttgt...</td>\n",
       "      <td>chr10</td>\n",
       "      <td>118814584</td>\n",
       "      <td>118814784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17:54532798-54532998\\ncaaatgatatttctgcctcta...</td>\n",
       "      <td>chr17:54532798-54532998</td>\n",
       "      <td>caaatgatatttctgcctctatgttttcttttgtttaattttgagt...</td>\n",
       "      <td>chr17</td>\n",
       "      <td>54532798</td>\n",
       "      <td>54532998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7:70496778-70496978\\ntcagacacacacacacacccca...</td>\n",
       "      <td>chr7:70496778-70496978</td>\n",
       "      <td>tcagacacacacacacaccccacacatgcacacgtcacatcagcat...</td>\n",
       "      <td>chr7</td>\n",
       "      <td>70496778</td>\n",
       "      <td>70496978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2:2558151-2558351\\nGAAGGTGAACAGACTGCCCTTCAA...</td>\n",
       "      <td>chr2:2558151-2558351</td>\n",
       "      <td>GAAGGTGAACAGACTGCCCTTCAAATCAAATTTGGAGTAATGGAGA...</td>\n",
       "      <td>chr2</td>\n",
       "      <td>2558151</td>\n",
       "      <td>2558351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31841</th>\n",
       "      <td>chrX:100943277-100943477\\ncataataggatattgtttta...</td>\n",
       "      <td>chrX:100943277-100943477</td>\n",
       "      <td>cataataggatattgttttatgtttttatactttacgttaatacta...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>100943277</td>\n",
       "      <td>100943477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31842</th>\n",
       "      <td>chr5:102140158-102140358\\nCCATATGCACTCTTATACAG...</td>\n",
       "      <td>chr5:102140158-102140358</td>\n",
       "      <td>CCATATGCACTCTTATACAGAGAAAAGCCACCACATATCAAGCTAT...</td>\n",
       "      <td>chr5</td>\n",
       "      <td>102140158</td>\n",
       "      <td>102140358</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31843</th>\n",
       "      <td>chr16:2979521-2979721\\nAGGGTGTTGGGATCCTGGAGGTC...</td>\n",
       "      <td>chr16:2979521-2979721</td>\n",
       "      <td>AGGGTGTTGGGATCCTGGAGGTCACTGGTCACAAGCACAGCCTCCA...</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2979521</td>\n",
       "      <td>2979721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31844</th>\n",
       "      <td>chr3:141405823-141406023\\nCTCTGTTGAGGAAAGAAGAT...</td>\n",
       "      <td>chr3:141405823-141406023</td>\n",
       "      <td>CTCTGTTGAGGAAAGAAGATATGCAGAAATAACCACGTGATAAATG...</td>\n",
       "      <td>chr3</td>\n",
       "      <td>141405823</td>\n",
       "      <td>141406023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31845</th>\n",
       "      <td>chr20:7284852-7285052\\nCTAAGGAAAATGGTACCCATTAA...</td>\n",
       "      <td>chr20:7284852-7285052</td>\n",
       "      <td>CTAAGGAAAATGGTACCCATTAAGAATGATTTTTTTAAAAGCAAAT...</td>\n",
       "      <td>chr20</td>\n",
       "      <td>7284852</td>\n",
       "      <td>7285052</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30316 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    info   \n",
       "0      chr1:171273699-171273899\\nagagataaggtctcactatg...  \\\n",
       "1      chr10:118814584-118814784\\ngtgagcccccttgcccaag...   \n",
       "2      chr17:54532798-54532998\\ncaaatgatatttctgcctcta...   \n",
       "3      chr7:70496778-70496978\\ntcagacacacacacacacccca...   \n",
       "4      chr2:2558151-2558351\\nGAAGGTGAACAGACTGCCCTTCAA...   \n",
       "...                                                  ...   \n",
       "31841  chrX:100943277-100943477\\ncataataggatattgtttta...   \n",
       "31842  chr5:102140158-102140358\\nCCATATGCACTCTTATACAG...   \n",
       "31843  chr16:2979521-2979721\\nAGGGTGTTGGGATCCTGGAGGTC...   \n",
       "31844  chr3:141405823-141406023\\nCTCTGTTGAGGAAAGAAGAT...   \n",
       "31845  chr20:7284852-7285052\\nCTAAGGAAAATGGTACCCATTAA...   \n",
       "\n",
       "                      coordinate   \n",
       "0       chr1:171273699-171273899  \\\n",
       "1      chr10:118814584-118814784   \n",
       "2        chr17:54532798-54532998   \n",
       "3         chr7:70496778-70496978   \n",
       "4           chr2:2558151-2558351   \n",
       "...                          ...   \n",
       "31841   chrX:100943277-100943477   \n",
       "31842   chr5:102140158-102140358   \n",
       "31843      chr16:2979521-2979721   \n",
       "31844   chr3:141405823-141406023   \n",
       "31845      chr20:7284852-7285052   \n",
       "\n",
       "                                                sequence  chrom      start   \n",
       "0      agagataaggtctcactatgccgcccagggtggtcttgaactcctg...   chr1  171273699  \\\n",
       "1      gtgagcccccttgcccaaggtctcacctccttcttagagagcttgt...  chr10  118814584   \n",
       "2      caaatgatatttctgcctctatgttttcttttgtttaattttgagt...  chr17   54532798   \n",
       "3      tcagacacacacacacaccccacacatgcacacgtcacatcagcat...   chr7   70496778   \n",
       "4      GAAGGTGAACAGACTGCCCTTCAAATCAAATTTGGAGTAATGGAGA...   chr2    2558151   \n",
       "...                                                  ...    ...        ...   \n",
       "31841  cataataggatattgttttatgtttttatactttacgttaatacta...   chrX  100943277   \n",
       "31842  CCATATGCACTCTTATACAGAGAAAAGCCACCACATATCAAGCTAT...   chr5  102140158   \n",
       "31843  AGGGTGTTGGGATCCTGGAGGTCACTGGTCACAAGCACAGCCTCCA...  chr16    2979521   \n",
       "31844  CTCTGTTGAGGAAAGAAGATATGCAGAAATAACCACGTGATAAATG...   chr3  141405823   \n",
       "31845  CTAAGGAAAATGGTACCCATTAAGAATGATTTTTTTAAAAGCAAAT...  chr20    7284852   \n",
       "\n",
       "             end  \n",
       "0      171273899  \n",
       "1      118814784  \n",
       "2       54532998  \n",
       "3       70496978  \n",
       "4        2558351  \n",
       "...          ...  \n",
       "31841  100943477  \n",
       "31842  102140358  \n",
       "31843    2979721  \n",
       "31844  141406023  \n",
       "31845    7285052  \n",
       "\n",
       "[30316 rows x 6 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_random[['chrom', 'start', 'end']] = df_random['coordinate'].str.replace(':', '\\t').str.replace('-', '\\t').str.split('\\t', expand=True)\n",
    "df_random[['start', 'end']] = df_random[['start', 'end']].astype(int)\n",
    "df_random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d02cc06c-f0be-4272-ac5a-703f03c048f9",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:47:42.582323Z",
     "iopub.status.busy": "2023-05-22T07:47:42.581672Z",
     "iopub.status.idle": "2023-05-22T07:47:42.680916Z",
     "shell.execute_reply": "2023-05-22T07:47:42.679408Z",
     "shell.execute_reply.started": "2023-05-22T07:47:42.582265Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>info</th>\n",
       "      <th>coordinate</th>\n",
       "      <th>sequence</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:171273699-171273899\\nagagataaggtctcactatg...</td>\n",
       "      <td>chr1:171273699-171273899</td>\n",
       "      <td>agagataaggtctcactatgccgcccagggtggtcttgaactcctg...</td>\n",
       "      <td>chr1</td>\n",
       "      <td>171273699</td>\n",
       "      <td>171273899</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10:118814584-118814784\\ngtgagcccccttgcccaag...</td>\n",
       "      <td>chr10:118814584-118814784</td>\n",
       "      <td>gtgagcccccttgcccaaggtctcacctccttcttagagagcttgt...</td>\n",
       "      <td>chr10</td>\n",
       "      <td>118814584</td>\n",
       "      <td>118814784</td>\n",
       "      <td>2</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17:54532798-54532998\\ncaaatgatatttctgcctcta...</td>\n",
       "      <td>chr17:54532798-54532998</td>\n",
       "      <td>caaatgatatttctgcctctatgttttcttttgtttaattttgagt...</td>\n",
       "      <td>chr17</td>\n",
       "      <td>54532798</td>\n",
       "      <td>54532998</td>\n",
       "      <td>3</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7:70496778-70496978\\ntcagacacacacacacacccca...</td>\n",
       "      <td>chr7:70496778-70496978</td>\n",
       "      <td>tcagacacacacacacaccccacacatgcacacgtcacatcagcat...</td>\n",
       "      <td>chr7</td>\n",
       "      <td>70496778</td>\n",
       "      <td>70496978</td>\n",
       "      <td>4</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2:2558151-2558351\\nGAAGGTGAACAGACTGCCCTTCAA...</td>\n",
       "      <td>chr2:2558151-2558351</td>\n",
       "      <td>GAAGGTGAACAGACTGCCCTTCAAATCAAATTTGGAGTAATGGAGA...</td>\n",
       "      <td>chr2</td>\n",
       "      <td>2558151</td>\n",
       "      <td>2558351</td>\n",
       "      <td>5</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30311</th>\n",
       "      <td>chrX:100943277-100943477\\ncataataggatattgtttta...</td>\n",
       "      <td>chrX:100943277-100943477</td>\n",
       "      <td>cataataggatattgttttatgtttttatactttacgttaatacta...</td>\n",
       "      <td>chrX</td>\n",
       "      <td>100943277</td>\n",
       "      <td>100943477</td>\n",
       "      <td>32996</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30312</th>\n",
       "      <td>chr5:102140158-102140358\\nCCATATGCACTCTTATACAG...</td>\n",
       "      <td>chr5:102140158-102140358</td>\n",
       "      <td>CCATATGCACTCTTATACAGAGAAAAGCCACCACATATCAAGCTAT...</td>\n",
       "      <td>chr5</td>\n",
       "      <td>102140158</td>\n",
       "      <td>102140358</td>\n",
       "      <td>32997</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30313</th>\n",
       "      <td>chr16:2979521-2979721\\nAGGGTGTTGGGATCCTGGAGGTC...</td>\n",
       "      <td>chr16:2979521-2979721</td>\n",
       "      <td>AGGGTGTTGGGATCCTGGAGGTCACTGGTCACAAGCACAGCCTCCA...</td>\n",
       "      <td>chr16</td>\n",
       "      <td>2979521</td>\n",
       "      <td>2979721</td>\n",
       "      <td>32998</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30314</th>\n",
       "      <td>chr3:141405823-141406023\\nCTCTGTTGAGGAAAGAAGAT...</td>\n",
       "      <td>chr3:141405823-141406023</td>\n",
       "      <td>CTCTGTTGAGGAAAGAAGATATGCAGAAATAACCACGTGATAAATG...</td>\n",
       "      <td>chr3</td>\n",
       "      <td>141405823</td>\n",
       "      <td>141406023</td>\n",
       "      <td>32999</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30315</th>\n",
       "      <td>chr20:7284852-7285052\\nCTAAGGAAAATGGTACCCATTAA...</td>\n",
       "      <td>chr20:7284852-7285052</td>\n",
       "      <td>CTAAGGAAAATGGTACCCATTAAGAATGATTTTTTTAAAAGCAAAT...</td>\n",
       "      <td>chr20</td>\n",
       "      <td>7284852</td>\n",
       "      <td>7285052</td>\n",
       "      <td>33000</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30316 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    info   \n",
       "0      chr1:171273699-171273899\\nagagataaggtctcactatg...  \\\n",
       "1      chr10:118814584-118814784\\ngtgagcccccttgcccaag...   \n",
       "2      chr17:54532798-54532998\\ncaaatgatatttctgcctcta...   \n",
       "3      chr7:70496778-70496978\\ntcagacacacacacacacccca...   \n",
       "4      chr2:2558151-2558351\\nGAAGGTGAACAGACTGCCCTTCAA...   \n",
       "...                                                  ...   \n",
       "30311  chrX:100943277-100943477\\ncataataggatattgtttta...   \n",
       "30312  chr5:102140158-102140358\\nCCATATGCACTCTTATACAG...   \n",
       "30313  chr16:2979521-2979721\\nAGGGTGTTGGGATCCTGGAGGTC...   \n",
       "30314  chr3:141405823-141406023\\nCTCTGTTGAGGAAAGAAGAT...   \n",
       "30315  chr20:7284852-7285052\\nCTAAGGAAAATGGTACCCATTAA...   \n",
       "\n",
       "                      coordinate   \n",
       "0       chr1:171273699-171273899  \\\n",
       "1      chr10:118814584-118814784   \n",
       "2        chr17:54532798-54532998   \n",
       "3         chr7:70496778-70496978   \n",
       "4           chr2:2558151-2558351   \n",
       "...                          ...   \n",
       "30311   chrX:100943277-100943477   \n",
       "30312   chr5:102140158-102140358   \n",
       "30313      chr16:2979521-2979721   \n",
       "30314   chr3:141405823-141406023   \n",
       "30315      chr20:7284852-7285052   \n",
       "\n",
       "                                                sequence  chrom      start   \n",
       "0      agagataaggtctcactatgccgcccagggtggtcttgaactcctg...   chr1  171273699  \\\n",
       "1      gtgagcccccttgcccaaggtctcacctccttcttagagagcttgt...  chr10  118814584   \n",
       "2      caaatgatatttctgcctctatgttttcttttgtttaattttgagt...  chr17   54532798   \n",
       "3      tcagacacacacacacaccccacacatgcacacgtcacatcagcat...   chr7   70496778   \n",
       "4      GAAGGTGAACAGACTGCCCTTCAAATCAAATTTGGAGTAATGGAGA...   chr2    2558151   \n",
       "...                                                  ...    ...        ...   \n",
       "30311  cataataggatattgttttatgtttttatactttacgttaatacta...   chrX  100943277   \n",
       "30312  CCATATGCACTCTTATACAGAGAAAAGCCACCACATATCAAGCTAT...   chr5  102140158   \n",
       "30313  AGGGTGTTGGGATCCTGGAGGTCACTGGTCACAAGCACAGCCTCCA...  chr16    2979521   \n",
       "30314  CTCTGTTGAGGAAAGAAGATATGCAGAAATAACCACGTGATAAATG...   chr3  141405823   \n",
       "30315  CTAAGGAAAATGGTACCCATTAAGAATGATTTTTTTAAAAGCAAAT...  chr20    7284852   \n",
       "\n",
       "             end   name  score strand  \n",
       "0      171273899      1    200      +  \n",
       "1      118814784      2    200      -  \n",
       "2       54532998      3    200      -  \n",
       "3       70496978      4    200      -  \n",
       "4        2558351      5    200      -  \n",
       "...          ...    ...    ...    ...  \n",
       "30311  100943477  32996    200      -  \n",
       "30312  102140358  32997    200      -  \n",
       "30313    2979721  32998    200      +  \n",
       "30314  141406023  32999    200      +  \n",
       "30315    7285052  33000    200      +  \n",
       "\n",
       "[30316 rows x 9 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_random = df_random.merge(random.to_dataframe(), on=['chrom', 'start', 'end'], how='left')\n",
    "df_random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "bd273465-193f-4003-b988-7c8d65e2f99a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:47:52.073362Z",
     "iopub.status.busy": "2023-05-22T07:47:52.072711Z",
     "iopub.status.idle": "2023-05-22T07:47:52.109578Z",
     "shell.execute_reply": "2023-05-22T07:47:52.107673Z",
     "shell.execute_reply.started": "2023-05-22T07:47:52.073305Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>171273699</td>\n",
       "      <td>171273899</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>118814584</td>\n",
       "      <td>118814784</td>\n",
       "      <td>2</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17</td>\n",
       "      <td>54532798</td>\n",
       "      <td>54532998</td>\n",
       "      <td>3</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7</td>\n",
       "      <td>70496778</td>\n",
       "      <td>70496978</td>\n",
       "      <td>4</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>2558151</td>\n",
       "      <td>2558351</td>\n",
       "      <td>5</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30311</th>\n",
       "      <td>chrX</td>\n",
       "      <td>100943277</td>\n",
       "      <td>100943477</td>\n",
       "      <td>32996</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30312</th>\n",
       "      <td>chr5</td>\n",
       "      <td>102140158</td>\n",
       "      <td>102140358</td>\n",
       "      <td>32997</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30313</th>\n",
       "      <td>chr16</td>\n",
       "      <td>2979521</td>\n",
       "      <td>2979721</td>\n",
       "      <td>32998</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30314</th>\n",
       "      <td>chr3</td>\n",
       "      <td>141405823</td>\n",
       "      <td>141406023</td>\n",
       "      <td>32999</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30315</th>\n",
       "      <td>chr20</td>\n",
       "      <td>7284852</td>\n",
       "      <td>7285052</td>\n",
       "      <td>33000</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30316 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom      start        end   name  score strand\n",
       "0       chr1  171273699  171273899      1    200      +\n",
       "1      chr10  118814584  118814784      2    200      -\n",
       "2      chr17   54532798   54532998      3    200      -\n",
       "3       chr7   70496778   70496978      4    200      -\n",
       "4       chr2    2558151    2558351      5    200      -\n",
       "...      ...        ...        ...    ...    ...    ...\n",
       "30311   chrX  100943277  100943477  32996    200      -\n",
       "30312   chr5  102140158  102140358  32997    200      -\n",
       "30313  chr16    2979521    2979721  32998    200      +\n",
       "30314   chr3  141405823  141406023  32999    200      +\n",
       "30315  chr20    7284852    7285052  33000    200      +\n",
       "\n",
       "[30316 rows x 6 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_random = df_random[['chrom', 'start', 'end', 'name', 'score', 'strand']].copy()\n",
    "df_random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "5a4039a6-b1a7-4ec1-8bcf-a1b17765e025",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:49:10.244738Z",
     "iopub.status.busy": "2023-05-22T07:49:10.244026Z",
     "iopub.status.idle": "2023-05-22T07:49:11.006382Z",
     "shell.execute_reply": "2023-05-22T07:49:11.005513Z",
     "shell.execute_reply.started": "2023-05-22T07:49:10.244677Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>171273699</td>\n",
       "      <td>171273899</td>\n",
       "      <td>1</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr10</td>\n",
       "      <td>118814584</td>\n",
       "      <td>118814784</td>\n",
       "      <td>2</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr17</td>\n",
       "      <td>54532798</td>\n",
       "      <td>54532998</td>\n",
       "      <td>3</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr7</td>\n",
       "      <td>70496778</td>\n",
       "      <td>70496978</td>\n",
       "      <td>4</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr2</td>\n",
       "      <td>2558151</td>\n",
       "      <td>2558351</td>\n",
       "      <td>5</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30065</th>\n",
       "      <td>chrX</td>\n",
       "      <td>100943277</td>\n",
       "      <td>100943477</td>\n",
       "      <td>32996</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30066</th>\n",
       "      <td>chr5</td>\n",
       "      <td>102140158</td>\n",
       "      <td>102140358</td>\n",
       "      <td>32997</td>\n",
       "      <td>200</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30067</th>\n",
       "      <td>chr16</td>\n",
       "      <td>2979521</td>\n",
       "      <td>2979721</td>\n",
       "      <td>32998</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30068</th>\n",
       "      <td>chr3</td>\n",
       "      <td>141405823</td>\n",
       "      <td>141406023</td>\n",
       "      <td>32999</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30069</th>\n",
       "      <td>chr20</td>\n",
       "      <td>7284852</td>\n",
       "      <td>7285052</td>\n",
       "      <td>33000</td>\n",
       "      <td>200</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30070 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom      start        end   name  score strand\n",
       "0       chr1  171273699  171273899      1    200      +\n",
       "1      chr10  118814584  118814784      2    200      -\n",
       "2      chr17   54532798   54532998      3    200      -\n",
       "3       chr7   70496778   70496978      4    200      -\n",
       "4       chr2    2558151    2558351      5    200      -\n",
       "...      ...        ...        ...    ...    ...    ...\n",
       "30065   chrX  100943277  100943477  32996    200      -\n",
       "30066   chr5  102140158  102140358  32997    200      -\n",
       "30067  chr16    2979521    2979721  32998    200      +\n",
       "30068   chr3  141405823  141406023  32999    200      +\n",
       "30069  chr20    7284852    7285052  33000    200      +\n",
       "\n",
       "[30070 rows x 6 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rm overlapped with df_ctcf, df_offs\n",
    "a = BedTool.from_dataframe(df_ctcf.iloc[:,:3])\n",
    "b = BedTool.from_dataframe(df_random.iloc[:,:3])\n",
    "df_random = b.intersect(a, loj=True).to_dataframe().query('score==-1').iloc[:, :3].merge(df_random, on=['chrom', 'start', 'end'], how='left')\n",
    "\n",
    "\n",
    "\n",
    "df_offs = pd.concat(\n",
    "    [\n",
    "        df_old_offs.iloc[:,:3],\n",
    "        df_new_offs.iloc[:,:3]\n",
    "    ]\n",
    ")\n",
    "# df_offs\n",
    "\n",
    "\n",
    "\n",
    "a = BedTool.from_dataframe(df_offs.iloc[:,:3])\n",
    "b = BedTool.from_dataframe(df_random.iloc[:,:3])\n",
    "df_random = b.intersect(a, loj=True).to_dataframe().query('score==-1').iloc[:, :3].merge(df_random, on=['chrom', 'start', 'end'], how='left')\n",
    "\n",
    "\n",
    "df_random"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "7d5789b9-6d6c-495b-b838-8484759a260e",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:49:15.350304Z",
     "iopub.status.busy": "2023-05-22T07:49:15.349627Z",
     "iopub.status.idle": "2023-05-22T07:49:15.502590Z",
     "shell.execute_reply": "2023-05-22T07:49:15.501704Z",
     "shell.execute_reply.started": "2023-05-22T07:49:15.350251Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_random.to_csv(f'../bed_for_count/2023-05-12_random_for_count.bed', header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60046497-1014-4d63-8935-878f21c57849",
   "metadata": {},
   "source": [
    "### df_tss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f53b8036-968b-4e61-8aea-a7723a861a22",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:49:17.165251Z",
     "iopub.status.busy": "2023-05-22T07:49:17.164620Z",
     "iopub.status.idle": "2023-05-22T07:49:17.299278Z",
     "shell.execute_reply": "2023-05-22T07:49:17.298105Z",
     "shell.execute_reply.started": "2023-05-22T07:49:17.165196Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>63418</td>\n",
       "      <td>67418</td>\n",
       "      <td>NM_001005484</td>\n",
       "      <td>65418</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>449678</td>\n",
       "      <td>453678</td>\n",
       "      <td>NM_001005221</td>\n",
       "      <td>451678</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>684654</td>\n",
       "      <td>688654</td>\n",
       "      <td>NM_001005277</td>\n",
       "      <td>686654</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>921922</td>\n",
       "      <td>925922</td>\n",
       "      <td>NM_001385641</td>\n",
       "      <td>923922</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>923730</td>\n",
       "      <td>927730</td>\n",
       "      <td>NM_152486</td>\n",
       "      <td>925730</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28009</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24811393</td>\n",
       "      <td>24815393</td>\n",
       "      <td>NM_020364</td>\n",
       "      <td>24813393</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28010</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24831819</td>\n",
       "      <td>24835819</td>\n",
       "      <td>NM_020420</td>\n",
       "      <td>24833819</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28011</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24831918</td>\n",
       "      <td>24835918</td>\n",
       "      <td>NM_001388484</td>\n",
       "      <td>24833918</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28012</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25050104</td>\n",
       "      <td>25054104</td>\n",
       "      <td>NM_001002761</td>\n",
       "      <td>25052104</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28013</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25620116</td>\n",
       "      <td>25624116</td>\n",
       "      <td>NM_170723</td>\n",
       "      <td>25622116</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>28014 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom     start       end          name     score strand\n",
       "0      chr1     63418     67418  NM_001005484     65418      +\n",
       "1      chr1    449678    453678  NM_001005221    451678      -\n",
       "2      chr1    684654    688654  NM_001005277    686654      -\n",
       "3      chr1    921922    925922  NM_001385641    923922      +\n",
       "4      chr1    923730    927730     NM_152486    925730      +\n",
       "...     ...       ...       ...           ...       ...    ...\n",
       "28009  chrY  24811393  24815393     NM_020364  24813393      -\n",
       "28010  chrY  24831819  24835819     NM_020420  24833819      +\n",
       "28011  chrY  24831918  24835918  NM_001388484  24833918      +\n",
       "28012  chrY  25050104  25054104  NM_001002761  25052104      -\n",
       "28013  chrY  25620116  25624116     NM_170723  25622116      +\n",
       "\n",
       "[28014 rows x 6 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tss = pd.read_csv('ref_data/hg38_ucsc_refseq_NM_TSS_extend2kb.sorted.bed', sep='\\t', header=None,\n",
    "                    names=['chrom', 'start', 'end', 'name', 'score', 'strand'])\n",
    "df_tss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "80c21a1f-2512-4670-887a-a9f85a08c7fd",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:49:20.545970Z",
     "iopub.status.busy": "2023-05-22T07:49:20.545350Z",
     "iopub.status.idle": "2023-05-22T07:49:20.876424Z",
     "shell.execute_reply": "2023-05-22T07:49:20.874951Z",
     "shell.execute_reply.started": "2023-05-22T07:49:20.545913Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>63418</td>\n",
       "      <td>67418</td>\n",
       "      <td>NM_001005484</td>\n",
       "      <td>65418</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>449678</td>\n",
       "      <td>453678</td>\n",
       "      <td>NM_001005221</td>\n",
       "      <td>451678</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>684654</td>\n",
       "      <td>688654</td>\n",
       "      <td>NM_001005277</td>\n",
       "      <td>686654</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>921922</td>\n",
       "      <td>925922</td>\n",
       "      <td>NM_001385641</td>\n",
       "      <td>923922</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>923730</td>\n",
       "      <td>927730</td>\n",
       "      <td>NM_152486</td>\n",
       "      <td>925730</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25976</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24811393</td>\n",
       "      <td>24815393</td>\n",
       "      <td>NM_020364</td>\n",
       "      <td>24813393</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25977</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24831819</td>\n",
       "      <td>24835819</td>\n",
       "      <td>NM_020420</td>\n",
       "      <td>24833819</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25978</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24831918</td>\n",
       "      <td>24835918</td>\n",
       "      <td>NM_001388484</td>\n",
       "      <td>24833918</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25979</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25050104</td>\n",
       "      <td>25054104</td>\n",
       "      <td>NM_001002761</td>\n",
       "      <td>25052104</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25980</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25620116</td>\n",
       "      <td>25624116</td>\n",
       "      <td>NM_170723</td>\n",
       "      <td>25622116</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25981 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom     start       end          name     score strand\n",
       "0      chr1     63418     67418  NM_001005484     65418      +\n",
       "1      chr1    449678    453678  NM_001005221    451678      -\n",
       "2      chr1    684654    688654  NM_001005277    686654      -\n",
       "3      chr1    921922    925922  NM_001385641    923922      +\n",
       "4      chr1    923730    927730     NM_152486    925730      +\n",
       "...     ...       ...       ...           ...       ...    ...\n",
       "25976  chrY  24811393  24815393     NM_020364  24813393      -\n",
       "25977  chrY  24831819  24835819     NM_020420  24833819      +\n",
       "25978  chrY  24831918  24835918  NM_001388484  24833918      +\n",
       "25979  chrY  25050104  25054104  NM_001002761  25052104      -\n",
       "25980  chrY  25620116  25624116     NM_170723  25622116      +\n",
       "\n",
       "[25981 rows x 6 columns]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rm overlapped with df_offs\n",
    "df_offs = pd.concat(\n",
    "    [\n",
    "        df_old_offs.iloc[:,:3],\n",
    "        df_new_offs.iloc[:,:3]\n",
    "    ]\n",
    ")\n",
    "# df_offs\n",
    "a = BedTool.from_dataframe(df_offs.iloc[:,:3])\n",
    "b = BedTool.from_dataframe(df_tss.iloc[:,:3])\n",
    "df_tss = b.intersect(a, loj=True).to_dataframe().query('score==-1').iloc[:, :3].merge(df_tss, on=['chrom', 'start', 'end'], how='left')\n",
    "df_tss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "2fc6a202-cdaa-48ab-b8c5-fa02dc225e8b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:49:23.407175Z",
     "iopub.status.busy": "2023-05-22T07:49:23.406505Z",
     "iopub.status.idle": "2023-05-22T07:49:24.013435Z",
     "shell.execute_reply": "2023-05-22T07:49:24.012221Z",
     "shell.execute_reply.started": "2023-05-22T07:49:23.407117Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# with df_ctcf, without, df_ctcf\n",
    "a = BedTool.from_dataframe(df_ctcf.iloc[:,:3])\n",
    "b = BedTool.from_dataframe(df_tss.iloc[:,:3])\n",
    "df_tss_has_ctcf = b.intersect(a, loj=True).to_dataframe().query('score!=-1').iloc[:, :3].merge(df_tss, on=['chrom', 'start', 'end'], how='left')\n",
    "df_tss_no_ctcf = b.intersect(a, loj=True).to_dataframe().query('score==-1').iloc[:, :3].merge(df_tss, on=['chrom', 'start', 'end'], how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "6c83f6f5-4c38-46e8-8799-b8abadf0f512",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:49:24.015684Z",
     "iopub.status.busy": "2023-05-22T07:49:24.015335Z",
     "iopub.status.idle": "2023-05-22T07:49:24.040532Z",
     "shell.execute_reply": "2023-05-22T07:49:24.039755Z",
     "shell.execute_reply.started": "2023-05-22T07:49:24.015651Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>921922</td>\n",
       "      <td>925922</td>\n",
       "      <td>NM_001385641</td>\n",
       "      <td>923922</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>923730</td>\n",
       "      <td>927730</td>\n",
       "      <td>NM_152486</td>\n",
       "      <td>925730</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chr1</td>\n",
       "      <td>998097</td>\n",
       "      <td>1002097</td>\n",
       "      <td>NM_001142467</td>\n",
       "      <td>1000097</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1018119</td>\n",
       "      <td>1022119</td>\n",
       "      <td>NM_198576</td>\n",
       "      <td>1020119</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chr1</td>\n",
       "      <td>1114089</td>\n",
       "      <td>1118089</td>\n",
       "      <td>NM_001330306</td>\n",
       "      <td>1116089</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15822</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154749077</td>\n",
       "      <td>154753077</td>\n",
       "      <td>NM_001081573</td>\n",
       "      <td>154751077</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15823</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154760863</td>\n",
       "      <td>154764863</td>\n",
       "      <td>NM_001142463</td>\n",
       "      <td>154762863</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15824</th>\n",
       "      <td>chrX</td>\n",
       "      <td>154803485</td>\n",
       "      <td>154807485</td>\n",
       "      <td>NM_001166462</td>\n",
       "      <td>154805485</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15825</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155262491</td>\n",
       "      <td>155266491</td>\n",
       "      <td>NM_171998</td>\n",
       "      <td>155264491</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15826</th>\n",
       "      <td>chrX</td>\n",
       "      <td>155879344</td>\n",
       "      <td>155883344</td>\n",
       "      <td>NM_001145149</td>\n",
       "      <td>155881344</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10943 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom      start        end          name      score strand\n",
       "0      chr1     921922     925922  NM_001385641     923922      +\n",
       "3      chr1     923730     927730     NM_152486     925730      +\n",
       "5      chr1     998097    1002097  NM_001142467    1000097      -\n",
       "10     chr1    1018119    1022119     NM_198576    1020119      +\n",
       "11     chr1    1114089    1118089  NM_001330306    1116089      -\n",
       "...     ...        ...        ...           ...        ...    ...\n",
       "15822  chrX  154749077  154753077  NM_001081573  154751077      -\n",
       "15823  chrX  154760863  154764863  NM_001142463  154762863      +\n",
       "15824  chrX  154803485  154807485  NM_001166462  154805485      -\n",
       "15825  chrX  155262491  155266491     NM_171998  155264491      -\n",
       "15826  chrX  155879344  155883344  NM_001145149  155881344      +\n",
       "\n",
       "[10943 rows x 6 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tss_has_ctcf = df_tss_has_ctcf.drop_duplicates(keep='first')\n",
    "df_tss_has_ctcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "9041a46a-53aa-4dc5-ba66-3a58690039a7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:49:24.782545Z",
     "iopub.status.busy": "2023-05-22T07:49:24.781912Z",
     "iopub.status.idle": "2023-05-22T07:49:24.824826Z",
     "shell.execute_reply": "2023-05-22T07:49:24.823584Z",
     "shell.execute_reply.started": "2023-05-22T07:49:24.782490Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>63418</td>\n",
       "      <td>67418</td>\n",
       "      <td>NM_001005484</td>\n",
       "      <td>65418</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>449678</td>\n",
       "      <td>453678</td>\n",
       "      <td>NM_001005221</td>\n",
       "      <td>451678</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>684654</td>\n",
       "      <td>688654</td>\n",
       "      <td>NM_001005277</td>\n",
       "      <td>686654</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>957256</td>\n",
       "      <td>961256</td>\n",
       "      <td>NM_015658</td>\n",
       "      <td>959256</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>958583</td>\n",
       "      <td>962583</td>\n",
       "      <td>NM_198317</td>\n",
       "      <td>960583</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15033</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24811393</td>\n",
       "      <td>24815393</td>\n",
       "      <td>NM_020364</td>\n",
       "      <td>24813393</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15034</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24831819</td>\n",
       "      <td>24835819</td>\n",
       "      <td>NM_020420</td>\n",
       "      <td>24833819</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15035</th>\n",
       "      <td>chrY</td>\n",
       "      <td>24831918</td>\n",
       "      <td>24835918</td>\n",
       "      <td>NM_001388484</td>\n",
       "      <td>24833918</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15036</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25050104</td>\n",
       "      <td>25054104</td>\n",
       "      <td>NM_001002761</td>\n",
       "      <td>25052104</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15037</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25620116</td>\n",
       "      <td>25624116</td>\n",
       "      <td>NM_170723</td>\n",
       "      <td>25622116</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15038 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      chrom     start       end          name     score strand\n",
       "0      chr1     63418     67418  NM_001005484     65418      +\n",
       "1      chr1    449678    453678  NM_001005221    451678      -\n",
       "2      chr1    684654    688654  NM_001005277    686654      -\n",
       "3      chr1    957256    961256     NM_015658    959256      -\n",
       "4      chr1    958583    962583     NM_198317    960583      +\n",
       "...     ...       ...       ...           ...       ...    ...\n",
       "15033  chrY  24811393  24815393     NM_020364  24813393      -\n",
       "15034  chrY  24831819  24835819     NM_020420  24833819      +\n",
       "15035  chrY  24831918  24835918  NM_001388484  24833918      +\n",
       "15036  chrY  25050104  25054104  NM_001002761  25052104      -\n",
       "15037  chrY  25620116  25624116     NM_170723  25622116      +\n",
       "\n",
       "[15038 rows x 6 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_tss_no_ctcf = df_tss_no_ctcf.drop_duplicates(keep='first')\n",
    "df_tss_no_ctcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "ab3c2687-5d48-4c3b-a37a-bb2a0a0048e3",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:00:51.813193Z",
     "iopub.status.busy": "2023-05-22T08:00:51.812502Z",
     "iopub.status.idle": "2023-05-22T08:00:51.975562Z",
     "shell.execute_reply": "2023-05-22T08:00:51.974563Z",
     "shell.execute_reply.started": "2023-05-22T08:00:51.813135Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df_tss_has_ctcf.to_csv(f'../bed_for_count/2023-05-12_tss_has_ctcf_for_count.bed', header=False, index=False, sep='\\t')\n",
    "df_tss_no_ctcf.to_csv(f'../bed_for_count/2023-05-12_tss_no_ctcf_for_count.bed', header=False, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82c852e9-05cf-4297-872e-54abbcdb4b19",
   "metadata": {},
   "source": [
    "## 准备count-table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "43510c38-5e2a-425d-9003-600e871ab8f8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:02:21.564599Z",
     "iopub.status.busy": "2023-05-22T08:02:21.563933Z",
     "iopub.status.idle": "2023-05-22T08:02:21.572307Z",
     "shell.execute_reply": "2023-05-22T08:02:21.570372Z",
     "shell.execute_reply.started": "2023-05-22T08:02:21.564542Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from glob import glob"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "a78c4b7d-9109-407b-9e2d-59e8509699d8",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:02:21.933138Z",
     "iopub.status.busy": "2023-05-22T08:02:21.932504Z",
     "iopub.status.idle": "2023-05-22T08:02:21.940112Z",
     "shell.execute_reply": "2023-05-22T08:02:21.938161Z",
     "shell.execute_reply.started": "2023-05-22T08:02:21.933066Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# ls_bam = sorted(glob('../bam/ATACSeq_*.sortp_genome_rmchrM_rmdup.bam'))\n",
    "# ls_bam\n",
    "# dt_bam = {i.split('Seq_')[-1].split('.sortp')[0]: i for i in ls_bam}\n",
    "# dt_bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "50d67f0e-b90f-404d-8534-d687cd11b2e2",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:02:22.408065Z",
     "iopub.status.busy": "2023-05-22T08:02:22.407451Z",
     "iopub.status.idle": "2023-05-22T08:02:22.415065Z",
     "shell.execute_reply": "2023-05-22T08:02:22.413170Z",
     "shell.execute_reply.started": "2023-05-22T08:02:22.408010Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# ls_bed = sorted(glob('../bed_for_count/*.bed'))\n",
    "# ls_bed\n",
    "# dt_bed = {i.split('2023-05-12_')[-1].split('_for_count')[0]: i for i in ls_bed}\n",
    "# dt_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "f49bcce9-eeec-483f-a6bd-db9d6a3dc392",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:02:22.691282Z",
     "iopub.status.busy": "2023-05-22T08:02:22.690484Z",
     "iopub.status.idle": "2023-05-22T08:02:22.698555Z",
     "shell.execute_reply": "2023-05-22T08:02:22.696611Z",
     "shell.execute_reply.started": "2023-05-22T08:02:22.691227Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# for bed_name, bed_path in dt_bed.items():\n",
    "#     for bam_name, bam_path in dt_bam.items():\n",
    "#         # print(bed_name)\n",
    "#         pass\n",
    "#         # print(bam_name)\n",
    "#     #     count_table = f\"../bed_for_count/AnalysisChanges_BAM_{bam_name}_BED_{bed_name}.count_table\"\n",
    "#     #     cmd = f\"sleep 1\\necho submit {count_table}\\npkurun-cnlong 1 10 bedtools coverage -a {bed_path} -b {bam_path} > {count_table}\\n\"\n",
    "#         # print(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "56af4d09-6dc7-4bbd-ae07-3c03c180825f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:02:22.875592Z",
     "iopub.status.busy": "2023-05-22T08:02:22.874903Z",
     "iopub.status.idle": "2023-05-22T08:02:22.887182Z",
     "shell.execute_reply": "2023-05-22T08:02:22.885593Z",
     "shell.execute_reply.started": "2023-05-22T08:02:22.875531Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['GFP-NLS_REP-1', 'GFP_REP-1', 'ND6-DddAwt_REP-1', 'ND6-DddAwt_REP-2', 'SIRT6-DddA11_REP-1', 'SIRT6-DddA11_REP-2']\n",
      "['CTCF_medium', 'CTCF_strong', 'CTCF_weak', 'CTCF_within-offs', 'offs-ND6-DEP', 'offs-ND6-IND', 'random', 'tss_has_ctcf', 'tss_no_ctcf']\n"
     ]
    }
   ],
   "source": [
    "ls_bam = sorted(glob('../bam/ATACSeq_*.sortp_genome_rmchrM_rmdup.bam'))\n",
    "bams = [i.split('Seq_')[-1].split('.sortp')[0] for i in ls_bam]\n",
    "ls_bed = sorted(glob('../bed_for_count/*.bed'))\n",
    "beds = [i.split('2023-05-12_')[-1].split('_for_count')[0] for i in ls_bed]\n",
    "print(bams)\n",
    "print(beds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "0acf7947-d142-4738-9ad6-0981a0f02c69",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:49:31.583291Z",
     "iopub.status.busy": "2023-05-22T07:49:31.582658Z",
     "iopub.status.idle": "2023-05-22T07:49:31.590214Z",
     "shell.execute_reply": "2023-05-22T07:49:31.588273Z",
     "shell.execute_reply.started": "2023-05-22T07:49:31.583236Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "# # run snakemake\n",
    "# BAMs = ['GFP-NLS_REP-1', 'GFP_REP-1', 'ND6-DddAwt_REP-1', 'ND6-DddAwt_REP-2', 'SIRT6-DddA11_REP-1', 'SIRT6-DddA11_REP-2']\n",
    "# BEDs = ['CTCF_medium', 'CTCF_strong', 'CTCF_weak', 'CTCF_within-offs', 'offs-ND6-DEP', 'offs-ND6-IND', 'random']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75a80f24-215f-4b18-a4dd-d7a222973ba1",
   "metadata": {},
   "source": [
    "## 分析比较ATAC-seq数据变化"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "e03dccd0-741a-4021-9c33-20d9e91a91b4",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:02:38.754705Z",
     "iopub.status.busy": "2023-05-22T08:02:38.754064Z",
     "iopub.status.idle": "2023-05-22T08:02:38.762214Z",
     "shell.execute_reply": "2023-05-22T08:02:38.760224Z",
     "shell.execute_reply.started": "2023-05-22T08:02:38.754650Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "from glob import glob\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "ce46ada3-7815-49af-b41c-4db336720917",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:02:39.228703Z",
     "iopub.status.busy": "2023-05-22T08:02:39.228074Z",
     "iopub.status.idle": "2023-05-22T08:02:39.236927Z",
     "shell.execute_reply": "2023-05-22T08:02:39.235063Z",
     "shell.execute_reply.started": "2023-05-22T08:02:39.228648Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "ls_counts = sorted(glob('../bed_for_count/AnalysisChanges_*.count_table'))\n",
    "# ls_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "0a4fbe4b-84aa-4344-8aa1-1d8900e91d7a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:02:39.783737Z",
     "iopub.status.busy": "2023-05-22T08:02:39.783124Z",
     "iopub.status.idle": "2023-05-22T08:02:39.792168Z",
     "shell.execute_reply": "2023-05-22T08:02:39.790237Z",
     "shell.execute_reply.started": "2023-05-22T08:02:39.783683Z"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "df = pd.DataFrame(ls_counts, columns=['file_path'])\n",
    "# df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "ca650645-7be6-44c0-bdc9-7f058505a646",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:02:40.278694Z",
     "iopub.status.busy": "2023-05-22T08:02:40.278125Z",
     "iopub.status.idle": "2023-05-22T08:02:40.311183Z",
     "shell.execute_reply": "2023-05-22T08:02:40.309907Z",
     "shell.execute_reply.started": "2023-05-22T08:02:40.278643Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_path</th>\n",
       "      <th>sample</th>\n",
       "      <th>region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_GFP-NLS_R...</td>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>CTCF_medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_GFP-NLS_R...</td>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>CTCF_strong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_GFP-NLS_R...</td>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>CTCF_weak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_GFP-NLS_R...</td>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>CTCF_within-offs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_GFP-NLS_R...</td>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>offs-ND6-DEP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_GFP-NLS_R...</td>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>offs-ND6-IND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_GFP-NLS_R...</td>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_GFP-NLS_R...</td>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>tss_has_ctcf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_GFP-NLS_R...</td>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_GFP_REP-1...</td>\n",
       "      <td>GFP_REP-1</td>\n",
       "      <td>CTCF_medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_GFP_REP-1...</td>\n",
       "      <td>GFP_REP-1</td>\n",
       "      <td>CTCF_strong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_GFP_REP-1...</td>\n",
       "      <td>GFP_REP-1</td>\n",
       "      <td>CTCF_weak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_GFP_REP-1...</td>\n",
       "      <td>GFP_REP-1</td>\n",
       "      <td>CTCF_within-offs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_GFP_REP-1...</td>\n",
       "      <td>GFP_REP-1</td>\n",
       "      <td>offs-ND6-DEP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_GFP_REP-1...</td>\n",
       "      <td>GFP_REP-1</td>\n",
       "      <td>offs-ND6-IND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_GFP_REP-1...</td>\n",
       "      <td>GFP_REP-1</td>\n",
       "      <td>random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_GFP_REP-1...</td>\n",
       "      <td>GFP_REP-1</td>\n",
       "      <td>tss_has_ctcf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_GFP_REP-1...</td>\n",
       "      <td>GFP_REP-1</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...</td>\n",
       "      <td>ND6-DddAwt_REP-1</td>\n",
       "      <td>CTCF_medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...</td>\n",
       "      <td>ND6-DddAwt_REP-1</td>\n",
       "      <td>CTCF_strong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...</td>\n",
       "      <td>ND6-DddAwt_REP-1</td>\n",
       "      <td>CTCF_weak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...</td>\n",
       "      <td>ND6-DddAwt_REP-1</td>\n",
       "      <td>CTCF_within-offs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...</td>\n",
       "      <td>ND6-DddAwt_REP-1</td>\n",
       "      <td>offs-ND6-DEP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...</td>\n",
       "      <td>ND6-DddAwt_REP-1</td>\n",
       "      <td>offs-ND6-IND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...</td>\n",
       "      <td>ND6-DddAwt_REP-1</td>\n",
       "      <td>random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...</td>\n",
       "      <td>ND6-DddAwt_REP-1</td>\n",
       "      <td>tss_has_ctcf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...</td>\n",
       "      <td>ND6-DddAwt_REP-1</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...</td>\n",
       "      <td>ND6-DddAwt_REP-2</td>\n",
       "      <td>CTCF_medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...</td>\n",
       "      <td>ND6-DddAwt_REP-2</td>\n",
       "      <td>CTCF_strong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...</td>\n",
       "      <td>ND6-DddAwt_REP-2</td>\n",
       "      <td>CTCF_weak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...</td>\n",
       "      <td>ND6-DddAwt_REP-2</td>\n",
       "      <td>CTCF_within-offs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...</td>\n",
       "      <td>ND6-DddAwt_REP-2</td>\n",
       "      <td>offs-ND6-DEP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...</td>\n",
       "      <td>ND6-DddAwt_REP-2</td>\n",
       "      <td>offs-ND6-IND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...</td>\n",
       "      <td>ND6-DddAwt_REP-2</td>\n",
       "      <td>random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...</td>\n",
       "      <td>ND6-DddAwt_REP-2</td>\n",
       "      <td>tss_has_ctcf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...</td>\n",
       "      <td>ND6-DddAwt_REP-2</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...</td>\n",
       "      <td>SIRT6-DddA11_REP-1</td>\n",
       "      <td>CTCF_medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...</td>\n",
       "      <td>SIRT6-DddA11_REP-1</td>\n",
       "      <td>CTCF_strong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...</td>\n",
       "      <td>SIRT6-DddA11_REP-1</td>\n",
       "      <td>CTCF_weak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...</td>\n",
       "      <td>SIRT6-DddA11_REP-1</td>\n",
       "      <td>CTCF_within-offs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...</td>\n",
       "      <td>SIRT6-DddA11_REP-1</td>\n",
       "      <td>offs-ND6-DEP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...</td>\n",
       "      <td>SIRT6-DddA11_REP-1</td>\n",
       "      <td>offs-ND6-IND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...</td>\n",
       "      <td>SIRT6-DddA11_REP-1</td>\n",
       "      <td>random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...</td>\n",
       "      <td>SIRT6-DddA11_REP-1</td>\n",
       "      <td>tss_has_ctcf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...</td>\n",
       "      <td>SIRT6-DddA11_REP-1</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...</td>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>CTCF_medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...</td>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>CTCF_strong</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...</td>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>CTCF_weak</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...</td>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>CTCF_within-offs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...</td>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>offs-ND6-DEP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...</td>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>offs-ND6-IND</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...</td>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>random</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...</td>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>tss_has_ctcf</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...</td>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            file_path              sample   \n",
       "0   ../bed_for_count/AnalysisChanges_BAM_GFP-NLS_R...       GFP-NLS_REP-1  \\\n",
       "1   ../bed_for_count/AnalysisChanges_BAM_GFP-NLS_R...       GFP-NLS_REP-1   \n",
       "2   ../bed_for_count/AnalysisChanges_BAM_GFP-NLS_R...       GFP-NLS_REP-1   \n",
       "3   ../bed_for_count/AnalysisChanges_BAM_GFP-NLS_R...       GFP-NLS_REP-1   \n",
       "4   ../bed_for_count/AnalysisChanges_BAM_GFP-NLS_R...       GFP-NLS_REP-1   \n",
       "5   ../bed_for_count/AnalysisChanges_BAM_GFP-NLS_R...       GFP-NLS_REP-1   \n",
       "6   ../bed_for_count/AnalysisChanges_BAM_GFP-NLS_R...       GFP-NLS_REP-1   \n",
       "7   ../bed_for_count/AnalysisChanges_BAM_GFP-NLS_R...       GFP-NLS_REP-1   \n",
       "8   ../bed_for_count/AnalysisChanges_BAM_GFP-NLS_R...       GFP-NLS_REP-1   \n",
       "9   ../bed_for_count/AnalysisChanges_BAM_GFP_REP-1...           GFP_REP-1   \n",
       "10  ../bed_for_count/AnalysisChanges_BAM_GFP_REP-1...           GFP_REP-1   \n",
       "11  ../bed_for_count/AnalysisChanges_BAM_GFP_REP-1...           GFP_REP-1   \n",
       "12  ../bed_for_count/AnalysisChanges_BAM_GFP_REP-1...           GFP_REP-1   \n",
       "13  ../bed_for_count/AnalysisChanges_BAM_GFP_REP-1...           GFP_REP-1   \n",
       "14  ../bed_for_count/AnalysisChanges_BAM_GFP_REP-1...           GFP_REP-1   \n",
       "15  ../bed_for_count/AnalysisChanges_BAM_GFP_REP-1...           GFP_REP-1   \n",
       "16  ../bed_for_count/AnalysisChanges_BAM_GFP_REP-1...           GFP_REP-1   \n",
       "17  ../bed_for_count/AnalysisChanges_BAM_GFP_REP-1...           GFP_REP-1   \n",
       "18  ../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...    ND6-DddAwt_REP-1   \n",
       "19  ../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...    ND6-DddAwt_REP-1   \n",
       "20  ../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...    ND6-DddAwt_REP-1   \n",
       "21  ../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...    ND6-DddAwt_REP-1   \n",
       "22  ../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...    ND6-DddAwt_REP-1   \n",
       "23  ../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...    ND6-DddAwt_REP-1   \n",
       "24  ../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...    ND6-DddAwt_REP-1   \n",
       "25  ../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...    ND6-DddAwt_REP-1   \n",
       "26  ../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...    ND6-DddAwt_REP-1   \n",
       "27  ../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...    ND6-DddAwt_REP-2   \n",
       "28  ../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...    ND6-DddAwt_REP-2   \n",
       "29  ../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...    ND6-DddAwt_REP-2   \n",
       "30  ../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...    ND6-DddAwt_REP-2   \n",
       "31  ../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...    ND6-DddAwt_REP-2   \n",
       "32  ../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...    ND6-DddAwt_REP-2   \n",
       "33  ../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...    ND6-DddAwt_REP-2   \n",
       "34  ../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...    ND6-DddAwt_REP-2   \n",
       "35  ../bed_for_count/AnalysisChanges_BAM_ND6-DddAw...    ND6-DddAwt_REP-2   \n",
       "36  ../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...  SIRT6-DddA11_REP-1   \n",
       "37  ../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...  SIRT6-DddA11_REP-1   \n",
       "38  ../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...  SIRT6-DddA11_REP-1   \n",
       "39  ../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...  SIRT6-DddA11_REP-1   \n",
       "40  ../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...  SIRT6-DddA11_REP-1   \n",
       "41  ../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...  SIRT6-DddA11_REP-1   \n",
       "42  ../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...  SIRT6-DddA11_REP-1   \n",
       "43  ../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...  SIRT6-DddA11_REP-1   \n",
       "44  ../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...  SIRT6-DddA11_REP-1   \n",
       "45  ../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...  SIRT6-DddA11_REP-2   \n",
       "46  ../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...  SIRT6-DddA11_REP-2   \n",
       "47  ../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...  SIRT6-DddA11_REP-2   \n",
       "48  ../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...  SIRT6-DddA11_REP-2   \n",
       "49  ../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...  SIRT6-DddA11_REP-2   \n",
       "50  ../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...  SIRT6-DddA11_REP-2   \n",
       "51  ../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...  SIRT6-DddA11_REP-2   \n",
       "52  ../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...  SIRT6-DddA11_REP-2   \n",
       "53  ../bed_for_count/AnalysisChanges_BAM_SIRT6-Ddd...  SIRT6-DddA11_REP-2   \n",
       "\n",
       "              region  \n",
       "0        CTCF_medium  \n",
       "1        CTCF_strong  \n",
       "2          CTCF_weak  \n",
       "3   CTCF_within-offs  \n",
       "4       offs-ND6-DEP  \n",
       "5       offs-ND6-IND  \n",
       "6             random  \n",
       "7       tss_has_ctcf  \n",
       "8        tss_no_ctcf  \n",
       "9        CTCF_medium  \n",
       "10       CTCF_strong  \n",
       "11         CTCF_weak  \n",
       "12  CTCF_within-offs  \n",
       "13      offs-ND6-DEP  \n",
       "14      offs-ND6-IND  \n",
       "15            random  \n",
       "16      tss_has_ctcf  \n",
       "17       tss_no_ctcf  \n",
       "18       CTCF_medium  \n",
       "19       CTCF_strong  \n",
       "20         CTCF_weak  \n",
       "21  CTCF_within-offs  \n",
       "22      offs-ND6-DEP  \n",
       "23      offs-ND6-IND  \n",
       "24            random  \n",
       "25      tss_has_ctcf  \n",
       "26       tss_no_ctcf  \n",
       "27       CTCF_medium  \n",
       "28       CTCF_strong  \n",
       "29         CTCF_weak  \n",
       "30  CTCF_within-offs  \n",
       "31      offs-ND6-DEP  \n",
       "32      offs-ND6-IND  \n",
       "33            random  \n",
       "34      tss_has_ctcf  \n",
       "35       tss_no_ctcf  \n",
       "36       CTCF_medium  \n",
       "37       CTCF_strong  \n",
       "38         CTCF_weak  \n",
       "39  CTCF_within-offs  \n",
       "40      offs-ND6-DEP  \n",
       "41      offs-ND6-IND  \n",
       "42            random  \n",
       "43      tss_has_ctcf  \n",
       "44       tss_no_ctcf  \n",
       "45       CTCF_medium  \n",
       "46       CTCF_strong  \n",
       "47         CTCF_weak  \n",
       "48  CTCF_within-offs  \n",
       "49      offs-ND6-DEP  \n",
       "50      offs-ND6-IND  \n",
       "51            random  \n",
       "52      tss_has_ctcf  \n",
       "53       tss_no_ctcf  "
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['sample'] = df.file_path.str.split('BAM_').str[-1].str.split('_BED').str[0]\n",
    "df['region'] = df.file_path.str.split('BED_').str[-1].str.split('.count').str[0]\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "594d00cc-2554-4895-9fd5-7ccbfe8ee763",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:02:41.619388Z",
     "iopub.status.busy": "2023-05-22T08:02:41.618746Z",
     "iopub.status.idle": "2023-05-22T08:02:41.632352Z",
     "shell.execute_reply": "2023-05-22T08:02:41.630621Z",
     "shell.execute_reply.started": "2023-05-22T08:02:41.619332Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CTCF_medium',\n",
       " 'CTCF_strong',\n",
       " 'CTCF_weak',\n",
       " 'CTCF_within-offs',\n",
       " 'offs-ND6-DEP',\n",
       " 'offs-ND6-IND',\n",
       " 'random',\n",
       " 'tss_has_ctcf',\n",
       " 'tss_no_ctcf']"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "samples = sorted(df['sample'].unique())\n",
    "regions = sorted(df['region'].unique())\n",
    "regions\n",
    "# samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "1b26e426-910c-4770-9e76-d91784ca88a7",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:02:44.100321Z",
     "iopub.status.busy": "2023-05-22T08:02:44.099699Z",
     "iopub.status.idle": "2023-05-22T08:02:44.133130Z",
     "shell.execute_reply": "2023-05-22T08:02:44.131405Z",
     "shell.execute_reply.started": "2023-05-22T08:02:44.100266Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>total_reads_M</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>28.781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GFP_REP-1</td>\n",
       "      <td>24.925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ND6-DddAwt_REP-1</td>\n",
       "      <td>24.676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ND6-DddAwt_REP-2</td>\n",
       "      <td>32.489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIRT6-DddA11_REP-1</td>\n",
       "      <td>27.194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>26.968</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample  total_reads_M\n",
       "0       GFP-NLS_REP-1         28.781\n",
       "1           GFP_REP-1         24.925\n",
       "2    ND6-DddAwt_REP-1         24.676\n",
       "3    ND6-DddAwt_REP-2         32.489\n",
       "4  SIRT6-DddA11_REP-1         27.194\n",
       "5  SIRT6-DddA11_REP-2         26.968"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_total_reads = pd.read_csv('2023-05-12_step.03.MappingQC_result.csv', index_col=0).reset_index()\n",
    "df_total_reads = df_total_reads[['index', 'map to genome remove duplicates (M)']].copy()\n",
    "df_total_reads.columns = ['sample', 'total_reads_M']\n",
    "df_total_reads = df_total_reads.query('sample!=\"test\"').copy()\n",
    "df_total_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "fa24549a-2446-4911-9d38-ce7399ecc942",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:02:45.285355Z",
     "iopub.status.busy": "2023-05-22T08:02:45.284789Z",
     "iopub.status.idle": "2023-05-22T08:02:46.469124Z",
     "shell.execute_reply": "2023-05-22T08:02:46.468269Z",
     "shell.execute_reply.started": "2023-05-22T08:02:45.285303Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>region</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>depth</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>chr1</td>\n",
       "      <td>778863</td>\n",
       "      <td>779287</td>\n",
       "      <td>chr1_778863_779287_medium_peak</td>\n",
       "      <td>2.778652e+00</td>\n",
       "      <td>.</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>chr1</td>\n",
       "      <td>919511</td>\n",
       "      <td>919935</td>\n",
       "      <td>chr1_919511_919935_medium_peak</td>\n",
       "      <td>2.369525e+00</td>\n",
       "      <td>.</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>chr1</td>\n",
       "      <td>923010</td>\n",
       "      <td>923434</td>\n",
       "      <td>chr1_923010_923434_medium_peak</td>\n",
       "      <td>3.034356e+00</td>\n",
       "      <td>.</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>chr1</td>\n",
       "      <td>939801</td>\n",
       "      <td>940225</td>\n",
       "      <td>chr1_939801_940225_medium_peak</td>\n",
       "      <td>2.608183e+00</td>\n",
       "      <td>.</td>\n",
       "      <td>69</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>chr1</td>\n",
       "      <td>951368</td>\n",
       "      <td>951792</td>\n",
       "      <td>chr1_951368_951792_medium_peak</td>\n",
       "      <td>2.659323e+00</td>\n",
       "      <td>.</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15033</th>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>chrY</td>\n",
       "      <td>24811393</td>\n",
       "      <td>24815393</td>\n",
       "      <td>NM_020364</td>\n",
       "      <td>2.481339e+07</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15034</th>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>chrY</td>\n",
       "      <td>24831819</td>\n",
       "      <td>24835819</td>\n",
       "      <td>NM_020420</td>\n",
       "      <td>2.483382e+07</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15035</th>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>chrY</td>\n",
       "      <td>24831918</td>\n",
       "      <td>24835918</td>\n",
       "      <td>NM_001388484</td>\n",
       "      <td>2.483392e+07</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15036</th>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>chrY</td>\n",
       "      <td>25050104</td>\n",
       "      <td>25054104</td>\n",
       "      <td>NM_001002761</td>\n",
       "      <td>2.505210e+07</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15037</th>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>chrY</td>\n",
       "      <td>25620116</td>\n",
       "      <td>25624116</td>\n",
       "      <td>NM_170723</td>\n",
       "      <td>2.562212e+07</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>570432 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   sample       region chrom     start       end   \n",
       "0           GFP-NLS_REP-1  CTCF_medium  chr1    778863    779287  \\\n",
       "1           GFP-NLS_REP-1  CTCF_medium  chr1    919511    919935   \n",
       "2           GFP-NLS_REP-1  CTCF_medium  chr1    923010    923434   \n",
       "3           GFP-NLS_REP-1  CTCF_medium  chr1    939801    940225   \n",
       "4           GFP-NLS_REP-1  CTCF_medium  chr1    951368    951792   \n",
       "...                   ...          ...   ...       ...       ...   \n",
       "15033  SIRT6-DddA11_REP-2  tss_no_ctcf  chrY  24811393  24815393   \n",
       "15034  SIRT6-DddA11_REP-2  tss_no_ctcf  chrY  24831819  24835819   \n",
       "15035  SIRT6-DddA11_REP-2  tss_no_ctcf  chrY  24831918  24835918   \n",
       "15036  SIRT6-DddA11_REP-2  tss_no_ctcf  chrY  25050104  25054104   \n",
       "15037  SIRT6-DddA11_REP-2  tss_no_ctcf  chrY  25620116  25624116   \n",
       "\n",
       "                                 name         score strand  depth  \n",
       "0      chr1_778863_779287_medium_peak  2.778652e+00      .    203  \n",
       "1      chr1_919511_919935_medium_peak  2.369525e+00      .     18  \n",
       "2      chr1_923010_923434_medium_peak  3.034356e+00      .     38  \n",
       "3      chr1_939801_940225_medium_peak  2.608183e+00      .     69  \n",
       "4      chr1_951368_951792_medium_peak  2.659323e+00      .     41  \n",
       "...                               ...           ...    ...    ...  \n",
       "15033                       NM_020364  2.481339e+07      -      0  \n",
       "15034                       NM_020420  2.483382e+07      +      0  \n",
       "15035                    NM_001388484  2.483392e+07      +      0  \n",
       "15036                    NM_001002761  2.505210e+07      -      0  \n",
       "15037                       NM_170723  2.562212e+07      +      0  \n",
       "\n",
       "[570432 rows x 9 columns]"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ls = []\n",
    "for idx, line in df.iterrows():\n",
    "    file_path, sample, region = line\n",
    "    # print(file_path, sample, region)\n",
    "    tmpdf = pd.read_csv(file_path, sep='\\t', header=None)\n",
    "    tmpdf['sample'] = sample\n",
    "    tmpdf['region'] = region\n",
    "    ls.append(tmpdf)\n",
    "df_counts = pd.concat(ls)\n",
    "df_counts.columns = ['chrom', 'start', 'end', 'name', 'score', 'strand', 'depth', '_1', '_2', '_3', 'sample', 'region']\n",
    "df_counts = df_counts[['sample', 'region', 'chrom', 'start', 'end', 'name', 'score', 'strand', 'depth']].copy()\n",
    "df_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "3c9e4ec3-4ac9-4b62-84ea-b4a98a5f2c2d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:02:52.615176Z",
     "iopub.status.busy": "2023-05-22T08:02:52.614520Z",
     "iopub.status.idle": "2023-05-22T08:02:52.633282Z",
     "shell.execute_reply": "2023-05-22T08:02:52.631501Z",
     "shell.execute_reply.started": "2023-05-22T08:02:52.615121Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>total_reads_M</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>28.781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GFP_REP-1</td>\n",
       "      <td>24.925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ND6-DddAwt_REP-1</td>\n",
       "      <td>24.676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ND6-DddAwt_REP-2</td>\n",
       "      <td>32.489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SIRT6-DddA11_REP-1</td>\n",
       "      <td>27.194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>26.968</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               sample  total_reads_M\n",
       "0       GFP-NLS_REP-1         28.781\n",
       "1           GFP_REP-1         24.925\n",
       "2    ND6-DddAwt_REP-1         24.676\n",
       "3    ND6-DddAwt_REP-2         32.489\n",
       "4  SIRT6-DddA11_REP-1         27.194\n",
       "5  SIRT6-DddA11_REP-2         26.968"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_total_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "021834d3-c142-4fcc-94e8-29f92efe9111",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:02:52.973260Z",
     "iopub.status.busy": "2023-05-22T08:02:52.972683Z",
     "iopub.status.idle": "2023-05-22T08:02:53.124793Z",
     "shell.execute_reply": "2023-05-22T08:02:53.123946Z",
     "shell.execute_reply.started": "2023-05-22T08:02:52.973210Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>region</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>depth</th>\n",
       "      <th>total_reads_M</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>chr1</td>\n",
       "      <td>778863</td>\n",
       "      <td>779287</td>\n",
       "      <td>chr1_778863_779287_medium_peak</td>\n",
       "      <td>2.778652e+00</td>\n",
       "      <td>.</td>\n",
       "      <td>203</td>\n",
       "      <td>28.781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>chr1</td>\n",
       "      <td>919511</td>\n",
       "      <td>919935</td>\n",
       "      <td>chr1_919511_919935_medium_peak</td>\n",
       "      <td>2.369525e+00</td>\n",
       "      <td>.</td>\n",
       "      <td>18</td>\n",
       "      <td>28.781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>chr1</td>\n",
       "      <td>923010</td>\n",
       "      <td>923434</td>\n",
       "      <td>chr1_923010_923434_medium_peak</td>\n",
       "      <td>3.034356e+00</td>\n",
       "      <td>.</td>\n",
       "      <td>38</td>\n",
       "      <td>28.781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>chr1</td>\n",
       "      <td>939801</td>\n",
       "      <td>940225</td>\n",
       "      <td>chr1_939801_940225_medium_peak</td>\n",
       "      <td>2.608183e+00</td>\n",
       "      <td>.</td>\n",
       "      <td>69</td>\n",
       "      <td>28.781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>chr1</td>\n",
       "      <td>951368</td>\n",
       "      <td>951792</td>\n",
       "      <td>chr1_951368_951792_medium_peak</td>\n",
       "      <td>2.659323e+00</td>\n",
       "      <td>.</td>\n",
       "      <td>41</td>\n",
       "      <td>28.781</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570427</th>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>chrY</td>\n",
       "      <td>24811393</td>\n",
       "      <td>24815393</td>\n",
       "      <td>NM_020364</td>\n",
       "      <td>2.481339e+07</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>26.968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570428</th>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>chrY</td>\n",
       "      <td>24831819</td>\n",
       "      <td>24835819</td>\n",
       "      <td>NM_020420</td>\n",
       "      <td>2.483382e+07</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>26.968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570429</th>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>chrY</td>\n",
       "      <td>24831918</td>\n",
       "      <td>24835918</td>\n",
       "      <td>NM_001388484</td>\n",
       "      <td>2.483392e+07</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>26.968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570430</th>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>chrY</td>\n",
       "      <td>25050104</td>\n",
       "      <td>25054104</td>\n",
       "      <td>NM_001002761</td>\n",
       "      <td>2.505210e+07</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>26.968</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>570431</th>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>chrY</td>\n",
       "      <td>25620116</td>\n",
       "      <td>25624116</td>\n",
       "      <td>NM_170723</td>\n",
       "      <td>2.562212e+07</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>26.968</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>570432 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    sample       region chrom     start       end   \n",
       "0            GFP-NLS_REP-1  CTCF_medium  chr1    778863    779287  \\\n",
       "1            GFP-NLS_REP-1  CTCF_medium  chr1    919511    919935   \n",
       "2            GFP-NLS_REP-1  CTCF_medium  chr1    923010    923434   \n",
       "3            GFP-NLS_REP-1  CTCF_medium  chr1    939801    940225   \n",
       "4            GFP-NLS_REP-1  CTCF_medium  chr1    951368    951792   \n",
       "...                    ...          ...   ...       ...       ...   \n",
       "570427  SIRT6-DddA11_REP-2  tss_no_ctcf  chrY  24811393  24815393   \n",
       "570428  SIRT6-DddA11_REP-2  tss_no_ctcf  chrY  24831819  24835819   \n",
       "570429  SIRT6-DddA11_REP-2  tss_no_ctcf  chrY  24831918  24835918   \n",
       "570430  SIRT6-DddA11_REP-2  tss_no_ctcf  chrY  25050104  25054104   \n",
       "570431  SIRT6-DddA11_REP-2  tss_no_ctcf  chrY  25620116  25624116   \n",
       "\n",
       "                                  name         score strand  depth   \n",
       "0       chr1_778863_779287_medium_peak  2.778652e+00      .    203  \\\n",
       "1       chr1_919511_919935_medium_peak  2.369525e+00      .     18   \n",
       "2       chr1_923010_923434_medium_peak  3.034356e+00      .     38   \n",
       "3       chr1_939801_940225_medium_peak  2.608183e+00      .     69   \n",
       "4       chr1_951368_951792_medium_peak  2.659323e+00      .     41   \n",
       "...                                ...           ...    ...    ...   \n",
       "570427                       NM_020364  2.481339e+07      -      0   \n",
       "570428                       NM_020420  2.483382e+07      +      0   \n",
       "570429                    NM_001388484  2.483392e+07      +      0   \n",
       "570430                    NM_001002761  2.505210e+07      -      0   \n",
       "570431                       NM_170723  2.562212e+07      +      0   \n",
       "\n",
       "        total_reads_M  \n",
       "0              28.781  \n",
       "1              28.781  \n",
       "2              28.781  \n",
       "3              28.781  \n",
       "4              28.781  \n",
       "...               ...  \n",
       "570427         26.968  \n",
       "570428         26.968  \n",
       "570429         26.968  \n",
       "570430         26.968  \n",
       "570431         26.968  \n",
       "\n",
       "[570432 rows x 10 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_counts = df_counts.merge(df_total_reads, on='sample', how='left')\n",
    "df_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "3e4a4e0d-7ef7-4ac7-b941-852eca920604",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:02:53.703967Z",
     "iopub.status.busy": "2023-05-22T08:02:53.703341Z",
     "iopub.status.idle": "2023-05-22T08:02:55.032640Z",
     "shell.execute_reply": "2023-05-22T08:02:55.031938Z",
     "shell.execute_reply.started": "2023-05-22T08:02:53.703896Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>sample</th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>depth</th>\n",
       "      <th>total_reads_M</th>\n",
       "      <th>region_length</th>\n",
       "      <th>RPKM</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>region_index</th>\n",
       "      <th>region</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr1_778863_779287</th>\n",
       "      <th>CTCF_medium</th>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>778863</td>\n",
       "      <td>779287</td>\n",
       "      <td>chr1_778863_779287_medium_peak</td>\n",
       "      <td>2.778652e+00</td>\n",
       "      <td>.</td>\n",
       "      <td>203</td>\n",
       "      <td>28.781</td>\n",
       "      <td>424</td>\n",
       "      <td>16.635057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_919511_919935</th>\n",
       "      <th>CTCF_medium</th>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>919511</td>\n",
       "      <td>919935</td>\n",
       "      <td>chr1_919511_919935_medium_peak</td>\n",
       "      <td>2.369525e+00</td>\n",
       "      <td>.</td>\n",
       "      <td>18</td>\n",
       "      <td>28.781</td>\n",
       "      <td>424</td>\n",
       "      <td>1.475030</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_923010_923434</th>\n",
       "      <th>CTCF_medium</th>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>923010</td>\n",
       "      <td>923434</td>\n",
       "      <td>chr1_923010_923434_medium_peak</td>\n",
       "      <td>3.034356e+00</td>\n",
       "      <td>.</td>\n",
       "      <td>38</td>\n",
       "      <td>28.781</td>\n",
       "      <td>424</td>\n",
       "      <td>3.113952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_939801_940225</th>\n",
       "      <th>CTCF_medium</th>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>939801</td>\n",
       "      <td>940225</td>\n",
       "      <td>chr1_939801_940225_medium_peak</td>\n",
       "      <td>2.608183e+00</td>\n",
       "      <td>.</td>\n",
       "      <td>69</td>\n",
       "      <td>28.781</td>\n",
       "      <td>424</td>\n",
       "      <td>5.654281</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr1_951368_951792</th>\n",
       "      <th>CTCF_medium</th>\n",
       "      <td>GFP-NLS_REP-1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>951368</td>\n",
       "      <td>951792</td>\n",
       "      <td>chr1_951368_951792_medium_peak</td>\n",
       "      <td>2.659323e+00</td>\n",
       "      <td>.</td>\n",
       "      <td>41</td>\n",
       "      <td>28.781</td>\n",
       "      <td>424</td>\n",
       "      <td>3.359790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY_24811393_24815393</th>\n",
       "      <th>tss_no_ctcf</th>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>chrY</td>\n",
       "      <td>24811393</td>\n",
       "      <td>24815393</td>\n",
       "      <td>NM_020364</td>\n",
       "      <td>2.481339e+07</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>26.968</td>\n",
       "      <td>4000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY_24831819_24835819</th>\n",
       "      <th>tss_no_ctcf</th>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>chrY</td>\n",
       "      <td>24831819</td>\n",
       "      <td>24835819</td>\n",
       "      <td>NM_020420</td>\n",
       "      <td>2.483382e+07</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>26.968</td>\n",
       "      <td>4000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY_24831918_24835918</th>\n",
       "      <th>tss_no_ctcf</th>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>chrY</td>\n",
       "      <td>24831918</td>\n",
       "      <td>24835918</td>\n",
       "      <td>NM_001388484</td>\n",
       "      <td>2.483392e+07</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>26.968</td>\n",
       "      <td>4000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY_25050104_25054104</th>\n",
       "      <th>tss_no_ctcf</th>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>chrY</td>\n",
       "      <td>25050104</td>\n",
       "      <td>25054104</td>\n",
       "      <td>NM_001002761</td>\n",
       "      <td>2.505210e+07</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>26.968</td>\n",
       "      <td>4000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrY_25620116_25624116</th>\n",
       "      <th>tss_no_ctcf</th>\n",
       "      <td>SIRT6-DddA11_REP-2</td>\n",
       "      <td>chrY</td>\n",
       "      <td>25620116</td>\n",
       "      <td>25624116</td>\n",
       "      <td>NM_170723</td>\n",
       "      <td>2.562212e+07</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>26.968</td>\n",
       "      <td>4000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>568470 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                sample chrom     start   \n",
       "region_index           region                                            \n",
       "chr1_778863_779287     CTCF_medium       GFP-NLS_REP-1  chr1    778863  \\\n",
       "chr1_919511_919935     CTCF_medium       GFP-NLS_REP-1  chr1    919511   \n",
       "chr1_923010_923434     CTCF_medium       GFP-NLS_REP-1  chr1    923010   \n",
       "chr1_939801_940225     CTCF_medium       GFP-NLS_REP-1  chr1    939801   \n",
       "chr1_951368_951792     CTCF_medium       GFP-NLS_REP-1  chr1    951368   \n",
       "...                                                ...   ...       ...   \n",
       "chrY_24811393_24815393 tss_no_ctcf  SIRT6-DddA11_REP-2  chrY  24811393   \n",
       "chrY_24831819_24835819 tss_no_ctcf  SIRT6-DddA11_REP-2  chrY  24831819   \n",
       "chrY_24831918_24835918 tss_no_ctcf  SIRT6-DddA11_REP-2  chrY  24831918   \n",
       "chrY_25050104_25054104 tss_no_ctcf  SIRT6-DddA11_REP-2  chrY  25050104   \n",
       "chrY_25620116_25624116 tss_no_ctcf  SIRT6-DddA11_REP-2  chrY  25620116   \n",
       "\n",
       "                                         end                            name   \n",
       "region_index           region                                                  \n",
       "chr1_778863_779287     CTCF_medium    779287  chr1_778863_779287_medium_peak  \\\n",
       "chr1_919511_919935     CTCF_medium    919935  chr1_919511_919935_medium_peak   \n",
       "chr1_923010_923434     CTCF_medium    923434  chr1_923010_923434_medium_peak   \n",
       "chr1_939801_940225     CTCF_medium    940225  chr1_939801_940225_medium_peak   \n",
       "chr1_951368_951792     CTCF_medium    951792  chr1_951368_951792_medium_peak   \n",
       "...                                      ...                             ...   \n",
       "chrY_24811393_24815393 tss_no_ctcf  24815393                       NM_020364   \n",
       "chrY_24831819_24835819 tss_no_ctcf  24835819                       NM_020420   \n",
       "chrY_24831918_24835918 tss_no_ctcf  24835918                    NM_001388484   \n",
       "chrY_25050104_25054104 tss_no_ctcf  25054104                    NM_001002761   \n",
       "chrY_25620116_25624116 tss_no_ctcf  25624116                       NM_170723   \n",
       "\n",
       "                                           score strand  depth  total_reads_M   \n",
       "region_index           region                                                   \n",
       "chr1_778863_779287     CTCF_medium  2.778652e+00      .    203         28.781  \\\n",
       "chr1_919511_919935     CTCF_medium  2.369525e+00      .     18         28.781   \n",
       "chr1_923010_923434     CTCF_medium  3.034356e+00      .     38         28.781   \n",
       "chr1_939801_940225     CTCF_medium  2.608183e+00      .     69         28.781   \n",
       "chr1_951368_951792     CTCF_medium  2.659323e+00      .     41         28.781   \n",
       "...                                          ...    ...    ...            ...   \n",
       "chrY_24811393_24815393 tss_no_ctcf  2.481339e+07      -      0         26.968   \n",
       "chrY_24831819_24835819 tss_no_ctcf  2.483382e+07      +      0         26.968   \n",
       "chrY_24831918_24835918 tss_no_ctcf  2.483392e+07      +      0         26.968   \n",
       "chrY_25050104_25054104 tss_no_ctcf  2.505210e+07      -      0         26.968   \n",
       "chrY_25620116_25624116 tss_no_ctcf  2.562212e+07      +      0         26.968   \n",
       "\n",
       "                                    region_length       RPKM  \n",
       "region_index           region                                 \n",
       "chr1_778863_779287     CTCF_medium            424  16.635057  \n",
       "chr1_919511_919935     CTCF_medium            424   1.475030  \n",
       "chr1_923010_923434     CTCF_medium            424   3.113952  \n",
       "chr1_939801_940225     CTCF_medium            424   5.654281  \n",
       "chr1_951368_951792     CTCF_medium            424   3.359790  \n",
       "...                                           ...        ...  \n",
       "chrY_24811393_24815393 tss_no_ctcf           4000   0.000000  \n",
       "chrY_24831819_24835819 tss_no_ctcf           4000   0.000000  \n",
       "chrY_24831918_24835918 tss_no_ctcf           4000   0.000000  \n",
       "chrY_25050104_25054104 tss_no_ctcf           4000   0.000000  \n",
       "chrY_25620116_25624116 tss_no_ctcf           4000   0.000000  \n",
       "\n",
       "[568470 rows x 11 columns]"
      ]
     },
     "execution_count": 82,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_counts['region_length'] = df_counts['end'] -  df_counts['start']\n",
    "df_counts['RPKM'] = df_counts['depth'] / (df_counts['region_length']/1_000 * df_counts['total_reads_M'])\n",
    "df_counts['region_index'] = df_counts['chrom'] + '_' + df_counts['start'].astype(str) + '_' + df_counts['end'].astype(str)\n",
    "df_counts.drop_duplicates(subset=['sample', 'region', 'region_index'], keep='first', inplace=True)\n",
    "df_counts.set_index(['region_index', 'region'], inplace=True)\n",
    "df_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "580527c2-05be-412b-a44a-b3d4f3321101",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:02:59.015891Z",
     "iopub.status.busy": "2023-05-22T08:02:59.015254Z",
     "iopub.status.idle": "2023-05-22T08:02:59.069598Z",
     "shell.execute_reply": "2023-05-22T08:02:59.068395Z",
     "shell.execute_reply.started": "2023-05-22T08:02:59.015832Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    568470.000000\n",
       "mean          2.950206\n",
       "std           6.291637\n",
       "min           0.000000\n",
       "25%           0.401204\n",
       "50%           1.234090\n",
       "75%           3.217199\n",
       "max         387.884121\n",
       "Name: RPKM, dtype: float64"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_counts.RPKM.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "2b812648-dd5c-4457-b9ca-3ac6bcbb66ca",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:03:14.883616Z",
     "iopub.status.busy": "2023-05-22T08:03:14.882952Z",
     "iopub.status.idle": "2023-05-22T08:03:14.894981Z",
     "shell.execute_reply": "2023-05-22T08:03:14.893276Z",
     "shell.execute_reply.started": "2023-05-22T08:03:14.883562Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "            <div id=\"lp97jb\"></div>\n",
       "            <script type=\"text/javascript\" data-lets-plot-script=\"library\">\n",
       "                if(!window.letsPlotCallQueue) {\n",
       "                    window.letsPlotCallQueue = [];\n",
       "                }; \n",
       "                window.letsPlotCall = function(f) {\n",
       "                    window.letsPlotCallQueue.push(f);\n",
       "                };\n",
       "                (function() {\n",
       "                    var script = document.createElement(\"script\");\n",
       "                    script.type = \"text/javascript\";\n",
       "                    script.src = \"https://cdn.jsdelivr.net/gh/JetBrains/lets-plot@v3.2.0/js-package/distr/lets-plot.min.js\";\n",
       "                    script.onload = function() {\n",
       "                        window.letsPlotCall = function(f) {f();};\n",
       "                        window.letsPlotCallQueue.forEach(function(f) {f();});\n",
       "                        window.letsPlotCallQueue = [];\n",
       "                        \n",
       "                    };\n",
       "                    script.onerror = function(event) {\n",
       "                        window.letsPlotCall = function(f) {};    // noop\n",
       "                        window.letsPlotCallQueue = [];\n",
       "                        var div = document.createElement(\"div\");\n",
       "                        div.style.color = 'darkred';\n",
       "                        div.textContent = 'Error loading Lets-Plot JS';\n",
       "                        document.getElementById(\"lp97jb\").appendChild(div);\n",
       "                    };\n",
       "                    var e = document.getElementById(\"lp97jb\");\n",
       "                    e.appendChild(script);\n",
       "                })()\n",
       "            </script>\n",
       "            "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import itertools as it\n",
    "import math\n",
    "from lets_plot import *\n",
    "LetsPlot.setup_html()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "bca0a5d4-ca64-478c-8307-8754f540526f",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:09:01.668274Z",
     "iopub.status.busy": "2023-05-22T08:09:01.667625Z",
     "iopub.status.idle": "2023-05-22T08:09:04.041999Z",
     "shell.execute_reply": "2023-05-22T08:09:04.041327Z",
     "shell.execute_reply.started": "2023-05-22T08:09:01.668216Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>region_index</th>\n",
       "      <th>region</th>\n",
       "      <th>Log2_FC_RPKM</th>\n",
       "      <th>Geometric_mean</th>\n",
       "      <th>compare</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_778863_779287</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>0.137815</td>\n",
       "      <td>14.194386</td>\n",
       "      <td>GFP-NLS_REP-1 v.s. GFP_REP-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_919511_919935</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>-0.284319</td>\n",
       "      <td>2.046259</td>\n",
       "      <td>GFP-NLS_REP-1 v.s. GFP_REP-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_923010_923434</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>0.025951</td>\n",
       "      <td>3.022291</td>\n",
       "      <td>GFP-NLS_REP-1 v.s. GFP_REP-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_939801_940225</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>0.060375</td>\n",
       "      <td>5.274604</td>\n",
       "      <td>GFP-NLS_REP-1 v.s. GFP_REP-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_951368_951792</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>0.103155</td>\n",
       "      <td>2.983559</td>\n",
       "      <td>GFP-NLS_REP-1 v.s. GFP_REP-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60950</th>\n",
       "      <td>chrX_155378708_155382708</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>-0.075912</td>\n",
       "      <td>1.274168</td>\n",
       "      <td>SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60951</th>\n",
       "      <td>chrX_155380094_155384094</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>-0.090775</td>\n",
       "      <td>1.653367</td>\n",
       "      <td>SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60952</th>\n",
       "      <td>chrX_155456620_155460620</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>-0.036049</td>\n",
       "      <td>1.236179</td>\n",
       "      <td>SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60953</th>\n",
       "      <td>chrX_155458005_155462005</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>-0.060447</td>\n",
       "      <td>0.916587</td>\n",
       "      <td>SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60954</th>\n",
       "      <td>chrX_155765795_155769795</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>-0.072506</td>\n",
       "      <td>0.639587</td>\n",
       "      <td>SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>896978 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   region_index       region  Log2_FC_RPKM  Geometric_mean   \n",
       "0            chr1_778863_779287  CTCF_medium      0.137815       14.194386  \\\n",
       "1            chr1_919511_919935  CTCF_medium     -0.284319        2.046259   \n",
       "2            chr1_923010_923434  CTCF_medium      0.025951        3.022291   \n",
       "3            chr1_939801_940225  CTCF_medium      0.060375        5.274604   \n",
       "4            chr1_951368_951792  CTCF_medium      0.103155        2.983559   \n",
       "...                         ...          ...           ...             ...   \n",
       "60950  chrX_155378708_155382708  tss_no_ctcf     -0.075912        1.274168   \n",
       "60951  chrX_155380094_155384094  tss_no_ctcf     -0.090775        1.653367   \n",
       "60952  chrX_155456620_155460620  tss_no_ctcf     -0.036049        1.236179   \n",
       "60953  chrX_155458005_155462005  tss_no_ctcf     -0.060447        0.916587   \n",
       "60954  chrX_155765795_155769795  tss_no_ctcf     -0.072506        0.639587   \n",
       "\n",
       "                                          compare  \n",
       "0                    GFP-NLS_REP-1 v.s. GFP_REP-1  \n",
       "1                    GFP-NLS_REP-1 v.s. GFP_REP-1  \n",
       "2                    GFP-NLS_REP-1 v.s. GFP_REP-1  \n",
       "3                    GFP-NLS_REP-1 v.s. GFP_REP-1  \n",
       "4                    GFP-NLS_REP-1 v.s. GFP_REP-1  \n",
       "...                                           ...  \n",
       "60950  SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2  \n",
       "60951  SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2  \n",
       "60952  SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2  \n",
       "60953  SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2  \n",
       "60954  SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2  \n",
       "\n",
       "[896978 rows x 5 columns]"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ls = []\n",
    "for sample_2 in it.combinations(samples, 2):\n",
    "    s1, s2 = sample_2\n",
    "    s_s1 = df_counts.query('sample==@s1')['RPKM']\n",
    "    s_s2 = df_counts.query('sample==@s2')['RPKM']\n",
    "    df_sample_2 = pd.concat([s_s1, s_s2], axis=1)\n",
    "    df_sample_2.columns = ['RPKM1', 'RPKM2']\n",
    "    df_sample_2 = df_sample_2[(df_sample_2.RPKM1 >= 0.5) & (df_sample_2.RPKM2 >= 0.5)].copy()\n",
    "    df_sample_2['Log2_FC_RPKM'] = (df_sample_2['RPKM1'] / df_sample_2['RPKM2']).map(math.log10)\n",
    "    df_sample_2['Geometric_mean'] = df_sample_2['RPKM1'].map(math.sqrt) * df_sample_2['RPKM2'].map(math.sqrt)\n",
    "    df_sample_2 = df_sample_2[['Log2_FC_RPKM', 'Geometric_mean']].reset_index().copy()\n",
    "    df_sample_2['compare'] = f'{s1} v.s. {s2}'\n",
    "    ls.append(df_sample_2)\n",
    "df_plot = pd.concat(ls)\n",
    "df_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "23a06617-d9bd-4fbf-8e47-3b974203713b",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:09:08.037711Z",
     "iopub.status.busy": "2023-05-22T08:09:08.037092Z",
     "iopub.status.idle": "2023-05-22T08:09:08.308584Z",
     "shell.execute_reply": "2023-05-22T08:09:08.307615Z",
     "shell.execute_reply.started": "2023-05-22T08:09:08.037653Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                        compare            region  region_index\n",
      "0                  GFP-NLS_REP-1 v.s. GFP_REP-1       CTCF_medium         11463\n",
      "1                  GFP-NLS_REP-1 v.s. GFP_REP-1       CTCF_strong         12340\n",
      "2                  GFP-NLS_REP-1 v.s. GFP_REP-1         CTCF_weak          9318\n",
      "3                  GFP-NLS_REP-1 v.s. GFP_REP-1  CTCF_within-offs          1218\n",
      "4                  GFP-NLS_REP-1 v.s. GFP_REP-1      offs-ND6-DEP            66\n",
      "5                  GFP-NLS_REP-1 v.s. GFP_REP-1      offs-ND6-IND           526\n",
      "6                  GFP-NLS_REP-1 v.s. GFP_REP-1            random          5182\n",
      "7                  GFP-NLS_REP-1 v.s. GFP_REP-1      tss_has_ctcf         10617\n",
      "8                  GFP-NLS_REP-1 v.s. GFP_REP-1       tss_no_ctcf          8051\n",
      "9           GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-1       CTCF_medium         11459\n",
      "10          GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-1       CTCF_strong         12337\n",
      "11          GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-1         CTCF_weak          9518\n",
      "12          GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-1  CTCF_within-offs          1223\n",
      "13          GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-1      offs-ND6-DEP            67\n",
      "14          GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-1      offs-ND6-IND           530\n",
      "15          GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-1            random          5286\n",
      "16          GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-1      tss_has_ctcf         10631\n",
      "17          GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-1       tss_no_ctcf          8135\n",
      "18          GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-2       CTCF_medium         11508\n",
      "19          GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-2       CTCF_strong         12360\n",
      "20          GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-2         CTCF_weak          9630\n",
      "21          GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-2  CTCF_within-offs          1224\n",
      "22          GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-2      offs-ND6-DEP            76\n",
      "23          GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-2      offs-ND6-IND           530\n",
      "24          GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-2            random          5656\n",
      "25          GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-2      tss_has_ctcf         10636\n",
      "26          GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-2       tss_no_ctcf          8252\n",
      "27        GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-1       CTCF_medium         11570\n",
      "28        GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-1       CTCF_strong         12368\n",
      "29        GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-1         CTCF_weak          9736\n",
      "30        GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-1  CTCF_within-offs          1226\n",
      "31        GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-1      offs-ND6-DEP            63\n",
      "32        GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-1      offs-ND6-IND           528\n",
      "33        GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-1            random          5762\n",
      "34        GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-1      tss_has_ctcf         10640\n",
      "35        GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-1       tss_no_ctcf          8189\n",
      "36        GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-2       CTCF_medium         11545\n",
      "37        GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-2       CTCF_strong         12363\n",
      "38        GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-2         CTCF_weak          9702\n",
      "39        GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-2  CTCF_within-offs          1228\n",
      "40        GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-2      offs-ND6-DEP            68\n",
      "41        GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-2      offs-ND6-IND           530\n",
      "42        GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-2            random          5621\n",
      "43        GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-2      tss_has_ctcf         10632\n",
      "44        GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-2       tss_no_ctcf          8130\n",
      "45              GFP_REP-1 v.s. ND6-DddAwt_REP-1       CTCF_medium         11414\n",
      "46              GFP_REP-1 v.s. ND6-DddAwt_REP-1       CTCF_strong         12327\n",
      "47              GFP_REP-1 v.s. ND6-DddAwt_REP-1         CTCF_weak          9304\n",
      "48              GFP_REP-1 v.s. ND6-DddAwt_REP-1  CTCF_within-offs          1215\n",
      "49              GFP_REP-1 v.s. ND6-DddAwt_REP-1      offs-ND6-DEP            61\n",
      "50              GFP_REP-1 v.s. ND6-DddAwt_REP-1      offs-ND6-IND           522\n",
      "51              GFP_REP-1 v.s. ND6-DddAwt_REP-1            random          5184\n",
      "52              GFP_REP-1 v.s. ND6-DddAwt_REP-1      tss_has_ctcf         10589\n",
      "53              GFP_REP-1 v.s. ND6-DddAwt_REP-1       tss_no_ctcf          8082\n",
      "54              GFP_REP-1 v.s. ND6-DddAwt_REP-2       CTCF_medium         11451\n",
      "55              GFP_REP-1 v.s. ND6-DddAwt_REP-2       CTCF_strong         12340\n",
      "56              GFP_REP-1 v.s. ND6-DddAwt_REP-2         CTCF_weak          9379\n",
      "57              GFP_REP-1 v.s. ND6-DddAwt_REP-2  CTCF_within-offs          1213\n",
      "58              GFP_REP-1 v.s. ND6-DddAwt_REP-2      offs-ND6-DEP            73\n",
      "59              GFP_REP-1 v.s. ND6-DddAwt_REP-2      offs-ND6-IND           523\n",
      "60              GFP_REP-1 v.s. ND6-DddAwt_REP-2            random          5541\n",
      "61              GFP_REP-1 v.s. ND6-DddAwt_REP-2      tss_has_ctcf         10601\n",
      "62              GFP_REP-1 v.s. ND6-DddAwt_REP-2       tss_no_ctcf          8155\n",
      "63            GFP_REP-1 v.s. SIRT6-DddA11_REP-1       CTCF_medium         11504\n",
      "64            GFP_REP-1 v.s. SIRT6-DddA11_REP-1       CTCF_strong         12350\n",
      "65            GFP_REP-1 v.s. SIRT6-DddA11_REP-1         CTCF_weak          9524\n",
      "66            GFP_REP-1 v.s. SIRT6-DddA11_REP-1  CTCF_within-offs          1217\n",
      "67            GFP_REP-1 v.s. SIRT6-DddA11_REP-1      offs-ND6-DEP            61\n",
      "68            GFP_REP-1 v.s. SIRT6-DddA11_REP-1      offs-ND6-IND           521\n",
      "69            GFP_REP-1 v.s. SIRT6-DddA11_REP-1            random          5660\n",
      "70            GFP_REP-1 v.s. SIRT6-DddA11_REP-1      tss_has_ctcf         10602\n",
      "71            GFP_REP-1 v.s. SIRT6-DddA11_REP-1       tss_no_ctcf          8122\n",
      "72            GFP_REP-1 v.s. SIRT6-DddA11_REP-2       CTCF_medium         11488\n",
      "73            GFP_REP-1 v.s. SIRT6-DddA11_REP-2       CTCF_strong         12342\n",
      "74            GFP_REP-1 v.s. SIRT6-DddA11_REP-2         CTCF_weak          9470\n",
      "75            GFP_REP-1 v.s. SIRT6-DddA11_REP-2  CTCF_within-offs          1219\n",
      "76            GFP_REP-1 v.s. SIRT6-DddA11_REP-2      offs-ND6-DEP            65\n",
      "77            GFP_REP-1 v.s. SIRT6-DddA11_REP-2      offs-ND6-IND           523\n",
      "78            GFP_REP-1 v.s. SIRT6-DddA11_REP-2            random          5567\n",
      "79            GFP_REP-1 v.s. SIRT6-DddA11_REP-2      tss_has_ctcf         10584\n",
      "80            GFP_REP-1 v.s. SIRT6-DddA11_REP-2       tss_no_ctcf          8090\n",
      "81       ND6-DddAwt_REP-1 v.s. ND6-DddAwt_REP-2       CTCF_medium         11468\n",
      "82       ND6-DddAwt_REP-1 v.s. ND6-DddAwt_REP-2       CTCF_strong         12336\n",
      "83       ND6-DddAwt_REP-1 v.s. ND6-DddAwt_REP-2         CTCF_weak          9671\n",
      "84       ND6-DddAwt_REP-1 v.s. ND6-DddAwt_REP-2  CTCF_within-offs          1217\n",
      "85       ND6-DddAwt_REP-1 v.s. ND6-DddAwt_REP-2      offs-ND6-DEP            74\n",
      "86       ND6-DddAwt_REP-1 v.s. ND6-DddAwt_REP-2      offs-ND6-IND           528\n",
      "87       ND6-DddAwt_REP-1 v.s. ND6-DddAwt_REP-2            random          5694\n",
      "88       ND6-DddAwt_REP-1 v.s. ND6-DddAwt_REP-2      tss_has_ctcf         10613\n",
      "89       ND6-DddAwt_REP-1 v.s. ND6-DddAwt_REP-2       tss_no_ctcf          8298\n",
      "90     ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-1       CTCF_medium         11522\n",
      "91     ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-1       CTCF_strong         12347\n",
      "92     ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-1         CTCF_weak          9818\n",
      "93     ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-1  CTCF_within-offs          1221\n",
      "94     ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-1      offs-ND6-DEP            65\n",
      "95     ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-1      offs-ND6-IND           526\n",
      "96     ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-1            random          5781\n",
      "97     ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-1      tss_has_ctcf         10615\n",
      "98     ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-1       tss_no_ctcf          8244\n",
      "99     ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-2       CTCF_medium         11505\n",
      "100    ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-2       CTCF_strong         12344\n",
      "101    ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-2         CTCF_weak          9753\n",
      "102    ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-2  CTCF_within-offs          1223\n",
      "103    ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-2      offs-ND6-DEP            65\n",
      "104    ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-2      offs-ND6-IND           526\n",
      "105    ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-2            random          5703\n",
      "106    ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-2      tss_has_ctcf         10606\n",
      "107    ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-2       tss_no_ctcf          8242\n",
      "108    ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-1       CTCF_medium         11570\n",
      "109    ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-1       CTCF_strong         12380\n",
      "110    ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-1         CTCF_weak          9888\n",
      "111    ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-1  CTCF_within-offs          1224\n",
      "112    ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-1      offs-ND6-DEP            74\n",
      "113    ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-1      offs-ND6-IND           523\n",
      "114    ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-1            random          6235\n",
      "115    ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-1      tss_has_ctcf         10631\n",
      "116    ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-1       tss_no_ctcf          8334\n",
      "117    ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-2       CTCF_medium         11538\n",
      "118    ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-2       CTCF_strong         12369\n",
      "119    ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-2         CTCF_weak          9824\n",
      "120    ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-2  CTCF_within-offs          1223\n",
      "121    ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-2      offs-ND6-DEP            80\n",
      "122    ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-2      offs-ND6-IND           526\n",
      "123    ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-2            random          6064\n",
      "124    ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-2      tss_has_ctcf         10609\n",
      "125    ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-2       tss_no_ctcf          8303\n",
      "126  SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2       CTCF_medium         11603\n",
      "127  SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2       CTCF_strong         12374\n",
      "128  SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2         CTCF_weak          9996\n",
      "129  SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2  CTCF_within-offs          1228\n",
      "130  SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2      offs-ND6-DEP            62\n",
      "131  SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2      offs-ND6-IND           524\n",
      "132  SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2            random          6259\n",
      "133  SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2      tss_has_ctcf         10612\n",
      "134  SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2       tss_no_ctcf          8297\n"
     ]
    }
   ],
   "source": [
    "print(df_plot.groupby(['compare', 'region']).region_index.count().to_frame().reset_index().to_string())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "ba750b23-3b2d-41b9-a73d-610badf47efc",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T07:51:13.583498Z",
     "iopub.status.busy": "2023-05-22T07:51:13.582776Z",
     "iopub.status.idle": "2023-05-22T07:51:13.819537Z",
     "shell.execute_reply": "2023-05-22T07:51:13.818700Z",
     "shell.execute_reply.started": "2023-05-22T07:51:13.583440Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>compare</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GFP-NLS_REP-1 v.s. GFP_REP-1</th>\n",
       "      <td>58781.0</td>\n",
       "      <td>0.109124</td>\n",
       "      <td>0.161037</td>\n",
       "      <td>-0.965561</td>\n",
       "      <td>0.028449</td>\n",
       "      <td>0.113621</td>\n",
       "      <td>0.197723</td>\n",
       "      <td>1.051473</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-1</th>\n",
       "      <td>59186.0</td>\n",
       "      <td>0.147606</td>\n",
       "      <td>0.189976</td>\n",
       "      <td>-0.820159</td>\n",
       "      <td>0.042313</td>\n",
       "      <td>0.167252</td>\n",
       "      <td>0.271092</td>\n",
       "      <td>1.188441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-2</th>\n",
       "      <td>59872.0</td>\n",
       "      <td>0.188708</td>\n",
       "      <td>0.187448</td>\n",
       "      <td>-0.947370</td>\n",
       "      <td>0.086054</td>\n",
       "      <td>0.213998</td>\n",
       "      <td>0.309306</td>\n",
       "      <td>1.228722</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-1</th>\n",
       "      <td>60082.0</td>\n",
       "      <td>0.172706</td>\n",
       "      <td>0.197298</td>\n",
       "      <td>-0.869731</td>\n",
       "      <td>0.061848</td>\n",
       "      <td>0.199763</td>\n",
       "      <td>0.302726</td>\n",
       "      <td>1.168492</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-2</th>\n",
       "      <td>59819.0</td>\n",
       "      <td>0.175846</td>\n",
       "      <td>0.197334</td>\n",
       "      <td>-0.873355</td>\n",
       "      <td>0.066299</td>\n",
       "      <td>0.200222</td>\n",
       "      <td>0.305820</td>\n",
       "      <td>1.175863</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GFP_REP-1 v.s. ND6-DddAwt_REP-1</th>\n",
       "      <td>58698.0</td>\n",
       "      <td>0.040026</td>\n",
       "      <td>0.169678</td>\n",
       "      <td>-1.045753</td>\n",
       "      <td>-0.049111</td>\n",
       "      <td>0.058691</td>\n",
       "      <td>0.142573</td>\n",
       "      <td>0.898730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GFP_REP-1 v.s. ND6-DddAwt_REP-2</th>\n",
       "      <td>59276.0</td>\n",
       "      <td>0.082393</td>\n",
       "      <td>0.164931</td>\n",
       "      <td>-0.884899</td>\n",
       "      <td>-0.009838</td>\n",
       "      <td>0.103202</td>\n",
       "      <td>0.184548</td>\n",
       "      <td>0.939010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GFP_REP-1 v.s. SIRT6-DddA11_REP-1</th>\n",
       "      <td>59561.0</td>\n",
       "      <td>0.066935</td>\n",
       "      <td>0.176651</td>\n",
       "      <td>-0.916405</td>\n",
       "      <td>-0.032200</td>\n",
       "      <td>0.089425</td>\n",
       "      <td>0.179590</td>\n",
       "      <td>1.268287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GFP_REP-1 v.s. SIRT6-DddA11_REP-2</th>\n",
       "      <td>59348.0</td>\n",
       "      <td>0.069462</td>\n",
       "      <td>0.176351</td>\n",
       "      <td>-1.141878</td>\n",
       "      <td>-0.027934</td>\n",
       "      <td>0.092206</td>\n",
       "      <td>0.180342</td>\n",
       "      <td>1.210305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ND6-DddAwt_REP-1 v.s. ND6-DddAwt_REP-2</th>\n",
       "      <td>59899.0</td>\n",
       "      <td>0.043078</td>\n",
       "      <td>0.146651</td>\n",
       "      <td>-0.765145</td>\n",
       "      <td>-0.030301</td>\n",
       "      <td>0.044091</td>\n",
       "      <td>0.119462</td>\n",
       "      <td>0.932375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-1</th>\n",
       "      <td>60139.0</td>\n",
       "      <td>0.027299</td>\n",
       "      <td>0.155082</td>\n",
       "      <td>-0.860892</td>\n",
       "      <td>-0.052715</td>\n",
       "      <td>0.039129</td>\n",
       "      <td>0.112631</td>\n",
       "      <td>1.218290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-2</th>\n",
       "      <td>59967.0</td>\n",
       "      <td>0.030473</td>\n",
       "      <td>0.153192</td>\n",
       "      <td>-0.877880</td>\n",
       "      <td>-0.046856</td>\n",
       "      <td>0.038574</td>\n",
       "      <td>0.112600</td>\n",
       "      <td>1.184702</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-1</th>\n",
       "      <td>60859.0</td>\n",
       "      <td>-0.015375</td>\n",
       "      <td>0.149890</td>\n",
       "      <td>-1.006682</td>\n",
       "      <td>-0.088450</td>\n",
       "      <td>-0.008806</td>\n",
       "      <td>0.064066</td>\n",
       "      <td>1.001918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-2</th>\n",
       "      <td>60536.0</td>\n",
       "      <td>-0.012404</td>\n",
       "      <td>0.149278</td>\n",
       "      <td>-0.925986</td>\n",
       "      <td>-0.080888</td>\n",
       "      <td>-0.006720</td>\n",
       "      <td>0.065240</td>\n",
       "      <td>0.998294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2</th>\n",
       "      <td>60955.0</td>\n",
       "      <td>0.002762</td>\n",
       "      <td>0.150238</td>\n",
       "      <td>-1.282378</td>\n",
       "      <td>-0.073260</td>\n",
       "      <td>-0.003624</td>\n",
       "      <td>0.078911</td>\n",
       "      <td>1.142504</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              count      mean       std   \n",
       "compare                                                                   \n",
       "GFP-NLS_REP-1 v.s. GFP_REP-1                58781.0  0.109124  0.161037  \\\n",
       "GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-1         59186.0  0.147606  0.189976   \n",
       "GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-2         59872.0  0.188708  0.187448   \n",
       "GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-1       60082.0  0.172706  0.197298   \n",
       "GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-2       59819.0  0.175846  0.197334   \n",
       "GFP_REP-1 v.s. ND6-DddAwt_REP-1             58698.0  0.040026  0.169678   \n",
       "GFP_REP-1 v.s. ND6-DddAwt_REP-2             59276.0  0.082393  0.164931   \n",
       "GFP_REP-1 v.s. SIRT6-DddA11_REP-1           59561.0  0.066935  0.176651   \n",
       "GFP_REP-1 v.s. SIRT6-DddA11_REP-2           59348.0  0.069462  0.176351   \n",
       "ND6-DddAwt_REP-1 v.s. ND6-DddAwt_REP-2      59899.0  0.043078  0.146651   \n",
       "ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-1    60139.0  0.027299  0.155082   \n",
       "ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-2    59967.0  0.030473  0.153192   \n",
       "ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-1    60859.0 -0.015375  0.149890   \n",
       "ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-2    60536.0 -0.012404  0.149278   \n",
       "SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2  60955.0  0.002762  0.150238   \n",
       "\n",
       "                                                 min       25%       50%   \n",
       "compare                                                                    \n",
       "GFP-NLS_REP-1 v.s. GFP_REP-1               -0.965561  0.028449  0.113621  \\\n",
       "GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-1        -0.820159  0.042313  0.167252   \n",
       "GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-2        -0.947370  0.086054  0.213998   \n",
       "GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-1      -0.869731  0.061848  0.199763   \n",
       "GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-2      -0.873355  0.066299  0.200222   \n",
       "GFP_REP-1 v.s. ND6-DddAwt_REP-1            -1.045753 -0.049111  0.058691   \n",
       "GFP_REP-1 v.s. ND6-DddAwt_REP-2            -0.884899 -0.009838  0.103202   \n",
       "GFP_REP-1 v.s. SIRT6-DddA11_REP-1          -0.916405 -0.032200  0.089425   \n",
       "GFP_REP-1 v.s. SIRT6-DddA11_REP-2          -1.141878 -0.027934  0.092206   \n",
       "ND6-DddAwt_REP-1 v.s. ND6-DddAwt_REP-2     -0.765145 -0.030301  0.044091   \n",
       "ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-1   -0.860892 -0.052715  0.039129   \n",
       "ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-2   -0.877880 -0.046856  0.038574   \n",
       "ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-1   -1.006682 -0.088450 -0.008806   \n",
       "ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-2   -0.925986 -0.080888 -0.006720   \n",
       "SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2 -1.282378 -0.073260 -0.003624   \n",
       "\n",
       "                                                 75%       max  \n",
       "compare                                                         \n",
       "GFP-NLS_REP-1 v.s. GFP_REP-1                0.197723  1.051473  \n",
       "GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-1         0.271092  1.188441  \n",
       "GFP-NLS_REP-1 v.s. ND6-DddAwt_REP-2         0.309306  1.228722  \n",
       "GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-1       0.302726  1.168492  \n",
       "GFP-NLS_REP-1 v.s. SIRT6-DddA11_REP-2       0.305820  1.175863  \n",
       "GFP_REP-1 v.s. ND6-DddAwt_REP-1             0.142573  0.898730  \n",
       "GFP_REP-1 v.s. ND6-DddAwt_REP-2             0.184548  0.939010  \n",
       "GFP_REP-1 v.s. SIRT6-DddA11_REP-1           0.179590  1.268287  \n",
       "GFP_REP-1 v.s. SIRT6-DddA11_REP-2           0.180342  1.210305  \n",
       "ND6-DddAwt_REP-1 v.s. ND6-DddAwt_REP-2      0.119462  0.932375  \n",
       "ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-1    0.112631  1.218290  \n",
       "ND6-DddAwt_REP-1 v.s. SIRT6-DddA11_REP-2    0.112600  1.184702  \n",
       "ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-1    0.064066  1.001918  \n",
       "ND6-DddAwt_REP-2 v.s. SIRT6-DddA11_REP-2    0.065240  0.998294  \n",
       "SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2  0.078911  1.142504  "
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_plot.groupby('compare').Log2_FC_RPKM.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "6ec7839e-ceb3-46d6-825e-c9979b6fb67a",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-22T08:09:21.285831Z",
     "iopub.status.busy": "2023-05-22T08:09:21.285208Z",
     "iopub.status.idle": "2023-05-22T08:09:21.311265Z",
     "shell.execute_reply": "2023-05-22T08:09:21.309965Z",
     "shell.execute_reply.started": "2023-05-22T08:09:21.285775Z"
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>region_index</th>\n",
       "      <th>region</th>\n",
       "      <th>Log2_FC_RPKM</th>\n",
       "      <th>Geometric_mean</th>\n",
       "      <th>compare</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1_778863_779287</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>0.137815</td>\n",
       "      <td>14.194386</td>\n",
       "      <td>GFP-NLS_REP-1 v.s. GFP_REP-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1_919511_919935</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>-0.284319</td>\n",
       "      <td>2.046259</td>\n",
       "      <td>GFP-NLS_REP-1 v.s. GFP_REP-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1_923010_923434</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>0.025951</td>\n",
       "      <td>3.022291</td>\n",
       "      <td>GFP-NLS_REP-1 v.s. GFP_REP-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1_939801_940225</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>0.060375</td>\n",
       "      <td>5.274604</td>\n",
       "      <td>GFP-NLS_REP-1 v.s. GFP_REP-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1_951368_951792</td>\n",
       "      <td>CTCF_medium</td>\n",
       "      <td>0.103155</td>\n",
       "      <td>2.983559</td>\n",
       "      <td>GFP-NLS_REP-1 v.s. GFP_REP-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60950</th>\n",
       "      <td>chrX_155378708_155382708</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>-0.075912</td>\n",
       "      <td>1.274168</td>\n",
       "      <td>SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60951</th>\n",
       "      <td>chrX_155380094_155384094</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>-0.090775</td>\n",
       "      <td>1.653367</td>\n",
       "      <td>SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60952</th>\n",
       "      <td>chrX_155456620_155460620</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>-0.036049</td>\n",
       "      <td>1.236179</td>\n",
       "      <td>SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60953</th>\n",
       "      <td>chrX_155458005_155462005</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>-0.060447</td>\n",
       "      <td>0.916587</td>\n",
       "      <td>SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60954</th>\n",
       "      <td>chrX_155765795_155769795</td>\n",
       "      <td>tss_no_ctcf</td>\n",
       "      <td>-0.072506</td>\n",
       "      <td>0.639587</td>\n",
       "      <td>SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>896978 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   region_index       region  Log2_FC_RPKM  Geometric_mean   \n",
       "0            chr1_778863_779287  CTCF_medium      0.137815       14.194386  \\\n",
       "1            chr1_919511_919935  CTCF_medium     -0.284319        2.046259   \n",
       "2            chr1_923010_923434  CTCF_medium      0.025951        3.022291   \n",
       "3            chr1_939801_940225  CTCF_medium      0.060375        5.274604   \n",
       "4            chr1_951368_951792  CTCF_medium      0.103155        2.983559   \n",
       "...                         ...          ...           ...             ...   \n",
       "60950  chrX_155378708_155382708  tss_no_ctcf     -0.075912        1.274168   \n",
       "60951  chrX_155380094_155384094  tss_no_ctcf     -0.090775        1.653367   \n",
       "60952  chrX_155456620_155460620  tss_no_ctcf     -0.036049        1.236179   \n",
       "60953  chrX_155458005_155462005  tss_no_ctcf     -0.060447        0.916587   \n",
       "60954  chrX_155765795_155769795  tss_no_ctcf     -0.072506        0.639587   \n",
       "\n",
       "                                          compare  \n",
       "0                    GFP-NLS_REP-1 v.s. GFP_REP-1  \n",
       "1                    GFP-NLS_REP-1 v.s. GFP_REP-1  \n",
       "2                    GFP-NLS_REP-1 v.s. GFP_REP-1  \n",
       "3                    GFP-NLS_REP-1 v.s. GFP_REP-1  \n",
       "4                    GFP-NLS_REP-1 v.s. GFP_REP-1  \n",
       "...                                           ...  \n",
       "60950  SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2  \n",
       "60951  SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2  \n",
       "60952  SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2  \n",
       "60953  SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2  \n",
       "60954  SIRT6-DddA11_REP-1 v.s. SIRT6-DddA11_REP-2  \n",
       "\n",
       "[896978 rows x 5 columns]"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d870d9e4-c34f-481d-bb2f-a73a76f7b8fa",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "g = (\n",
    "    ggplot(data=df_plot, mapping=aes(x='Log2_FC_RPKM', y='Geometric_mean'))\n",
    "    + geom_point()\n",
    "    + geom_line(x=0, color='red')\n",
    "    + facet_grid(x='compare', y='region', scales='free')\n",
    "    + ggsize(width=5500, height=1500)\n",
    "    + theme_minimal()\n",
    ")\n",
    "# g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1457a96a-906f-4ca1-9247-a6836f9ae945",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "ggsave(g, filename='2023-05-12_ATAC-seq_LogFC_compare.html', path='.', iframe=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f29414b-0e2f-4bce-9fc7-ab7cf30bd40e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d1200343-58fc-49e6-90bf-a4614e4c3225",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:14:46.533772Z",
     "iopub.status.busy": "2023-05-22T09:14:46.533296Z",
     "iopub.status.idle": "2023-05-22T09:14:54.658329Z",
     "shell.execute_reply": "2023-05-22T09:14:54.656538Z",
     "shell.execute_reply.started": "2023-05-22T09:14:46.533728Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "097edfaf-da48-41de-9f1f-55ee813afc8c",
   "metadata": {},
   "source": [
    "\n",
    "# A simple PyDESeq2 workflow\n",
    "\n",
    "In this example, we show how to perform a simple differential expression analysis on bulk\n",
    "RNAseq data, using PyDESeq2.\n",
    "    :depth: 3\n",
    "\n",
    "We start by importing required packages and setting up an optional path to save results.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "71779e2e-49d3-432e-badc-c1fedbe77816",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:24:43.605747Z",
     "iopub.status.busy": "2023-05-22T09:24:43.605200Z",
     "iopub.status.idle": "2023-05-22T09:24:43.614472Z",
     "shell.execute_reply": "2023-05-22T09:24:43.613056Z",
     "shell.execute_reply.started": "2023-05-22T09:24:43.605701Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pickle as pkl\n",
    "import pandas as pd\n",
    "\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats\n",
    "# from pydeseq2.utils import load_example_data\n",
    "\n",
    "SAVE = False  # whether to save the outputs of this notebook\n",
    "\n",
    "if SAVE:\n",
    "    # Replace this with the path to directory where you would like results to be saved\n",
    "    OUTPUT_PATH = \"../output_files/synthetic_example\"\n",
    "    os.makedirs(OUTPUT_PATH, exist_ok=True)  # Create path if it doesn't exist"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83614665-58ef-47a6-a47d-88327bb9fe66",
   "metadata": {},
   "source": [
    "## Data loading\n",
    "\n",
    "To perform differential expression analysis (DEA), PyDESeq2 requires two types of\n",
    "inputs:\n",
    "\n",
    "  * A count matrix of shape 'number of samples' x 'number of genes', containing\n",
    "    read counts (non-negative integers),\n",
    "  * Clinical data (or \"column\" data) of shape 'number of samples' x\n",
    "    'number of variables', containing sample annotations that will be used\n",
    "    to split the data in cohorts.\n",
    "\n",
    "Both should be provided as [pandas dataframes](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html).\n",
    "\n",
    ".. currentmodule:: pydeseq2\n",
    "\n",
    "To illustrate the required data format, we load a synthetic example dataset that may be\n",
    "obtained through PyDESeq2's API using :func:`utils.load_example_data`.\n",
    "You may replace it with your own dataset.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "c20590e2-edac-4a96-b561-46f6e4dd651a",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:34:57.934650Z",
     "iopub.status.busy": "2023-05-22T09:34:57.933967Z",
     "iopub.status.idle": "2023-05-22T09:34:57.971115Z",
     "shell.execute_reply": "2023-05-22T09:34:57.969975Z",
     "shell.execute_reply.started": "2023-05-22T09:34:57.934592Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           gene1  gene2  gene3  gene4  gene5  gene6  gene7  gene8  gene9   \n",
      "sample1       12     21      4    130     18      0     16     54     49  \\\n",
      "sample2        1     44      2     63     11     10     70     32     57   \n",
      "sample3        4      4     11    180     21      3     28     34     65   \n",
      "sample4        1     10      2    100     44      9     28     16     33   \n",
      "sample5        1     11      6    135     16      2     32     29     31   \n",
      "...          ...    ...    ...    ...    ...    ...    ...    ...    ...   \n",
      "sample96       7     26      3     67     11      4     41     44     54   \n",
      "sample97       1     14      3     71     33      5     19     42     25   \n",
      "sample98      10     36      2     72     11      2     66     27     16   \n",
      "sample99      18     14      3     66     53     11     32     19     79   \n",
      "sample100     21      9      3     42     13     13     19     78     30   \n",
      "\n",
      "           gene10  \n",
      "sample1         3  \n",
      "sample2         9  \n",
      "sample3         2  \n",
      "sample4         9  \n",
      "sample5         5  \n",
      "...           ...  \n",
      "sample96        1  \n",
      "sample97        4  \n",
      "sample98        9  \n",
      "sample99       11  \n",
      "sample100       5  \n",
      "\n",
      "[100 rows x 10 columns]\n",
      "          condition group\n",
      "sample1           A     X\n",
      "sample2           A     Y\n",
      "sample3           A     X\n",
      "sample4           A     Y\n",
      "sample5           A     X\n",
      "...             ...   ...\n",
      "sample96          B     Y\n",
      "sample97          B     X\n",
      "sample98          B     Y\n",
      "sample99          B     X\n",
      "sample100         B     Y\n",
      "\n",
      "[100 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "# counts_df = load_example_data(\n",
    "#     modality=\"raw_counts\",\n",
    "#     dataset=\"synthetic\",\n",
    "#     debug=True,\n",
    "# )\n",
    "\n",
    "# clinical_df = load_example_data(\n",
    "#     modality=\"clinical\",\n",
    "#     dataset=\"synthetic\",\n",
    "#     debug=True,\n",
    "# )\n",
    "counts_df = pd.read_csv('test_counts.csv', index_col=0).T\n",
    "clinical_df = pd.read_csv('test_clinical.csv', index_col=0)\n",
    "print(counts_df)\n",
    "print(clinical_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9261f56-d5b0-4c90-9628-eb41bafe924d",
   "metadata": {},
   "source": [
    "In this example, the clinical data contains two columns, ``condition`` and ``group``,\n",
    "representing two types of bi-level annotations. In the first part, we will only use the\n",
    "``condition`` factor. Later on, we'll see how to use both the `condition` and the\n",
    "``group`` factors in our analysis (see `multifactor_ref`).\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "243bddb4-b970-4311-a6ae-e465b12a745f",
   "metadata": {},
   "source": [
    "### Data filtering\n",
    "\n",
    "Before proceeding with DEA, it is good practice to preprocess your data, e.g. to remove\n",
    "samples for which annotations are missing and exclude genes with very low levels of\n",
    "expression. This is not necessary in the case of our synthetic data, but don't forget\n",
    "this step if you are using real data. To this end you can use the code below.\n",
    "\n",
    "We start by removing samples for which ``condition`` is ``NaN``. If you are using\n",
    "another dataset, do not forget to change \"condition\" for the column of ``clinical_df``\n",
    "you wish to use as a design factor in your analysis.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "4c75b9ee-90d8-434f-8999-f54d2a0ad98e",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:35:22.998073Z",
     "iopub.status.busy": "2023-05-22T09:35:22.997460Z",
     "iopub.status.idle": "2023-05-22T09:35:23.021363Z",
     "shell.execute_reply": "2023-05-22T09:35:23.019827Z",
     "shell.execute_reply.started": "2023-05-22T09:35:22.998023Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           gene1  gene2  gene3  gene4  gene5  gene6  gene7  gene8  gene9   \n",
      "sample1       12     21      4    130     18      0     16     54     49  \\\n",
      "sample2        1     44      2     63     11     10     70     32     57   \n",
      "sample3        4      4     11    180     21      3     28     34     65   \n",
      "sample4        1     10      2    100     44      9     28     16     33   \n",
      "sample5        1     11      6    135     16      2     32     29     31   \n",
      "...          ...    ...    ...    ...    ...    ...    ...    ...    ...   \n",
      "sample96       7     26      3     67     11      4     41     44     54   \n",
      "sample97       1     14      3     71     33      5     19     42     25   \n",
      "sample98      10     36      2     72     11      2     66     27     16   \n",
      "sample99      18     14      3     66     53     11     32     19     79   \n",
      "sample100     21      9      3     42     13     13     19     78     30   \n",
      "\n",
      "           gene10  \n",
      "sample1         3  \n",
      "sample2         9  \n",
      "sample3         2  \n",
      "sample4         9  \n",
      "sample5         5  \n",
      "...           ...  \n",
      "sample96        1  \n",
      "sample97        4  \n",
      "sample98        9  \n",
      "sample99       11  \n",
      "sample100       5  \n",
      "\n",
      "[100 rows x 10 columns]\n",
      "          condition group\n",
      "sample1           A     X\n",
      "sample2           A     Y\n",
      "sample3           A     X\n",
      "sample4           A     Y\n",
      "sample5           A     X\n",
      "...             ...   ...\n",
      "sample96          B     Y\n",
      "sample97          B     X\n",
      "sample98          B     Y\n",
      "sample99          B     X\n",
      "sample100         B     Y\n",
      "\n",
      "[100 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "samples_to_keep = ~clinical_df.condition.isna()\n",
    "counts_df = counts_df.loc[samples_to_keep,:]\n",
    "clinical_df = clinical_df.loc[samples_to_keep, :]\n",
    "\n",
    "print(counts_df)\n",
    "print(clinical_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d4ac7fe-fbb0-452c-988c-2ec755acdb77",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\"><h4>Note</h4><p>In the case where the design factor contains ``NaN`` entries, PyDESeq2 will throw an\n",
    "  error when intializing a :class:`DeseqDataSet <dds.DeseqDataSet>`.</p></div>\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cdf4e6d8-20c7-49b2-9c01-516e97b82ae8",
   "metadata": {},
   "source": [
    "Next, we filter out genes that have less than 10 read counts in total. Note again that\n",
    "there are no such genes in this synthetic dataset.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "671563c8-768c-4ddd-acc2-4035fd110e28",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:35:32.982903Z",
     "iopub.status.busy": "2023-05-22T09:35:32.982278Z",
     "iopub.status.idle": "2023-05-22T09:35:33.013478Z",
     "shell.execute_reply": "2023-05-22T09:35:33.012268Z",
     "shell.execute_reply.started": "2023-05-22T09:35:32.982851Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gene1       838\n",
      "gene2      2259\n",
      "gene3       485\n",
      "gene4     10682\n",
      "gene5      2870\n",
      "gene6       545\n",
      "gene7      2966\n",
      "gene8      4277\n",
      "gene9      3886\n",
      "gene10     1170\n",
      "dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>gene3</th>\n",
       "      <th>gene4</th>\n",
       "      <th>gene5</th>\n",
       "      <th>gene6</th>\n",
       "      <th>gene7</th>\n",
       "      <th>gene8</th>\n",
       "      <th>gene9</th>\n",
       "      <th>gene10</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample1</th>\n",
       "      <td>12</td>\n",
       "      <td>21</td>\n",
       "      <td>4</td>\n",
       "      <td>130</td>\n",
       "      <td>18</td>\n",
       "      <td>0</td>\n",
       "      <td>16</td>\n",
       "      <td>54</td>\n",
       "      <td>49</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample2</th>\n",
       "      <td>1</td>\n",
       "      <td>44</td>\n",
       "      <td>2</td>\n",
       "      <td>63</td>\n",
       "      <td>11</td>\n",
       "      <td>10</td>\n",
       "      <td>70</td>\n",
       "      <td>32</td>\n",
       "      <td>57</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample3</th>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>180</td>\n",
       "      <td>21</td>\n",
       "      <td>3</td>\n",
       "      <td>28</td>\n",
       "      <td>34</td>\n",
       "      <td>65</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample4</th>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>100</td>\n",
       "      <td>44</td>\n",
       "      <td>9</td>\n",
       "      <td>28</td>\n",
       "      <td>16</td>\n",
       "      <td>33</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample5</th>\n",
       "      <td>1</td>\n",
       "      <td>11</td>\n",
       "      <td>6</td>\n",
       "      <td>135</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>32</td>\n",
       "      <td>29</td>\n",
       "      <td>31</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample96</th>\n",
       "      <td>7</td>\n",
       "      <td>26</td>\n",
       "      <td>3</td>\n",
       "      <td>67</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "      <td>41</td>\n",
       "      <td>44</td>\n",
       "      <td>54</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample97</th>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>3</td>\n",
       "      <td>71</td>\n",
       "      <td>33</td>\n",
       "      <td>5</td>\n",
       "      <td>19</td>\n",
       "      <td>42</td>\n",
       "      <td>25</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample98</th>\n",
       "      <td>10</td>\n",
       "      <td>36</td>\n",
       "      <td>2</td>\n",
       "      <td>72</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>66</td>\n",
       "      <td>27</td>\n",
       "      <td>16</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample99</th>\n",
       "      <td>18</td>\n",
       "      <td>14</td>\n",
       "      <td>3</td>\n",
       "      <td>66</td>\n",
       "      <td>53</td>\n",
       "      <td>11</td>\n",
       "      <td>32</td>\n",
       "      <td>19</td>\n",
       "      <td>79</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample100</th>\n",
       "      <td>21</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>42</td>\n",
       "      <td>13</td>\n",
       "      <td>13</td>\n",
       "      <td>19</td>\n",
       "      <td>78</td>\n",
       "      <td>30</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           gene1  gene2  gene3  gene4  gene5  gene6  gene7  gene8  gene9   \n",
       "sample1       12     21      4    130     18      0     16     54     49  \\\n",
       "sample2        1     44      2     63     11     10     70     32     57   \n",
       "sample3        4      4     11    180     21      3     28     34     65   \n",
       "sample4        1     10      2    100     44      9     28     16     33   \n",
       "sample5        1     11      6    135     16      2     32     29     31   \n",
       "...          ...    ...    ...    ...    ...    ...    ...    ...    ...   \n",
       "sample96       7     26      3     67     11      4     41     44     54   \n",
       "sample97       1     14      3     71     33      5     19     42     25   \n",
       "sample98      10     36      2     72     11      2     66     27     16   \n",
       "sample99      18     14      3     66     53     11     32     19     79   \n",
       "sample100     21      9      3     42     13     13     19     78     30   \n",
       "\n",
       "           gene10  \n",
       "sample1         3  \n",
       "sample2         9  \n",
       "sample3         2  \n",
       "sample4         9  \n",
       "sample5         5  \n",
       "...           ...  \n",
       "sample96        1  \n",
       "sample97        4  \n",
       "sample98        9  \n",
       "sample99       11  \n",
       "sample100       5  \n",
       "\n",
       "[100 rows x 10 columns]"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(counts_df.sum(axis=0))\n",
    "genes_to_keep = counts_df.columns[counts_df.sum(axis=0) >= 10]\n",
    "counts_df = counts_df[genes_to_keep]\n",
    "counts_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7de9b946-98e3-4e4b-80ba-a4abd44b0357",
   "metadata": {},
   "source": [
    "Now that we have loaded and filtered our data, we may proceed with the differential\n",
    "analysis.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c000e95-2ac8-4ec9-8d95-8f838b9ab118",
   "metadata": {},
   "source": [
    "## Single factor analysis\n",
    "\n",
    "In this first analysis, we ignore the ``group`` variable and use the ``condition``\n",
    "column as our design factor. That is, we compare gene expressions of samples that have\n",
    "``condition B`` to those that have ``condition A``.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "af45f5d7-2386-4fcf-b7a1-fdc9d97f149d",
   "metadata": {},
   "source": [
    ".. currentmodule:: pydeseq2.dds\n",
    "\n",
    "### Read counts modeling with the :class:`DeseqDataSet` class\n",
    "\n",
    "We start by creating a :class:`DeseqDataSet`\n",
    "object from the count and clinical data.\n",
    "A :class:`DeseqDataSet` fits dispersion and\n",
    "log-fold change (LFC) parameters from the data, and stores them.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "17698796-b503-4bb3-8d62-986b1b2294e8",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:35:51.484041Z",
     "iopub.status.busy": "2023-05-22T09:35:51.483420Z",
     "iopub.status.idle": "2023-05-22T09:35:51.567116Z",
     "shell.execute_reply": "2023-05-22T09:35:51.565905Z",
     "shell.execute_reply.started": "2023-05-22T09:35:51.483991Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "dds = DeseqDataSet(\n",
    "    counts=counts_df,\n",
    "    clinical=clinical_df,\n",
    "    design_factors=\"condition\",\n",
    "    refit_cooks=True,\n",
    "    n_cpus=8,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13f9cdbd-a7d3-4018-80d3-21a8377c35ba",
   "metadata": {},
   "source": [
    "A :class:`DeseqDataSet` has two mandatory\n",
    "arguments: a ``counts`` and a ``clinical`` dataframe, like the ones we've loaded in the\n",
    "first part of this tutorial.\n",
    "\n",
    "Next, we should specify the ``design_factor``, i.e. the column of the ``clinical``\n",
    "dataframe that will be used to compare samples. This can be a single string as above,\n",
    "or a list of strings, as in the\n",
    "`section on multifactor analysis<multifactor_ref>`.\n",
    "\n",
    "<div class=\"alert alert-info\"><h4>Note</h4><p>The ``\"condition\"`` argument passed to ``design_factors`` corresponds to a column\n",
    "  from the ``clinical_df`` dataframe we loaded earlier.\n",
    "  You might need to change it according to your own dataset.</p></div>\n",
    "\n",
    "Several other arguments may be optionally specified (see the :doc:`API documentation\n",
    "</api/docstrings/pydeseq2.dds.DeseqDataSet>`).\n",
    "Among those, the ``refit_cooks`` argument (set to ``True`` by default), controls\n",
    "whether Cooks outlier should be refitted (which is advised, in general) and ``n_cpus``\n",
    "sets the number of CPUs to use for computation. Here, we use 8 threads. Feel free to\n",
    "adapt this to your setup or to set to ``None`` to use all available CPUs.\n",
    "\n",
    "<div class=\"alert alert-info\"><h4>Note</h4><p>In the case of the provided synthetic data, there won't be any Cooks outliers.</p></div>\n",
    "\n",
    "Once a :class:`DeseqDataSet` was initialized,\n",
    "we may run the :meth:`deseq2() <DeseqDataSet.deseq2>` method\n",
    "to fit dispersions and LFCs.\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "34ce8dc4-6c15-4ace-8ea8-5aabfb8fca0f",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:35:56.280597Z",
     "iopub.status.busy": "2023-05-22T09:35:56.280087Z",
     "iopub.status.idle": "2023-05-22T09:35:58.142257Z",
     "shell.execute_reply": "2023-05-22T09:35:58.140648Z",
     "shell.execute_reply.started": "2023-05-22T09:35:56.280553Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.00 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.04 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Refitting 0 outliers.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dds.deseq2()\n",
    "\n",
    "\n",
    "if SAVE:\n",
    "    with open(os.path.join(OUTPUT_PATH, \"dds.pkl\"), \"wb\") as f:\n",
    "        pkl.dump(dds, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6feda5dc-427e-4b9a-8291-6baa9fac600b",
   "metadata": {},
   "source": [
    "The :class:`DeseqDataSet` class extends the\n",
    ":class:`AnnData <anndata.AnnData>`\n",
    "class.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "5af217d1-1f64-485a-8af8-18700f32c069",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:36:03.620654Z",
     "iopub.status.busy": "2023-05-22T09:36:03.619864Z",
     "iopub.status.idle": "2023-05-22T09:36:03.629239Z",
     "shell.execute_reply": "2023-05-22T09:36:03.627533Z",
     "shell.execute_reply.started": "2023-05-22T09:36:03.620590Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 100 × 10\n",
      "    obs: 'condition', 'group'\n",
      "    uns: 'trend_coeffs', '_squared_logres', 'prior_disp_var'\n",
      "    obsm: 'design_matrix', 'size_factors', 'replaceable'\n",
      "    varm: 'non_zero', '_rough_dispersions', 'genewise_dispersions', '_genewise_converged', '_normed_means', 'fitted_dispersions', 'MAP_dispersions', '_MAP_converged', 'dispersions', '_outlier_genes', 'LFC', '_LFC_converged', 'replaced'\n",
      "    layers: 'normed_counts', '_mu_hat', '_mu_LFC', '_hat_diagonals', 'cooks'\n"
     ]
    }
   ],
   "source": [
    "print(dds)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b37b0007-d1e2-40ae-99cc-dc998768ff58",
   "metadata": {},
   "source": [
    "Hence, parameters are stored according to the :class:`AnnData <anndata.AnnData>` data\n",
    "structure, with key-based data fields. In particular,\n",
    "\n",
    "- ``X`` stores the count data,\n",
    "- ``obs`` stores design factors,\n",
    "- ``obsm`` stores sample-level data, such as ``\"design_matrix\"`` and\n",
    "  ``\"size_factors\"``,\n",
    "- ``varm`` stores gene-level data, such as ``\"dispersions\"`` and ``\"LFC\"``.\n",
    "\n",
    "\n",
    "As an example, here is how we would access dispersions and LFCs\n",
    "(in natural log scale):\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "5acffa4c-9593-4074-a6d8-788d78e2a855",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:36:11.816696Z",
     "iopub.status.busy": "2023-05-22T09:36:11.815889Z",
     "iopub.status.idle": "2023-05-22T09:36:11.825549Z",
     "shell.execute_reply": "2023-05-22T09:36:11.823842Z",
     "shell.execute_reply.started": "2023-05-22T09:36:11.816646Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0.88259868 0.22257863 0.83723801 0.15897048 0.24992589 0.97364795\n",
      " 0.23515487 0.19878078 0.1865203  0.63189989]\n"
     ]
    }
   ],
   "source": [
    "print(dds.varm[\"dispersions\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "d1941ba2-68e8-41a9-b438-1aa3743c0ec3",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:36:13.135291Z",
     "iopub.status.busy": "2023-05-22T09:36:13.134039Z",
     "iopub.status.idle": "2023-05-22T09:36:13.146914Z",
     "shell.execute_reply": "2023-05-22T09:36:13.145358Z",
     "shell.execute_reply.started": "2023-05-22T09:36:13.135241Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        intercept  condition_B_vs_A\n",
      "gene1    1.891436          0.438632\n",
      "gene2    2.851662          0.373296\n",
      "gene3    1.787780         -0.438645\n",
      "gene4    4.741958         -0.285647\n",
      "gene5    3.077798          0.403457\n",
      "gene6    1.678536          0.001010\n",
      "gene7    3.291025          0.093116\n",
      "gene8    3.785129         -0.187604\n",
      "gene9    3.682882         -0.147443\n",
      "gene10   2.300515          0.267562\n"
     ]
    }
   ],
   "source": [
    "print(dds.varm[\"LFC\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1bda495-f1e7-4eef-be10-ef98634055ed",
   "metadata": {},
   "source": [
    ".. currentmodule:: pydeseq2.ds\n",
    "\n",
    "### Statistical analysis with the :class:`DeseqStats` class\n",
    "\n",
    "Now that dispersions and LFCs were fitted, we may proceed with statistical tests to\n",
    "compute p-values and adjusted p-values for differential expresion. This is the role of\n",
    "the :class:`DeseqStats` class. It has a unique mandatory argument, ``dds``, which\n",
    "should be a *fitted* :class:`DeseqDataSet <pydeseq2.dds.DeseqDataSet>`\n",
    "object.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "1b1723cb-c7dd-42a9-b183-947b57c1f81f",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:36:16.378141Z",
     "iopub.status.busy": "2023-05-22T09:36:16.377579Z",
     "iopub.status.idle": "2023-05-22T09:36:16.385847Z",
     "shell.execute_reply": "2023-05-22T09:36:16.384442Z",
     "shell.execute_reply.started": "2023-05-22T09:36:16.378090Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "stat_res = DeseqStats(dds, n_cpus=8)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c946358-68fb-4443-b235-2745cae444fd",
   "metadata": {},
   "source": [
    "It also has a set of optional keyword arguments (see the :doc:`API documentation\n",
    "</api/docstrings/pydeseq2.ds.DeseqStats>`), among which:\n",
    "\n",
    "- ``alpha``: the p-value and adjusted p-value significance threshold (``0.05``\n",
    "  by default),\n",
    "- ``cooks_filter``: whether to filter p-values based on cooks outliers\n",
    "  (``True`` by default),\n",
    "- ``independent_filter``: whether to perform independent filtering to correct\n",
    "  p-value trends (``True`` by default).\n",
    "\n",
    "In the `section on multifactor analysis<multifactor_ref>`, we will also see how\n",
    "to use the ``contrast`` argument to specify according to which variable samples should\n",
    "be compared.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a41497bd-79e1-4f4f-994c-85c5018a79e8",
   "metadata": {},
   "source": [
    "\n",
    "#### Wald test\n",
    "\n",
    "PyDESeq2 computes p-values using Wald tests. This can be done using the\n",
    ":meth:`summary() <DeseqStats.summary>` method, which runs the whole statistical\n",
    "analysis, cooks filtering and multiple testing adjustement included.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "47fcbf5e-44a2-4ba1-936b-9674060afa39",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:36:18.932335Z",
     "iopub.status.busy": "2023-05-22T09:36:18.931769Z",
     "iopub.status.idle": "2023-05-22T09:36:19.038417Z",
     "shell.execute_reply": "2023-05-22T09:36:19.037657Z",
     "shell.execute_reply.started": "2023-05-22T09:36:18.932286Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Wald tests...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Log2 fold change & Wald test p-value: condition B vs A\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene1</th>\n",
       "      <td>8.541317</td>\n",
       "      <td>0.632812</td>\n",
       "      <td>0.289101</td>\n",
       "      <td>2.188898</td>\n",
       "      <td>0.028604</td>\n",
       "      <td>0.064150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene2</th>\n",
       "      <td>21.281239</td>\n",
       "      <td>0.538552</td>\n",
       "      <td>0.149963</td>\n",
       "      <td>3.591235</td>\n",
       "      <td>0.000329</td>\n",
       "      <td>0.001646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene3</th>\n",
       "      <td>5.010123</td>\n",
       "      <td>-0.632830</td>\n",
       "      <td>0.295236</td>\n",
       "      <td>-2.143475</td>\n",
       "      <td>0.032075</td>\n",
       "      <td>0.064150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene4</th>\n",
       "      <td>100.517961</td>\n",
       "      <td>-0.412102</td>\n",
       "      <td>0.118629</td>\n",
       "      <td>-3.473867</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.001710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene5</th>\n",
       "      <td>27.142450</td>\n",
       "      <td>0.582065</td>\n",
       "      <td>0.154706</td>\n",
       "      <td>3.762408</td>\n",
       "      <td>0.000168</td>\n",
       "      <td>0.001646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene6</th>\n",
       "      <td>5.413043</td>\n",
       "      <td>0.001457</td>\n",
       "      <td>0.310311</td>\n",
       "      <td>0.004696</td>\n",
       "      <td>0.996253</td>\n",
       "      <td>0.996253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene7</th>\n",
       "      <td>28.294023</td>\n",
       "      <td>0.134338</td>\n",
       "      <td>0.149945</td>\n",
       "      <td>0.895917</td>\n",
       "      <td>0.370297</td>\n",
       "      <td>0.411441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene8</th>\n",
       "      <td>40.358344</td>\n",
       "      <td>-0.270656</td>\n",
       "      <td>0.136401</td>\n",
       "      <td>-1.984260</td>\n",
       "      <td>0.047227</td>\n",
       "      <td>0.078711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene9</th>\n",
       "      <td>37.166183</td>\n",
       "      <td>-0.212715</td>\n",
       "      <td>0.133243</td>\n",
       "      <td>-1.596437</td>\n",
       "      <td>0.110391</td>\n",
       "      <td>0.143148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene10</th>\n",
       "      <td>11.589325</td>\n",
       "      <td>0.386011</td>\n",
       "      <td>0.244588</td>\n",
       "      <td>1.578207</td>\n",
       "      <td>0.114518</td>\n",
       "      <td>0.143148</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
       "gene1     8.541317        0.632812  0.289101  2.188898  0.028604  0.064150\n",
       "gene2    21.281239        0.538552  0.149963  3.591235  0.000329  0.001646\n",
       "gene3     5.010123       -0.632830  0.295236 -2.143475  0.032075  0.064150\n",
       "gene4   100.517961       -0.412102  0.118629 -3.473867  0.000513  0.001710\n",
       "gene5    27.142450        0.582065  0.154706  3.762408  0.000168  0.001646\n",
       "gene6     5.413043        0.001457  0.310311  0.004696  0.996253  0.996253\n",
       "gene7    28.294023        0.134338  0.149945  0.895917  0.370297  0.411441\n",
       "gene8    40.358344       -0.270656  0.136401 -1.984260  0.047227  0.078711\n",
       "gene9    37.166183       -0.212715  0.133243 -1.596437  0.110391  0.143148\n",
       "gene10   11.589325        0.386011  0.244588  1.578207  0.114518  0.143148"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "stat_res.summary()\n",
    "\n",
    "if SAVE:\n",
    "    with open(os.path.join(OUTPUT_PATH, \"stat_results.pkl\"), \"wb\") as f:\n",
    "        pkl.dump(stat_res, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1dde452-3d76-47b5-ba83-11b2febb19aa",
   "metadata": {},
   "source": [
    "The results are then stored in the ``results_df`` attribute (``stat_res.results_df``).\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6967c274-e916-423d-a578-4d9c9ac72ab0",
   "metadata": {},
   "source": [
    "#### LFC shrinkage\n",
    "\n",
    "For visualization or post-processing purposes, it might be suitable to perform\n",
    "LFC shrinkage. This is implemented by the :meth:`lfc_shrink() <DeseqStats.lfc_shrink>`\n",
    "method.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "568f7747-335d-4bd0-ab16-e5553a1e535b",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:36:35.870841Z",
     "iopub.status.busy": "2023-05-22T09:36:35.870286Z",
     "iopub.status.idle": "2023-05-22T09:36:35.933035Z",
     "shell.execute_reply": "2023-05-22T09:36:35.932018Z",
     "shell.execute_reply.started": "2023-05-22T09:36:35.870792Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting MAP LFCs...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Shrunk Log2 fold change & Wald test p-value: condition B vs A\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene1</th>\n",
       "      <td>8.541317</td>\n",
       "      <td>0.408253</td>\n",
       "      <td>0.294276</td>\n",
       "      <td>2.188898</td>\n",
       "      <td>0.028604</td>\n",
       "      <td>0.064150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene2</th>\n",
       "      <td>21.281239</td>\n",
       "      <td>0.480145</td>\n",
       "      <td>0.151201</td>\n",
       "      <td>3.591235</td>\n",
       "      <td>0.000329</td>\n",
       "      <td>0.001646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene3</th>\n",
       "      <td>5.010123</td>\n",
       "      <td>-0.396066</td>\n",
       "      <td>0.300796</td>\n",
       "      <td>-2.143475</td>\n",
       "      <td>0.032075</td>\n",
       "      <td>0.064150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene4</th>\n",
       "      <td>100.517961</td>\n",
       "      <td>-0.374191</td>\n",
       "      <td>0.118704</td>\n",
       "      <td>-3.473867</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.001710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene5</th>\n",
       "      <td>27.142450</td>\n",
       "      <td>0.521487</td>\n",
       "      <td>0.156210</td>\n",
       "      <td>3.762408</td>\n",
       "      <td>0.000168</td>\n",
       "      <td>0.001646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene6</th>\n",
       "      <td>5.413043</td>\n",
       "      <td>0.000716</td>\n",
       "      <td>0.239203</td>\n",
       "      <td>0.004696</td>\n",
       "      <td>0.996253</td>\n",
       "      <td>0.996253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene7</th>\n",
       "      <td>28.294023</td>\n",
       "      <td>0.103421</td>\n",
       "      <td>0.141496</td>\n",
       "      <td>0.895917</td>\n",
       "      <td>0.370297</td>\n",
       "      <td>0.411441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene8</th>\n",
       "      <td>40.358344</td>\n",
       "      <td>-0.226288</td>\n",
       "      <td>0.133477</td>\n",
       "      <td>-1.984260</td>\n",
       "      <td>0.047227</td>\n",
       "      <td>0.078711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene9</th>\n",
       "      <td>37.166183</td>\n",
       "      <td>-0.175746</td>\n",
       "      <td>0.129138</td>\n",
       "      <td>-1.596437</td>\n",
       "      <td>0.110391</td>\n",
       "      <td>0.143148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene10</th>\n",
       "      <td>11.589325</td>\n",
       "      <td>0.239935</td>\n",
       "      <td>0.231986</td>\n",
       "      <td>1.578207</td>\n",
       "      <td>0.114518</td>\n",
       "      <td>0.143148</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
       "gene1     8.541317        0.408253  0.294276  2.188898  0.028604  0.064150\n",
       "gene2    21.281239        0.480145  0.151201  3.591235  0.000329  0.001646\n",
       "gene3     5.010123       -0.396066  0.300796 -2.143475  0.032075  0.064150\n",
       "gene4   100.517961       -0.374191  0.118704 -3.473867  0.000513  0.001710\n",
       "gene5    27.142450        0.521487  0.156210  3.762408  0.000168  0.001646\n",
       "gene6     5.413043        0.000716  0.239203  0.004696  0.996253  0.996253\n",
       "gene7    28.294023        0.103421  0.141496  0.895917  0.370297  0.411441\n",
       "gene8    40.358344       -0.226288  0.133477 -1.984260  0.047227  0.078711\n",
       "gene9    37.166183       -0.175746  0.129138 -1.596437  0.110391  0.143148\n",
       "gene10   11.589325        0.239935  0.231986  1.578207  0.114518  0.143148"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "stat_res.lfc_shrink(coeff=\"condition_B_vs_A\")\n",
    "\n",
    "if SAVE:\n",
    "    with open(os.path.join(OUTPUT_PATH, \"shrunk_stat_results.pkl\"), \"wb\") as f:\n",
    "        pkl.dump(stat_res, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e1c0db9-0917-4a94-b45f-2ac8689f397d",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\"><h4>Note</h4><p>Running :meth:`lfc_shrink() <DeseqStats.lfc_shrink>` will overwrite a\n",
    "  :class:`DeseqStats`' log fold changes (and standard errors) with shrunk values.\n",
    "  This can be checked using the ``shrunk_LFCs`` flag.</p></div>\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "971e9a66-8e19-479e-bdd6-c4e5b5020e5f",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:36:37.788175Z",
     "iopub.status.busy": "2023-05-22T09:36:37.787619Z",
     "iopub.status.idle": "2023-05-22T09:36:37.796000Z",
     "shell.execute_reply": "2023-05-22T09:36:37.794347Z",
     "shell.execute_reply.started": "2023-05-22T09:36:37.788125Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "print(stat_res.shrunk_LFCs)  # Will be True only if lfc_shrink() was run."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3b5384a-724c-426e-b966-c93432491da2",
   "metadata": {},
   "source": [
    "\n",
    "## Multifactor analysis\n",
    "\n",
    ".. currentmodule:: pydeseq2.dds\n",
    "\n",
    "So far, we have only used the ``condition`` column of ``clinical_df``, which divides\n",
    "samples between conditions ``A`` and ``B``. Yet, ``clinical_df`` contains second\n",
    "column, which separates samples according to ``group`` ``X`` and ``Y``.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "52d8be5d-936f-4ca6-b7dc-5ac4a5a73d73",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:36:40.379348Z",
     "iopub.status.busy": "2023-05-22T09:36:40.378786Z",
     "iopub.status.idle": "2023-05-22T09:36:40.391352Z",
     "shell.execute_reply": "2023-05-22T09:36:40.389886Z",
     "shell.execute_reply.started": "2023-05-22T09:36:40.379298Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          condition group\n",
      "sample1           A     X\n",
      "sample2           A     Y\n",
      "sample3           A     X\n",
      "sample4           A     Y\n",
      "sample5           A     X\n",
      "...             ...   ...\n",
      "sample96          B     Y\n",
      "sample97          B     X\n",
      "sample98          B     Y\n",
      "sample99          B     X\n",
      "sample100         B     Y\n",
      "\n",
      "[100 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "print(clinical_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9483ae1f-a346-493e-8a8c-a222bdf4dc66",
   "metadata": {},
   "source": [
    "The goal of multifactor analysis is to use *both* variables to fit LFCs.\n",
    "\n",
    "### Read counts modeling\n",
    "\n",
    "To perform multifactor analysis with PyDESeq2, we start by inializing a\n",
    ":class:`DeseqDataSet` as previously, but we provide the list of variables we would like\n",
    "to use in the ``design_factors`` argument.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "6c5f9432-8d69-4ba5-a0a2-df1627b32b67",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:36:43.719300Z",
     "iopub.status.busy": "2023-05-22T09:36:43.718731Z",
     "iopub.status.idle": "2023-05-22T09:36:43.752916Z",
     "shell.execute_reply": "2023-05-22T09:36:43.751707Z",
     "shell.execute_reply.started": "2023-05-22T09:36:43.719251Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "dds = DeseqDataSet(\n",
    "    counts=counts_df,\n",
    "    clinical=clinical_df,\n",
    "    design_factors=[\"group\", \"condition\"],\n",
    "    refit_cooks=True,\n",
    "    n_cpus=8,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f09d437-6d95-43d4-b819-9addf917fc08",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\"><h4>Note</h4><p>By default, the last variable in the list (here, ``\"condition\"``) will be the one for\n",
    "  which LFCs and p-values will be displayed, but this may be changed later on when\n",
    "  performing the statistical analysis.</p></div>\n",
    "\n",
    "As for the single-factor analysis, we fit dispersions and LFCs using the\n",
    ":meth:`deseq2() <DeseqDataSet.deseq2>` method.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "de723736-8ebe-42b0-9563-bfe83a7b21a3",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:36:44.998258Z",
     "iopub.status.busy": "2023-05-22T09:36:44.997254Z",
     "iopub.status.idle": "2023-05-22T09:36:45.117063Z",
     "shell.execute_reply": "2023-05-22T09:36:45.116261Z",
     "shell.execute_reply.started": "2023-05-22T09:36:44.998196Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.00 seconds.\n",
      "\n",
      "Fitting dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting dispersion trend curve...\n",
      "... done in 0.03 seconds.\n",
      "\n",
      "Fitting MAP dispersions...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Fitting LFCs...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Refitting 0 outliers.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dds.deseq2()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54f30d38-f1db-4135-850f-3bcf3cd9da35",
   "metadata": {},
   "source": [
    "Now, if we print log fold changes, we will have two columns in addition to the\n",
    "intercept: one corresponding to the ``group`` variable, and the other to ``condition``.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "3a103ade-0b78-4f7b-804b-eb07b3838050",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:36:46.623929Z",
     "iopub.status.busy": "2023-05-22T09:36:46.623375Z",
     "iopub.status.idle": "2023-05-22T09:36:46.636447Z",
     "shell.execute_reply": "2023-05-22T09:36:46.634982Z",
     "shell.execute_reply.started": "2023-05-22T09:36:46.623880Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        intercept  group_Y_vs_X  condition_B_vs_A\n",
      "gene1    1.560313      0.525572          0.507013\n",
      "gene2    2.812255      0.079229          0.371026\n",
      "gene3    2.059577     -0.602606         -0.467002\n",
      "gene4    4.919837     -0.385197         -0.293527\n",
      "gene5    2.973802      0.194078          0.407582\n",
      "gene6    1.846061     -0.352410         -0.013510\n",
      "gene7    3.235967      0.106773          0.093654\n",
      "gene8    3.640464      0.270819         -0.188248\n",
      "gene9    3.645334      0.074474         -0.148346\n",
      "gene10   2.136670      0.303009          0.270003\n"
     ]
    }
   ],
   "source": [
    "print(dds.varm[\"LFC\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6a48426-3ed2-4e75-b72c-3fdab4f0845a",
   "metadata": {},
   "source": [
    ".. currentmodule:: pydeseq2.ds\n",
    "\n",
    "### Statistical analysis\n",
    "\n",
    "P-values are computed as earlier from a :class:`DeseqStats` object with the\n",
    ":meth:`summary() <DeseqStats.summary>` method, with a new important argument:\n",
    "the ``contrast``.\n",
    "It is a list of three strings of the form\n",
    "``[\"variable\", \"tested level\", \"reference level\"]`` which determines which\n",
    "variable we want to compute LFCs and pvalues for.\n",
    "As an example, to compare the condition B to the condition A, we set\n",
    "``contrast=[\"condition\", \"B\", \"A\"]``.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "ad288ae2-d14e-46f4-965d-8fe5fb42bb8f",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:36:47.689429Z",
     "iopub.status.busy": "2023-05-22T09:36:47.688854Z",
     "iopub.status.idle": "2023-05-22T09:36:47.697822Z",
     "shell.execute_reply": "2023-05-22T09:36:47.696266Z",
     "shell.execute_reply.started": "2023-05-22T09:36:47.689378Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "stat_res_B_vs_A = DeseqStats(dds, contrast=[\"condition\", \"B\", \"A\"], n_cpus=8)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21f4eb13-a19d-4a7b-9656-7eccd563eac0",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-info\"><h4>Note</h4><p>If left blank, the variable of interest will be the last one provided in\n",
    "  the ``design_factors`` attribute of the corresponding\n",
    "  :class:`DeseqDataSet <pydeseq2.dds.DeseqDataSet>` object,\n",
    "  and the reference level will be picked alphabetically.\n",
    "  In any case, *both variables are still used*. This is due to the fact that ``dds``\n",
    "  was fit with both as design factors.</p></div>\n",
    "\n",
    "Let us fit p-values:\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "5df5e45c-bee5-4b13-9502-b28e3b257c3c",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:36:51.128836Z",
     "iopub.status.busy": "2023-05-22T09:36:51.128279Z",
     "iopub.status.idle": "2023-05-22T09:36:51.235831Z",
     "shell.execute_reply": "2023-05-22T09:36:51.235039Z",
     "shell.execute_reply.started": "2023-05-22T09:36:51.128786Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Wald tests...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Log2 fold change & Wald test p-value: condition B vs A\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene1</th>\n",
       "      <td>8.541317</td>\n",
       "      <td>0.731466</td>\n",
       "      <td>0.286275</td>\n",
       "      <td>2.555117</td>\n",
       "      <td>0.010615</td>\n",
       "      <td>0.026538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene2</th>\n",
       "      <td>21.281239</td>\n",
       "      <td>0.535278</td>\n",
       "      <td>0.149824</td>\n",
       "      <td>3.572711</td>\n",
       "      <td>0.000353</td>\n",
       "      <td>0.001178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene3</th>\n",
       "      <td>5.010123</td>\n",
       "      <td>-0.673742</td>\n",
       "      <td>0.287400</td>\n",
       "      <td>-2.344268</td>\n",
       "      <td>0.019064</td>\n",
       "      <td>0.038129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene4</th>\n",
       "      <td>100.517961</td>\n",
       "      <td>-0.423471</td>\n",
       "      <td>0.106221</td>\n",
       "      <td>-3.986712</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>0.000592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene5</th>\n",
       "      <td>27.142450</td>\n",
       "      <td>0.588016</td>\n",
       "      <td>0.152758</td>\n",
       "      <td>3.849332</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>0.000592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene6</th>\n",
       "      <td>5.413043</td>\n",
       "      <td>-0.019490</td>\n",
       "      <td>0.307845</td>\n",
       "      <td>-0.063313</td>\n",
       "      <td>0.949518</td>\n",
       "      <td>0.949518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene7</th>\n",
       "      <td>28.294023</td>\n",
       "      <td>0.135114</td>\n",
       "      <td>0.149561</td>\n",
       "      <td>0.903410</td>\n",
       "      <td>0.366308</td>\n",
       "      <td>0.407009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene8</th>\n",
       "      <td>40.358344</td>\n",
       "      <td>-0.271584</td>\n",
       "      <td>0.131512</td>\n",
       "      <td>-2.065084</td>\n",
       "      <td>0.038915</td>\n",
       "      <td>0.064858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene9</th>\n",
       "      <td>37.166183</td>\n",
       "      <td>-0.214018</td>\n",
       "      <td>0.132989</td>\n",
       "      <td>-1.609288</td>\n",
       "      <td>0.107553</td>\n",
       "      <td>0.139684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene10</th>\n",
       "      <td>11.589325</td>\n",
       "      <td>0.389532</td>\n",
       "      <td>0.244929</td>\n",
       "      <td>1.590388</td>\n",
       "      <td>0.111747</td>\n",
       "      <td>0.139684</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
       "gene1     8.541317        0.731466  0.286275  2.555117  0.010615  0.026538\n",
       "gene2    21.281239        0.535278  0.149824  3.572711  0.000353  0.001178\n",
       "gene3     5.010123       -0.673742  0.287400 -2.344268  0.019064  0.038129\n",
       "gene4   100.517961       -0.423471  0.106221 -3.986712  0.000067  0.000592\n",
       "gene5    27.142450        0.588016  0.152758  3.849332  0.000118  0.000592\n",
       "gene6     5.413043       -0.019490  0.307845 -0.063313  0.949518  0.949518\n",
       "gene7    28.294023        0.135114  0.149561  0.903410  0.366308  0.407009\n",
       "gene8    40.358344       -0.271584  0.131512 -2.065084  0.038915  0.064858\n",
       "gene9    37.166183       -0.214018  0.132989 -1.609288  0.107553  0.139684\n",
       "gene10   11.589325        0.389532  0.244929  1.590388  0.111747  0.139684"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "stat_res_B_vs_A.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02de6f17-4174-4def-abb9-5ae477803714",
   "metadata": {},
   "source": [
    "As we can see, although we are comparing the same cohorts (condition B vs A), the\n",
    "results differ from the `single-factor analysis <wald_ref>`. This is because the\n",
    "model uses information from both the ``condition`` and ``group`` variables.\n",
    "\n",
    "Let us now evaluate differential expression according to group Y vs X. To do so,\n",
    "we create a new :class:`DeseqStats` from the same\n",
    ":class:`DeseqDataSet <pydeseq2.dds.DeseqDataSet>`\n",
    "with ``contrast=[\"group\", \"Y\", \"X\"]``, and run the analysis again.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "bc492d60-a52f-4e67-9140-8a6dab8e0f36",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:36:59.759557Z",
     "iopub.status.busy": "2023-05-22T09:36:59.758992Z",
     "iopub.status.idle": "2023-05-22T09:36:59.869393Z",
     "shell.execute_reply": "2023-05-22T09:36:59.868585Z",
     "shell.execute_reply.started": "2023-05-22T09:36:59.759508Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Wald tests...\n",
      "... done in 0.01 seconds.\n",
      "\n",
      "Log2 fold change & Wald test p-value: group Y vs X\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene1</th>\n",
       "      <td>8.541317</td>\n",
       "      <td>0.758241</td>\n",
       "      <td>0.286328</td>\n",
       "      <td>2.648155</td>\n",
       "      <td>8.093232e-03</td>\n",
       "      <td>0.020233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene2</th>\n",
       "      <td>21.281239</td>\n",
       "      <td>0.114304</td>\n",
       "      <td>0.149771</td>\n",
       "      <td>0.763189</td>\n",
       "      <td>4.453508e-01</td>\n",
       "      <td>0.445351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene3</th>\n",
       "      <td>5.010123</td>\n",
       "      <td>-0.869376</td>\n",
       "      <td>0.287516</td>\n",
       "      <td>-3.023748</td>\n",
       "      <td>2.496646e-03</td>\n",
       "      <td>0.009903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene4</th>\n",
       "      <td>100.517961</td>\n",
       "      <td>-0.555721</td>\n",
       "      <td>0.106221</td>\n",
       "      <td>-5.231722</td>\n",
       "      <td>1.679380e-07</td>\n",
       "      <td>0.000002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene5</th>\n",
       "      <td>27.142450</td>\n",
       "      <td>0.279995</td>\n",
       "      <td>0.152731</td>\n",
       "      <td>1.833256</td>\n",
       "      <td>6.676450e-02</td>\n",
       "      <td>0.123874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene6</th>\n",
       "      <td>5.413043</td>\n",
       "      <td>-0.508421</td>\n",
       "      <td>0.307951</td>\n",
       "      <td>-1.650980</td>\n",
       "      <td>9.874262e-02</td>\n",
       "      <td>0.141061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene7</th>\n",
       "      <td>28.294023</td>\n",
       "      <td>0.154041</td>\n",
       "      <td>0.149565</td>\n",
       "      <td>1.029927</td>\n",
       "      <td>3.030443e-01</td>\n",
       "      <td>0.378805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene8</th>\n",
       "      <td>40.358344</td>\n",
       "      <td>0.390709</td>\n",
       "      <td>0.131519</td>\n",
       "      <td>2.970732</td>\n",
       "      <td>2.970912e-03</td>\n",
       "      <td>0.009903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene9</th>\n",
       "      <td>37.166183</td>\n",
       "      <td>0.107443</td>\n",
       "      <td>0.132982</td>\n",
       "      <td>0.807956</td>\n",
       "      <td>4.191157e-01</td>\n",
       "      <td>0.445351</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene10</th>\n",
       "      <td>11.589325</td>\n",
       "      <td>0.437149</td>\n",
       "      <td>0.244955</td>\n",
       "      <td>1.784611</td>\n",
       "      <td>7.432438e-02</td>\n",
       "      <td>0.123874</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          baseMean  log2FoldChange     lfcSE      stat        pvalue      padj\n",
       "gene1     8.541317        0.758241  0.286328  2.648155  8.093232e-03  0.020233\n",
       "gene2    21.281239        0.114304  0.149771  0.763189  4.453508e-01  0.445351\n",
       "gene3     5.010123       -0.869376  0.287516 -3.023748  2.496646e-03  0.009903\n",
       "gene4   100.517961       -0.555721  0.106221 -5.231722  1.679380e-07  0.000002\n",
       "gene5    27.142450        0.279995  0.152731  1.833256  6.676450e-02  0.123874\n",
       "gene6     5.413043       -0.508421  0.307951 -1.650980  9.874262e-02  0.141061\n",
       "gene7    28.294023        0.154041  0.149565  1.029927  3.030443e-01  0.378805\n",
       "gene8    40.358344        0.390709  0.131519  2.970732  2.970912e-03  0.009903\n",
       "gene9    37.166183        0.107443  0.132982  0.807956  4.191157e-01  0.445351\n",
       "gene10   11.589325        0.437149  0.244955  1.784611  7.432438e-02  0.123874"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "stat_res_Y_vs_X = DeseqStats(dds, contrast=[\"group\", \"Y\", \"X\"], n_cpus=8)\n",
    "stat_res_Y_vs_X.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d1a2067-1df6-4623-99a2-7f31743e5e0a",
   "metadata": {},
   "source": [
    "#### LFC shrinkage (multifactor)\n",
    "\n",
    "In a multifactor setting, LFC shrinkage works as in the single-factor case, but will\n",
    "only shrink the LFCs of a :class:`DeseqStats` object based on its\n",
    "``contrast`` argument.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "3ef4f997-5d99-4265-95e3-eed1db328a5b",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T09:37:04.400492Z",
     "iopub.status.busy": "2023-05-22T09:37:04.399972Z",
     "iopub.status.idle": "2023-05-22T09:37:04.457738Z",
     "shell.execute_reply": "2023-05-22T09:37:04.456833Z",
     "shell.execute_reply.started": "2023-05-22T09:37:04.400448Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting MAP LFCs...\n",
      "... done in 0.02 seconds.\n",
      "\n",
      "Shrunk Log2 fold change & Wald test p-value: condition B vs A\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene1</th>\n",
       "      <td>8.541317</td>\n",
       "      <td>0.526936</td>\n",
       "      <td>0.299297</td>\n",
       "      <td>2.555117</td>\n",
       "      <td>0.010615</td>\n",
       "      <td>0.026538</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene2</th>\n",
       "      <td>21.281239</td>\n",
       "      <td>0.479377</td>\n",
       "      <td>0.150834</td>\n",
       "      <td>3.572711</td>\n",
       "      <td>0.000353</td>\n",
       "      <td>0.001178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene3</th>\n",
       "      <td>5.010123</td>\n",
       "      <td>-0.463492</td>\n",
       "      <td>0.296136</td>\n",
       "      <td>-2.344268</td>\n",
       "      <td>0.019064</td>\n",
       "      <td>0.038129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene4</th>\n",
       "      <td>100.517961</td>\n",
       "      <td>-0.394989</td>\n",
       "      <td>0.106316</td>\n",
       "      <td>-3.986712</td>\n",
       "      <td>0.000067</td>\n",
       "      <td>0.000592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene5</th>\n",
       "      <td>27.142450</td>\n",
       "      <td>0.531655</td>\n",
       "      <td>0.154062</td>\n",
       "      <td>3.849332</td>\n",
       "      <td>0.000118</td>\n",
       "      <td>0.000592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene6</th>\n",
       "      <td>5.413043</td>\n",
       "      <td>-0.008636</td>\n",
       "      <td>0.241570</td>\n",
       "      <td>-0.063313</td>\n",
       "      <td>0.949518</td>\n",
       "      <td>0.949518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene7</th>\n",
       "      <td>28.294023</td>\n",
       "      <td>0.106695</td>\n",
       "      <td>0.141879</td>\n",
       "      <td>0.903410</td>\n",
       "      <td>0.366308</td>\n",
       "      <td>0.407009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene8</th>\n",
       "      <td>40.358344</td>\n",
       "      <td>-0.233346</td>\n",
       "      <td>0.129070</td>\n",
       "      <td>-2.065084</td>\n",
       "      <td>0.038915</td>\n",
       "      <td>0.064858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene9</th>\n",
       "      <td>37.166183</td>\n",
       "      <td>-0.180064</td>\n",
       "      <td>0.129246</td>\n",
       "      <td>-1.609288</td>\n",
       "      <td>0.107553</td>\n",
       "      <td>0.139684</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene10</th>\n",
       "      <td>11.589325</td>\n",
       "      <td>0.251471</td>\n",
       "      <td>0.233140</td>\n",
       "      <td>1.590388</td>\n",
       "      <td>0.111747</td>\n",
       "      <td>0.139684</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
       "gene1     8.541317        0.526936  0.299297  2.555117  0.010615  0.026538\n",
       "gene2    21.281239        0.479377  0.150834  3.572711  0.000353  0.001178\n",
       "gene3     5.010123       -0.463492  0.296136 -2.344268  0.019064  0.038129\n",
       "gene4   100.517961       -0.394989  0.106316 -3.986712  0.000067  0.000592\n",
       "gene5    27.142450        0.531655  0.154062  3.849332  0.000118  0.000592\n",
       "gene6     5.413043       -0.008636  0.241570 -0.063313  0.949518  0.949518\n",
       "gene7    28.294023        0.106695  0.141879  0.903410  0.366308  0.407009\n",
       "gene8    40.358344       -0.233346  0.129070 -2.065084  0.038915  0.064858\n",
       "gene9    37.166183       -0.180064  0.129246 -1.609288  0.107553  0.139684\n",
       "gene10   11.589325        0.251471  0.233140  1.590388  0.111747  0.139684"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "stat_res_B_vs_A.lfc_shrink(coeff=\"condition_B_vs_A\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6cd8fa2e-98f2-4ee8-8683-384f1b78dc11",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:30:23.155098Z",
     "iopub.status.busy": "2023-05-22T12:30:23.154752Z",
     "iopub.status.idle": "2023-05-22T12:30:33.718545Z",
     "shell.execute_reply": "2023-05-22T12:30:33.717611Z",
     "shell.execute_reply.started": "2023-05-22T12:30:23.155065Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4f77906-6856-4a72-97cf-0b8826ae7b25",
   "metadata": {},
   "source": [
    "\n",
    "# Step-by-step PyDESeq2 workflow\n",
    "\n",
    "This notebook details all the steps of the PyDESeq2 pipeline.\n",
    "\n",
    "It allows you to run the PyDESeq2 pipeline on the synthetic data provided\n",
    "in this repository.\n",
    "\n",
    "If this is your first contact with PyDESeq2, we recommend you first have a look at the\n",
    ":doc:`standard workflow example <plot_minimal_pydeseq2_pipeline>`.\n",
    "    :depth: 3\n",
    "\n",
    "We start by importing required packages and setting up an optional path to save\n",
    "results.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "98b2027f-ac16-4484-b8ea-5a949db78ce1",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:34:44.522377Z",
     "iopub.status.busy": "2023-05-22T12:34:44.517899Z",
     "iopub.status.idle": "2023-05-22T12:34:44.555565Z",
     "shell.execute_reply": "2023-05-22T12:34:44.554806Z",
     "shell.execute_reply.started": "2023-05-22T12:34:44.522342Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pickle as pkl\n",
    "import pandas as pd\n",
    "\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats\n",
    "# from pydeseq2.utils import load_example_data\n",
    "\n",
    "SAVE = False  # whether to save the outputs of this notebook\n",
    "\n",
    "if SAVE:\n",
    "    # Replace this with the path to directory where you would like results to be\n",
    "    # saved\n",
    "    OUTPUT_PATH = \"../output_files/synthetic_example\"\n",
    "    os.makedirs(OUTPUT_PATH, exist_ok=True)  # Create path if it doesn't exist"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "959aef0f-7d00-4968-aced-dc7238cf7776",
   "metadata": {},
   "source": [
    "## Data loading\n",
    "\n",
    "Note that we are also explaining this step in the 'getting started' example.\n",
    "To perform differential expression analysis (DEA), PyDESeq2 requires two types of\n",
    "inputs:\n",
    "\n",
    "  * A count matrix of shape 'number of samples' x 'number of genes', containing\n",
    "    read counts (non-negative integers),\n",
    "  * Clinical data (or \"column\" data) of shape 'number of samples' x\n",
    "    'number of variables', containing sample annotations that will be used\n",
    "    to split the data in cohorts.\n",
    "\n",
    "Both should be provided as [pandas dataframes](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.html).\n",
    "\n",
    ".. currentmodule:: pydeseq2\n",
    "\n",
    "To illustrate the required data format, we load a synthetic example dataset\n",
    "that may be\n",
    "obtained through PyDESeq2's API using :func:`utils.load_example_data`.\n",
    "You may replace it with your own dataset.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e097f22e-b31b-4300-9e21-6c972e0bb858",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:34:45.768162Z",
     "iopub.status.busy": "2023-05-22T12:34:45.766101Z",
     "iopub.status.idle": "2023-05-22T12:34:46.727317Z",
     "shell.execute_reply": "2023-05-22T12:34:46.726299Z",
     "shell.execute_reply.started": "2023-05-22T12:34:45.768134Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           gene1  gene2  gene3  gene4  gene5  gene6  gene7  gene8  gene9   \n",
      "sample1       12     21      4    130     18      0     16     54     49  \\\n",
      "sample2        1     44      2     63     11     10     70     32     57   \n",
      "sample3        4      4     11    180     21      3     28     34     65   \n",
      "sample4        1     10      2    100     44      9     28     16     33   \n",
      "sample5        1     11      6    135     16      2     32     29     31   \n",
      "...          ...    ...    ...    ...    ...    ...    ...    ...    ...   \n",
      "sample96       7     26      3     67     11      4     41     44     54   \n",
      "sample97       1     14      3     71     33      5     19     42     25   \n",
      "sample98      10     36      2     72     11      2     66     27     16   \n",
      "sample99      18     14      3     66     53     11     32     19     79   \n",
      "sample100     21      9      3     42     13     13     19     78     30   \n",
      "\n",
      "           gene10  \n",
      "sample1         3  \n",
      "sample2         9  \n",
      "sample3         2  \n",
      "sample4         9  \n",
      "sample5         5  \n",
      "...           ...  \n",
      "sample96        1  \n",
      "sample97        4  \n",
      "sample98        9  \n",
      "sample99       11  \n",
      "sample100       5  \n",
      "\n",
      "[100 rows x 10 columns]\n",
      "          condition group\n",
      "sample1           A     X\n",
      "sample2           A     Y\n",
      "sample3           A     X\n",
      "sample4           A     Y\n",
      "sample5           A     X\n",
      "...             ...   ...\n",
      "sample96          B     Y\n",
      "sample97          B     X\n",
      "sample98          B     Y\n",
      "sample99          B     X\n",
      "sample100         B     Y\n",
      "\n",
      "[100 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "# counts_df = load_example_data(\n",
    "#     modality=\"raw_counts\",\n",
    "#     dataset=\"synthetic\",\n",
    "#     debug=True,\n",
    "# )\n",
    "\n",
    "# clinical_df = load_example_data(\n",
    "#     modality=\"clinical\",\n",
    "#     dataset=\"synthetic\",\n",
    "#     debug=True,\n",
    "# )\n",
    "counts_df = pd.read_csv('test_counts.csv', index_col=0).T\n",
    "clinical_df = pd.read_csv('test_clinical.csv', index_col=0)\n",
    "print(counts_df)\n",
    "print(clinical_df)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5568dc86-d67d-4bb6-a817-aa0dba70151a",
   "metadata": {},
   "source": [
    "## 1. Read counts modeling\n",
    "Read counts modeling with the :class:`DeseqDataSet\n",
    "<dds.DeseqDataSet>` class\n",
    "\n",
    "The :class:`DeseqDataSet <dds.DeseqDataSet>` class has two mandatory\n",
    "arguments, `counts_df` and\n",
    "`clinical_df`, as well as a set of optional keyword arguments, among which:\n",
    "\n",
    "- `design_factor`: the name of the column of clinical to be used as a design\n",
    "  variable\n",
    "- `refit_cooks`: whether to refit cooks outliers – this is advised, in general.\n",
    "\n",
    "<div class=\"alert alert-info\"><h4>Note</h4><p>in the case of the provided synthetic data, there won't be any Cooks\n",
    "  outliers.</p></div>\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "724df854-4248-4b67-a591-ee882c256077",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:34:48.630196Z",
     "iopub.status.busy": "2023-05-22T12:34:48.628860Z",
     "iopub.status.idle": "2023-05-22T12:34:49.017778Z",
     "shell.execute_reply": "2023-05-22T12:34:49.005940Z",
     "shell.execute_reply.started": "2023-05-22T12:34:48.630167Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "dds = DeseqDataSet(\n",
    "    counts=counts_df,\n",
    "    clinical=clinical_df,\n",
    "    design_factors=\"condition\",  # compare samples based on the \"condition\"\n",
    "    # column (\"B\" vs \"A\")\n",
    "    refit_cooks=True,\n",
    "    n_cpus=8,\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c839b1e4-7a71-4047-adc5-1667cca89289",
   "metadata": {},
   "source": [
    "### Compute normalization factors\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5cf8ba67-5bef-4f17-b8b8-83f6bb3b6c87",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:34:54.295995Z",
     "iopub.status.busy": "2023-05-22T12:34:54.292928Z",
     "iopub.status.idle": "2023-05-22T12:34:54.417549Z",
     "shell.execute_reply": "2023-05-22T12:34:54.415942Z",
     "shell.execute_reply.started": "2023-05-22T12:34:54.295961Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting size factors...\n",
      "... done in 0.04 seconds.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([1.22898094, 1.18877375, 0.99722229, 1.00215773, 0.83457743,\n",
       "       1.10730382, 0.8999001 , 1.15343785, 0.68163849, 1.29764537,\n",
       "       1.04491511, 1.45930946, 1.14588441, 0.8049275 , 0.88402672,\n",
       "       0.88402672, 1.32879767, 0.82564657, 1.5978062 , 1.29764537,\n",
       "       1.31940196, 0.69919197, 1.10697146, 1.10214803, 1.19152118,\n",
       "       1.0624452 , 0.98548229, 0.76881428, 0.8939601 , 1.27135863,\n",
       "       1.61101905, 1.55084302, 0.83601298, 0.98213727, 1.27270212,\n",
       "       1.0510719 , 1.76078144, 1.08132885, 1.50390106, 1.0510719 ,\n",
       "       0.80280751, 0.70955247, 1.32602392, 0.98031899, 1.1078077 ,\n",
       "       0.68792508, 0.90429564, 1.56411155, 0.81918767, 1.19364837,\n",
       "       0.79492024, 1.84963565, 0.79694628, 0.79708276, 0.97287297,\n",
       "       1.16248554, 1.50489413, 1.41929759, 1.04612122, 1.05720226,\n",
       "       0.99635345, 1.84224912, 1.03801163, 0.89633874, 0.72952979,\n",
       "       1.33453944, 0.93968061, 1.14016425, 1.59166589, 1.08554239,\n",
       "       0.72370261, 0.91558563, 1.14183629, 1.33857618, 0.94450599,\n",
       "       0.85266438, 1.38005658, 1.01803293, 1.04472988, 1.25372879,\n",
       "       1.77488931, 0.92641092, 1.06341062, 1.05653145, 0.61973175,\n",
       "       0.75569423, 0.69781308, 1.38512317, 0.82165798, 0.81330537,\n",
       "       0.96864497, 1.08526199, 0.78205578, 0.98864287, 1.15322141,\n",
       "       1.26935914, 0.74399805, 0.73987825, 0.95887987, 0.63570796])"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dds.fit_size_factors()\n",
    "\n",
    "dds.obsm[\"size_factors\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1014bf8a-a6c2-4b24-84fc-e0432a758ddb",
   "metadata": {},
   "source": [
    "### Fit genewise dispersions\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "85e8ae3e-59b3-43ce-b5f8-3589f3a6a9fb",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:34:56.700604Z",
     "iopub.status.busy": "2023-05-22T12:34:56.699556Z",
     "iopub.status.idle": "2023-05-22T12:35:05.989764Z",
     "shell.execute_reply": "2023-05-22T12:35:05.988774Z",
     "shell.execute_reply.started": "2023-05-22T12:34:56.700574Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting dispersions...\n",
      "... done in 0.13 seconds.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([0.91013511, 0.21371805, 0.81359443, 0.1613621 , 0.24850424,\n",
       "       0.97307734, 0.23303023, 0.19810261, 0.18363604, 0.64637486])"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dds.fit_genewise_dispersions()\n",
    "\n",
    "dds.varm[\"genewise_dispersions\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "85a560c4-8ce9-4900-b861-8371fb431215",
   "metadata": {},
   "source": [
    "### Fit dispersion trend coefficients\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "1528f72e-8f4a-4f77-bcf0-d86a21e307e4",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:35:07.678175Z",
     "iopub.status.busy": "2023-05-22T12:35:07.661258Z",
     "iopub.status.idle": "2023-05-22T12:35:08.073708Z",
     "shell.execute_reply": "2023-05-22T12:35:08.072821Z",
     "shell.execute_reply.started": "2023-05-22T12:35:07.678122Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting dispersion trend curve...\n",
      "... done in 0.38 seconds.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([0.65142791, 0.31300248, 1.04987097, 0.13414632, 0.26400662,\n",
       "       0.97813348, 0.25676617, 0.20575176, 0.2160277 , 0.50274843])"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dds.fit_dispersion_trend()\n",
    "dds.uns[\"trend_coeffs\"]\n",
    "dds.varm[\"fitted_dispersions\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c5143f6-c3cb-4b06-905a-6d68ee984826",
   "metadata": {},
   "source": [
    "### Dispersion priors\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "730ce65b-4596-46fc-88ad-1058cfe6a38a",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:35:09.428366Z",
     "iopub.status.busy": "2023-05-22T12:35:09.426526Z",
     "iopub.status.idle": "2023-05-22T12:35:09.568233Z",
     "shell.execute_reply": "2023-05-22T12:35:09.566933Z",
     "shell.execute_reply.started": "2023-05-22T12:35:09.428334Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "logres_prior=0.055924632043151165, sigma_prior=0.25\n"
     ]
    }
   ],
   "source": [
    "dds.fit_dispersion_prior()\n",
    "print(\n",
    "    f\"logres_prior={dds.uns['_squared_logres']}, sigma_prior={dds.uns['prior_disp_var']}\"\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a019e1e8-707c-425d-956c-8b114a776774",
   "metadata": {},
   "source": [
    "### MAP Dispersions\n",
    "The `fit_MAP_dispersions` method filters the genes for which dispersion\n",
    "shrinkage is applied.\n",
    "Indeed, for genes whose MLE dispersions are too high above the trend curve,\n",
    "the original MLE value is kept.\n",
    "The final values of the dispersions that are used for downstream analysis is\n",
    "stored in `dds.dispersions`.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "81e88c7b-5518-43d9-95b9-474f76f0390f",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:35:12.589062Z",
     "iopub.status.busy": "2023-05-22T12:35:12.581153Z",
     "iopub.status.idle": "2023-05-22T12:35:12.683478Z",
     "shell.execute_reply": "2023-05-22T12:35:12.682793Z",
     "shell.execute_reply.started": "2023-05-22T12:35:12.588953Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting MAP dispersions...\n",
      "... done in 0.05 seconds.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "array([0.88259868, 0.22257863, 0.83723801, 0.15897048, 0.24992589,\n",
       "       0.97364795, 0.23515487, 0.19878078, 0.1865203 , 0.63189989])"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dds.fit_MAP_dispersions()\n",
    "dds.varm[\"MAP_dispersions\"]\n",
    "dds.varm[\"dispersions\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "254a36e9-5874-4f3c-a86e-beaf40f28ea1",
   "metadata": {},
   "source": [
    "### Fit log fold changes\n",
    "Note that in the `DeseqDataSet` object, the log-fold changes are stored in\n",
    "natural\n",
    "log scale, but that the results dataframe output by the `summary` method of\n",
    "`DeseqStats` displays LFCs in log2 scale (see later on).\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "d1bfde56-33ff-484a-95d8-5775919e4b21",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:35:14.177848Z",
     "iopub.status.busy": "2023-05-22T12:35:14.176555Z",
     "iopub.status.idle": "2023-05-22T12:35:14.983525Z",
     "shell.execute_reply": "2023-05-22T12:35:14.982617Z",
     "shell.execute_reply.started": "2023-05-22T12:35:14.177821Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting LFCs...\n",
      "... done in 0.45 seconds.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>intercept</th>\n",
       "      <th>condition_B_vs_A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene1</th>\n",
       "      <td>1.891436</td>\n",
       "      <td>0.438632</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene2</th>\n",
       "      <td>2.851662</td>\n",
       "      <td>0.373296</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene3</th>\n",
       "      <td>1.787780</td>\n",
       "      <td>-0.438645</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene4</th>\n",
       "      <td>4.741958</td>\n",
       "      <td>-0.285647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene5</th>\n",
       "      <td>3.077798</td>\n",
       "      <td>0.403457</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene6</th>\n",
       "      <td>1.678536</td>\n",
       "      <td>0.001010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene7</th>\n",
       "      <td>3.291025</td>\n",
       "      <td>0.093116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene8</th>\n",
       "      <td>3.785129</td>\n",
       "      <td>-0.187604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene9</th>\n",
       "      <td>3.682882</td>\n",
       "      <td>-0.147443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene10</th>\n",
       "      <td>2.300515</td>\n",
       "      <td>0.267562</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        intercept  condition_B_vs_A\n",
       "gene1    1.891436          0.438632\n",
       "gene2    2.851662          0.373296\n",
       "gene3    1.787780         -0.438645\n",
       "gene4    4.741958         -0.285647\n",
       "gene5    3.077798          0.403457\n",
       "gene6    1.678536          0.001010\n",
       "gene7    3.291025          0.093116\n",
       "gene8    3.785129         -0.187604\n",
       "gene9    3.682882         -0.147443\n",
       "gene10   2.300515          0.267562"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dds.fit_LFC()\n",
    "dds.varm[\"LFC\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "631e085a-af63-4c7f-a762-196ab9ab4785",
   "metadata": {},
   "source": [
    "### Calculate Cooks distances and refit\n",
    "Note that this step is optional.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "84e0c1a5-202b-4752-a109-067087aa455c",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:35:16.963891Z",
     "iopub.status.busy": "2023-05-22T12:35:16.962687Z",
     "iopub.status.idle": "2023-05-22T12:35:17.067580Z",
     "shell.execute_reply": "2023-05-22T12:35:17.066859Z",
     "shell.execute_reply.started": "2023-05-22T12:35:16.963866Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Refitting 0 outliers.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "dds.calculate_cooks()\n",
    "if dds.refit_cooks:\n",
    "    # Replace outlier counts\n",
    "    dds.refit()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0691de3e-db14-484d-9d26-99373d5bd3cf",
   "metadata": {},
   "source": [
    "Save everything\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "dfc23026-9f58-4f1f-9808-491f534562f5",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:35:19.667978Z",
     "iopub.status.busy": "2023-05-22T12:35:19.663341Z",
     "iopub.status.idle": "2023-05-22T12:35:19.678667Z",
     "shell.execute_reply": "2023-05-22T12:35:19.677955Z",
     "shell.execute_reply.started": "2023-05-22T12:35:19.667948Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "if SAVE:\n",
    "    with open(os.path.join(OUTPUT_PATH, \"dds_detailed_pipe.pkl\"), \"wb\") as f:\n",
    "        pkl.dump(dds, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c90f3fcc-5bb4-469b-be29-d793d80129db",
   "metadata": {},
   "source": [
    "## 2. Statistical analysis\n",
    "Statistical analysis with the :class:`DeseqStats <ds.DeseqStats>` class.\n",
    "The `DeseqDataSet` class has a unique mandatory arguments, `dds`, which should\n",
    "be a *fitted* `DeseqDataSet` object, as well as a set of optional keyword\n",
    "arguments, among which:\n",
    "\n",
    "- `alpha`: the p-value and adjusted p-value significance threshold\n",
    "- `cooks_filter`: whether to filter p-values based on cooks outliers\n",
    "- `independent_filter`: whether to perform independent filtering to correct\n",
    "  p-value trends.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a6eef388-1107-4ef4-9efc-511856aec413",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:35:21.962498Z",
     "iopub.status.busy": "2023-05-22T12:35:21.960845Z",
     "iopub.status.idle": "2023-05-22T12:35:22.016235Z",
     "shell.execute_reply": "2023-05-22T12:35:22.015438Z",
     "shell.execute_reply.started": "2023-05-22T12:35:21.962470Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "stat_res = DeseqStats(dds, alpha=0.05, cooks_filter=True, independent_filter=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f64ab30a-699c-4fb9-851f-aed8c111b997",
   "metadata": {},
   "source": [
    "### Wald tests\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "96c6a128-ce6f-47f3-ab22-9627250f9f77",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:35:24.495610Z",
     "iopub.status.busy": "2023-05-22T12:35:24.494121Z",
     "iopub.status.idle": "2023-05-22T12:35:24.935177Z",
     "shell.execute_reply": "2023-05-22T12:35:24.933968Z",
     "shell.execute_reply.started": "2023-05-22T12:35:24.495585Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Running Wald tests...\n",
      "... done in 0.29 seconds.\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "gene1     0.028604\n",
       "gene2     0.000329\n",
       "gene3     0.032075\n",
       "gene4     0.000513\n",
       "gene5     0.000168\n",
       "gene6     0.996253\n",
       "gene7     0.370297\n",
       "gene8     0.047227\n",
       "gene9     0.110391\n",
       "gene10    0.114518\n",
       "dtype: float64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "stat_res.run_wald_test()\n",
    "stat_res.p_values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "414c4be4-3c58-44a6-b5e6-e588a818b39c",
   "metadata": {},
   "source": [
    "### Cooks filtering\n",
    "This is optional\n",
    "\n",
    "<div class=\"alert alert-info\"><h4>Note</h4><p>Note: in the case of the provided synthetic data, there won't be any\n",
    "  outliers.</p></div>\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "1e141954-e5d5-4be0-b6e8-0bce2bf08cec",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:35:28.476537Z",
     "iopub.status.busy": "2023-05-22T12:35:28.475367Z",
     "iopub.status.idle": "2023-05-22T12:35:28.568563Z",
     "shell.execute_reply": "2023-05-22T12:35:28.567906Z",
     "shell.execute_reply.started": "2023-05-22T12:35:28.476506Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene1     0.028604\n",
       "gene2     0.000329\n",
       "gene3     0.032075\n",
       "gene4     0.000513\n",
       "gene5     0.000168\n",
       "gene6     0.996253\n",
       "gene7     0.370297\n",
       "gene8     0.047227\n",
       "gene9     0.110391\n",
       "gene10    0.114518\n",
       "dtype: float64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if stat_res.cooks_filter:\n",
    "    stat_res._cooks_filtering()\n",
    "stat_res.p_values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36115f6e-261d-4118-8ca7-73077aac2184",
   "metadata": {},
   "source": [
    "### P-value adjustment\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "6d8c5192-456a-4770-aef0-43f75e0c3fba",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:35:30.841750Z",
     "iopub.status.busy": "2023-05-22T12:35:30.828896Z",
     "iopub.status.idle": "2023-05-22T12:35:31.128164Z",
     "shell.execute_reply": "2023-05-22T12:35:31.127135Z",
     "shell.execute_reply.started": "2023-05-22T12:35:30.841708Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "gene1     0.064150\n",
       "gene2     0.001646\n",
       "gene3     0.064150\n",
       "gene4     0.001710\n",
       "gene5     0.001646\n",
       "gene6     0.996253\n",
       "gene7     0.411441\n",
       "gene8     0.078711\n",
       "gene9     0.143148\n",
       "gene10    0.143148\n",
       "Name: 0, dtype: float64"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if stat_res.independent_filter:\n",
    "    stat_res._independent_filtering()\n",
    "else:\n",
    "    stat_res._p_value_adjustment()\n",
    "\n",
    "stat_res.padj"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6062b13-33b9-411d-b987-d4a21d295b17",
   "metadata": {},
   "source": [
    "### Building a results dataframe\n",
    "This dataframe is stored in the `results_df` attribute of the `DeseqStats`\n",
    "class.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e80f939b-fba6-4f5a-9015-483bb3c4f520",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:35:33.599743Z",
     "iopub.status.busy": "2023-05-22T12:35:33.597763Z",
     "iopub.status.idle": "2023-05-22T12:35:33.859919Z",
     "shell.execute_reply": "2023-05-22T12:35:33.858157Z",
     "shell.execute_reply.started": "2023-05-22T12:35:33.599715Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Log2 fold change & Wald test p-value: condition B vs A\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene1</th>\n",
       "      <td>8.541317</td>\n",
       "      <td>0.632812</td>\n",
       "      <td>0.289101</td>\n",
       "      <td>2.188898</td>\n",
       "      <td>0.028604</td>\n",
       "      <td>0.064150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene2</th>\n",
       "      <td>21.281239</td>\n",
       "      <td>0.538552</td>\n",
       "      <td>0.149963</td>\n",
       "      <td>3.591235</td>\n",
       "      <td>0.000329</td>\n",
       "      <td>0.001646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene3</th>\n",
       "      <td>5.010123</td>\n",
       "      <td>-0.632830</td>\n",
       "      <td>0.295236</td>\n",
       "      <td>-2.143475</td>\n",
       "      <td>0.032075</td>\n",
       "      <td>0.064150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene4</th>\n",
       "      <td>100.517961</td>\n",
       "      <td>-0.412102</td>\n",
       "      <td>0.118629</td>\n",
       "      <td>-3.473867</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.001710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene5</th>\n",
       "      <td>27.142450</td>\n",
       "      <td>0.582065</td>\n",
       "      <td>0.154706</td>\n",
       "      <td>3.762408</td>\n",
       "      <td>0.000168</td>\n",
       "      <td>0.001646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene6</th>\n",
       "      <td>5.413043</td>\n",
       "      <td>0.001457</td>\n",
       "      <td>0.310311</td>\n",
       "      <td>0.004696</td>\n",
       "      <td>0.996253</td>\n",
       "      <td>0.996253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene7</th>\n",
       "      <td>28.294023</td>\n",
       "      <td>0.134338</td>\n",
       "      <td>0.149945</td>\n",
       "      <td>0.895917</td>\n",
       "      <td>0.370297</td>\n",
       "      <td>0.411441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene8</th>\n",
       "      <td>40.358344</td>\n",
       "      <td>-0.270656</td>\n",
       "      <td>0.136401</td>\n",
       "      <td>-1.984260</td>\n",
       "      <td>0.047227</td>\n",
       "      <td>0.078711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene9</th>\n",
       "      <td>37.166183</td>\n",
       "      <td>-0.212715</td>\n",
       "      <td>0.133243</td>\n",
       "      <td>-1.596437</td>\n",
       "      <td>0.110391</td>\n",
       "      <td>0.143148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene10</th>\n",
       "      <td>11.589325</td>\n",
       "      <td>0.386011</td>\n",
       "      <td>0.244588</td>\n",
       "      <td>1.578207</td>\n",
       "      <td>0.114518</td>\n",
       "      <td>0.143148</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
       "gene1     8.541317        0.632812  0.289101  2.188898  0.028604  0.064150\n",
       "gene2    21.281239        0.538552  0.149963  3.591235  0.000329  0.001646\n",
       "gene3     5.010123       -0.632830  0.295236 -2.143475  0.032075  0.064150\n",
       "gene4   100.517961       -0.412102  0.118629 -3.473867  0.000513  0.001710\n",
       "gene5    27.142450        0.582065  0.154706  3.762408  0.000168  0.001646\n",
       "gene6     5.413043        0.001457  0.310311  0.004696  0.996253  0.996253\n",
       "gene7    28.294023        0.134338  0.149945  0.895917  0.370297  0.411441\n",
       "gene8    40.358344       -0.270656  0.136401 -1.984260  0.047227  0.078711\n",
       "gene9    37.166183       -0.212715  0.133243 -1.596437  0.110391  0.143148\n",
       "gene10   11.589325        0.386011  0.244588  1.578207  0.114518  0.143148"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "stat_res.summary()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1d48d222-2202-4235-b4bf-43765f6a7d49",
   "metadata": {},
   "source": [
    "Save everything if SAVE is set to True\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "293eb424-3a21-461d-9659-09377a2ceb53",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:35:39.298344Z",
     "iopub.status.busy": "2023-05-22T12:35:39.296946Z",
     "iopub.status.idle": "2023-05-22T12:35:39.308503Z",
     "shell.execute_reply": "2023-05-22T12:35:39.307785Z",
     "shell.execute_reply.started": "2023-05-22T12:35:39.298318Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "if SAVE:\n",
    "    with open(os.path.join(OUTPUT_PATH, \"stat_results_detailed_pipe.pkl\"), \"wb\") as f:\n",
    "        pkl.dump(stat_res, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f94f360-d421-4866-807d-e24e0501bafd",
   "metadata": {},
   "source": [
    "### LFC Shrinkage\n",
    "For visualization or post-processing purposes, it might be suitable to perform\n",
    "LFC shrinkage. This is implemented by the `lfc_shrink` method.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ca8e460d-b1c6-4ccb-b0b8-47c43ea0a1e4",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:35:40.715010Z",
     "iopub.status.busy": "2023-05-22T12:35:40.714688Z",
     "iopub.status.idle": "2023-05-22T12:35:43.080781Z",
     "shell.execute_reply": "2023-05-22T12:35:43.080108Z",
     "shell.execute_reply.started": "2023-05-22T12:35:40.714986Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fitting MAP LFCs...\n",
      "... done in 2.29 seconds.\n",
      "\n",
      "Shrunk Log2 fold change & Wald test p-value: condition B vs A\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene1</th>\n",
       "      <td>8.541317</td>\n",
       "      <td>0.408253</td>\n",
       "      <td>0.294276</td>\n",
       "      <td>2.188898</td>\n",
       "      <td>0.028604</td>\n",
       "      <td>0.064150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene2</th>\n",
       "      <td>21.281239</td>\n",
       "      <td>0.480145</td>\n",
       "      <td>0.151201</td>\n",
       "      <td>3.591235</td>\n",
       "      <td>0.000329</td>\n",
       "      <td>0.001646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene3</th>\n",
       "      <td>5.010123</td>\n",
       "      <td>-0.396066</td>\n",
       "      <td>0.300796</td>\n",
       "      <td>-2.143475</td>\n",
       "      <td>0.032075</td>\n",
       "      <td>0.064150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene4</th>\n",
       "      <td>100.517961</td>\n",
       "      <td>-0.374191</td>\n",
       "      <td>0.118704</td>\n",
       "      <td>-3.473867</td>\n",
       "      <td>0.000513</td>\n",
       "      <td>0.001710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene5</th>\n",
       "      <td>27.142450</td>\n",
       "      <td>0.521487</td>\n",
       "      <td>0.156210</td>\n",
       "      <td>3.762408</td>\n",
       "      <td>0.000168</td>\n",
       "      <td>0.001646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene6</th>\n",
       "      <td>5.413043</td>\n",
       "      <td>0.000716</td>\n",
       "      <td>0.239203</td>\n",
       "      <td>0.004696</td>\n",
       "      <td>0.996253</td>\n",
       "      <td>0.996253</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene7</th>\n",
       "      <td>28.294023</td>\n",
       "      <td>0.103421</td>\n",
       "      <td>0.141496</td>\n",
       "      <td>0.895917</td>\n",
       "      <td>0.370297</td>\n",
       "      <td>0.411441</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene8</th>\n",
       "      <td>40.358344</td>\n",
       "      <td>-0.226288</td>\n",
       "      <td>0.133477</td>\n",
       "      <td>-1.984260</td>\n",
       "      <td>0.047227</td>\n",
       "      <td>0.078711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene9</th>\n",
       "      <td>37.166183</td>\n",
       "      <td>-0.175746</td>\n",
       "      <td>0.129138</td>\n",
       "      <td>-1.596437</td>\n",
       "      <td>0.110391</td>\n",
       "      <td>0.143148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene10</th>\n",
       "      <td>11.589325</td>\n",
       "      <td>0.239935</td>\n",
       "      <td>0.231986</td>\n",
       "      <td>1.578207</td>\n",
       "      <td>0.114518</td>\n",
       "      <td>0.143148</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          baseMean  log2FoldChange     lfcSE      stat    pvalue      padj\n",
       "gene1     8.541317        0.408253  0.294276  2.188898  0.028604  0.064150\n",
       "gene2    21.281239        0.480145  0.151201  3.591235  0.000329  0.001646\n",
       "gene3     5.010123       -0.396066  0.300796 -2.143475  0.032075  0.064150\n",
       "gene4   100.517961       -0.374191  0.118704 -3.473867  0.000513  0.001710\n",
       "gene5    27.142450        0.521487  0.156210  3.762408  0.000168  0.001646\n",
       "gene6     5.413043        0.000716  0.239203  0.004696  0.996253  0.996253\n",
       "gene7    28.294023        0.103421  0.141496  0.895917  0.370297  0.411441\n",
       "gene8    40.358344       -0.226288  0.133477 -1.984260  0.047227  0.078711\n",
       "gene9    37.166183       -0.175746  0.129138 -1.596437  0.110391  0.143148\n",
       "gene10   11.589325        0.239935  0.231986  1.578207  0.114518  0.143148"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "stat_res.lfc_shrink(coeff=\"condition_B_vs_A\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9738d9ec-6b5d-422b-9796-d270ced08d86",
   "metadata": {},
   "source": [
    "Save everything\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "624e22a2-570c-4c83-8cc5-058221f2a0e1",
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2023-05-22T12:35:43.082342Z",
     "iopub.status.busy": "2023-05-22T12:35:43.081970Z",
     "iopub.status.idle": "2023-05-22T12:35:43.086341Z",
     "shell.execute_reply": "2023-05-22T12:35:43.085625Z",
     "shell.execute_reply.started": "2023-05-22T12:35:43.082318Z"
    },
    "jupyter": {
     "outputs_hidden": false
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "if SAVE:\n",
    "    with open(\n",
    "        os.path.join(OUTPUT_PATH, \"shrunk_stat_results_detailed_pipe.pkl\"), \"wb\"\n",
    "    ) as f:\n",
    "        pkl.dump(stat_res, f)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "snakepipes_ATAC-seq",
   "language": "python",
   "name": "snakepipes_atac-seq"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.8"
  },
  "toc-autonumbering": true
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
